// @glideapps/glide-data-grid/dist/js/data-editor/use-cell-sizer.js
import * as React from "react";
import { CellRenderers } from "../data-grid/cells";
import { GridCellKind, isSizedGridColumn } from "../data-grid/data-grid-types";
const defaultSize = 150;

function measureCell(ctx, cell) {
  var _r$measure;

  if (cell.kind === GridCellKind.Custom) return defaultSize;
  const r = CellRenderers[cell.kind];
  return (_r$measure = r === null || r === void 0 ? void 0 : r.measure(ctx, cell)) !== null && _r$measure !== void 0 ? _r$measure : defaultSize;
}

export function useCellSizer(columns, rows, getCellsForSelection, theme) {
  const rowsRef = React.useRef(rows);
  const getCellsForSelectionRef = React.useRef(getCellsForSelection);
  const themeRef = React.useRef(theme);
  rowsRef.current = rows;
  getCellsForSelectionRef.current = getCellsForSelection;
  themeRef.current = theme;
  const [ctx] = React.useState(() => {
    const offscreen = document.createElement("canvas");
    return offscreen.getContext("2d", {
      alpha: false
    });
  });
  const memoMap = React.useRef({});
  return React.useMemo(() => {
    var _getCellsForSelection;

    if (columns.every(isSizedGridColumn)) {
      return columns;
    }

    if (ctx === null) {
      return columns.map(c => {
        if (isSizedGridColumn(c)) return c;
        return { ...c,
          width: defaultSize
        };
      });
    }

    ctx.font = `${themeRef.current.baseFontStyle} ${themeRef.current.fontFamily}`;
    const computeRows = Math.max(1, 10 - Math.floor(columns.length / 10000));
    const cells = (_getCellsForSelection = getCellsForSelectionRef.current) === null || _getCellsForSelection === void 0 ? void 0 : _getCellsForSelection.call(getCellsForSelectionRef, {
      x: 0,
      y: 0,
      width: columns.length,
      height: Math.min(rowsRef.current, computeRows)
    });
    return columns.map((c, colIndex) => {
      if (isSizedGridColumn(c)) return c;

      if (cells === undefined || c.id === undefined) {
        return { ...c,
          width: defaultSize
        };
      }

      if (memoMap.current[c.id] !== undefined) {
        return { ...c,
          width: memoMap.current[c.id]
        };
      }

      const sizes = cells.map(row => row[colIndex]).map(cell => measureCell(ctx, cell));
      sizes.push(ctx.measureText(c.title).width + 16 + (c.icon === undefined ? 0 : 28));
      const average = sizes.reduce((a, b) => a + b) / sizes.length;
      const biggest = sizes.reduce((a, acc) => a > average * 2 ? acc : Math.max(acc, a));
      const final = Math.min(500, Math.ceil(biggest));
      memoMap.current[c.id] = final;
      return { ...c,
        width: final
      };
    });
  }, [columns, ctx]);
}
