// .ignored/@labbs/openblock-react/dist/index.js
import { jsx, jsxs, Fragment } from "react/jsx-runtime";
import React, { useState, useRef, useEffect, useMemo, forwardRef, useImperativeHandle, useCallback, useLayoutEffect } from "react";
import { OpenBlockEditor, getDefaultSlashMenuItems, filterSlashMenuItems, executeSlashCommand, SLASH_MENU_PLUGIN_KEY, closeSlashMenu, BUBBLE_MENU_PLUGIN_KEY, addRowBefore, addRowAfter, deleteRow, addColumnBefore, addColumnAfter, deleteColumn, deleteTable, isInTable, getTableInfo, addRowAtIndex, deleteRowAtIndex, addColumnAtIndex, deleteColumnAtIndex, updateMediaAttrs, deleteMediaNode, MEDIA_MENU_PLUGIN_KEY } from "@labbs/openblock-core";
import { createRoot } from "react-dom/client";
function useOpenBlock(options = {}) {
  const [editor, setEditor] = useState(null);
  const optionsRef = useRef(options);
  const editorRef = useRef(null);
  useEffect(() => {
    var _a;
    const { customBlocks, ...editorOptions } = optionsRef.current;
    const nodeViews = {};
    if (customBlocks && customBlocks.length > 0) {
      for (const blockSpec of customBlocks) {
        nodeViews[blockSpec.type] = (node, view, getPos, decorations, innerDecorations) => {
          const editor2 = editorRef.current;
          if (!editor2) {
            throw new Error("Editor not initialized");
          }
          const nodeViewConstructor = blockSpec.createNodeView(editor2);
          return nodeViewConstructor(node, view, getPos, decorations, innerDecorations);
        };
      }
    }
    const prosemirrorConfig = {
      ...editorOptions.prosemirror,
      nodeViews: {
        ...(_a = editorOptions.prosemirror) == null ? void 0 : _a.nodeViews,
        ...nodeViews
      }
    };
    const newEditor = new OpenBlockEditor({
      ...editorOptions,
      prosemirror: Object.keys(nodeViews).length > 0 ? prosemirrorConfig : editorOptions.prosemirror
    });
    editorRef.current = newEditor;
    setEditor(newEditor);
    return () => {
      newEditor.destroy();
      editorRef.current = null;
    };
  }, []);
  return editor;
}
function useEditorContent(editor) {
  const [blocks, setBlocks] = useState([]);
  useEffect(() => {
    if (!editor || editor.isDestroyed) return;
    setBlocks(editor.getDocument());
    const unsubscribe = editor.on("change", ({ blocks: blocks2 }) => {
      setBlocks(blocks2);
    });
    return unsubscribe;
  }, [editor]);
  return blocks;
}
function useEditorSelection(editor) {
  const [selected, setSelected] = useState([]);
  useEffect(() => {
    if (!editor || editor.isDestroyed) return;
    setSelected(editor.getSelectedBlocks());
    const unsubscribe = editor.on("selectionChange", ({ blocks }) => {
      setSelected(blocks);
    });
    return unsubscribe;
  }, [editor]);
  return selected;
}
function useEditorFocus(editor) {
  const [focused, setFocused] = useState(false);
  useEffect(() => {
    if (!editor || editor.isDestroyed) return;
    setFocused(editor.hasFocus);
    const unsubscribeFocus = editor.on("focus", () => setFocused(true));
    const unsubscribeBlur = editor.on("blur", () => setFocused(false));
    return () => {
      unsubscribeFocus();
      unsubscribeBlur();
    };
  }, [editor]);
  return focused;
}
function useCustomSlashMenuItems(editor, customBlocks) {
  return useMemo(() => {
    if (!editor || editor.isDestroyed) return [];
    return customBlocks.filter(
      (block) => block.slashMenu !== void 0
    ).map((block) => {
      const { slashMenu, type, propSchema } = block;
      return {
        id: type,
        title: slashMenu.title,
        description: slashMenu.description,
        icon: slashMenu.icon,
        keywords: slashMenu.aliases,
        group: slashMenu.group || "Custom",
        action: (view) => {
          const attrs = {};
          for (const [key, value] of Object.entries(propSchema)) {
            attrs[key] = value.default;
          }
          const nodeType = view.state.schema.nodes[type];
          if (nodeType) {
            const node = nodeType.create(attrs);
            view.dispatch(view.state.tr.replaceSelectionWith(node));
          }
        }
      };
    });
  }, [editor, customBlocks]);
}
const OpenBlockView = forwardRef(
  function OpenBlockView2({ editor, className, style, children }, ref) {
    const containerRef = useRef(null);
    const mountedRef = useRef(false);
    useImperativeHandle(
      ref,
      () => ({
        container: containerRef.current,
        editor
      }),
      [editor]
    );
    useEffect(() => {
      const container = containerRef.current;
      if (!container || !editor || editor.isDestroyed) {
        return;
      }
      editor.mount(container);
      mountedRef.current = true;
    }, [editor]);
    return /* @__PURE__ */ jsx(
      "div",
      {
        ref: containerRef,
        className: className ? `openblock-container ${className}` : "openblock-container",
        style: {
          position: "relative",
          ...style
        },
        children
      }
    );
  }
);
OpenBlockView.displayName = "OpenBlockView";
const Icons$2 = {
  heading1: ({ className }) => /* @__PURE__ */ jsx("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M4 12h8M4 6v12M12 6v12M17 12l3-2v8" }) }),
  heading2: ({ className }) => /* @__PURE__ */ jsx("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M4 12h8M4 6v12M12 6v12M17 10c1.5-1 3 0 3 2s-3 3-3 5h3" }) }),
  heading3: ({ className }) => /* @__PURE__ */ jsx("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M4 12h8M4 6v12M12 6v12M17 10c1.5-1 3 0 3 1.5c0 1-1 1.5-1.5 1.5c.5 0 1.5.5 1.5 1.5c0 1.5-1.5 2.5-3 1.5" }) }),
  list: ({ className }) => /* @__PURE__ */ jsx("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M8 6h13M8 12h13M8 18h13M3 6h.01M3 12h.01M3 18h.01" }) }),
  listOrdered: ({ className }) => /* @__PURE__ */ jsx("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M10 6h11M10 12h11M10 18h11M3 5v3h2M3 10v1c0 1 2 2 2 2s-2 1-2 2v1h4M3 17v4h2l2-2" }) }),
  quote: ({ className }) => /* @__PURE__ */ jsxs("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("path", { d: "M3 21c3 0 7-1 7-8V5c0-1.25-.756-2.017-2-2H4c-1.25 0-2 .75-2 1.972V11c0 1.25.75 2 2 2 1 0 1 0 1 1v1c0 1-1 2-2 2s-1 .008-1 1.031V21" }),
    /* @__PURE__ */ jsx("path", { d: "M15 21c3 0 7-1 7-8V5c0-1.25-.757-2.017-2-2h-4c-1.25 0-2 .75-2 1.972V11c0 1.25.75 2 2 2h.75c0 2.25.25 4-2.75 4v4" })
  ] }),
  code: ({ className }) => /* @__PURE__ */ jsx("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M16 18l6-6-6-6M8 6l-6 6 6 6" }) }),
  minus: ({ className }) => /* @__PURE__ */ jsx("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M5 12h14" }) }),
  image: ({ className }) => /* @__PURE__ */ jsxs("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("rect", { x: "3", y: "3", width: "18", height: "18", rx: "2", ry: "2" }),
    /* @__PURE__ */ jsx("circle", { cx: "8.5", cy: "8.5", r: "1.5" }),
    /* @__PURE__ */ jsx("polyline", { points: "21 15 16 10 5 21" })
  ] }),
  checkSquare: ({ className }) => /* @__PURE__ */ jsxs("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("polyline", { points: "9 11 12 14 22 4" }),
    /* @__PURE__ */ jsx("path", { d: "M21 12v7a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11" })
  ] }),
  callout: ({ className }) => /* @__PURE__ */ jsx("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M21 15a2 2 0 0 1-2 2H7l-4 4V5a2 2 0 0 1 2-2h14a2 2 0 0 1 2 2z" }) }),
  table: ({ className }) => /* @__PURE__ */ jsxs("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("rect", { x: "3", y: "3", width: "18", height: "18", rx: "2", ry: "2" }),
    /* @__PURE__ */ jsx("line", { x1: "3", y1: "9", x2: "21", y2: "9" }),
    /* @__PURE__ */ jsx("line", { x1: "3", y1: "15", x2: "21", y2: "15" }),
    /* @__PURE__ */ jsx("line", { x1: "9", y1: "3", x2: "9", y2: "21" }),
    /* @__PURE__ */ jsx("line", { x1: "15", y1: "3", x2: "15", y2: "21" })
  ] }),
  embed: ({ className }) => /* @__PURE__ */ jsxs("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("rect", { x: "2", y: "3", width: "20", height: "14", rx: "2", ry: "2" }),
    /* @__PURE__ */ jsx("line", { x1: "8", y1: "21", x2: "16", y2: "21" }),
    /* @__PURE__ */ jsx("line", { x1: "12", y1: "17", x2: "12", y2: "21" })
  ] }),
  youtube: ({ className }) => /* @__PURE__ */ jsxs("svg", { className, width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("path", { d: "M22.54 6.42a2.78 2.78 0 0 0-1.94-2C18.88 4 12 4 12 4s-6.88 0-8.6.46a2.78 2.78 0 0 0-1.94 2A29 29 0 0 0 1 11.75a29 29 0 0 0 .46 5.33A2.78 2.78 0 0 0 3.4 19c1.72.46 8.6.46 8.6.46s6.88 0 8.6-.46a2.78 2.78 0 0 0 1.94-2 29 29 0 0 0 .46-5.25 29 29 0 0 0-.46-5.33z" }),
    /* @__PURE__ */ jsx("polygon", { points: "9.75 15.02 15.5 11.75 9.75 8.48 9.75 15.02" })
  ] })
};
function SlashMenu({
  editor,
  items,
  customItems,
  itemOrder,
  hideItems,
  renderItem,
  className
}) {
  const [menuState, setMenuState] = useState(null);
  const [selectedIndex, setSelectedIndex] = useState(0);
  const [openUpward, setOpenUpward] = useState(false);
  const menuRef = useRef(null);
  const allItems = useMemo(() => {
    if (!editor) return [];
    if (items) {
      return items;
    }
    const defaultItems = getDefaultSlashMenuItems(editor.pm.state.schema);
    const mergedItems = customItems ? [...defaultItems, ...customItems] : defaultItems;
    const itemMap = /* @__PURE__ */ new Map();
    mergedItems.forEach((item) => itemMap.set(item.id, item));
    if (itemOrder) {
      const orderedItems = [];
      itemOrder.forEach((id) => {
        const item = itemMap.get(id);
        if (item) {
          orderedItems.push(item);
        }
      });
      return orderedItems;
    }
    if (hideItems && hideItems.length > 0) {
      const hideSet = new Set(hideItems);
      return mergedItems.filter((item) => !hideSet.has(item.id));
    }
    return mergedItems;
  }, [editor, items, customItems, itemOrder, hideItems]);
  const filteredItems = menuState ? filterSlashMenuItems(allItems, menuState.query) : [];
  useEffect(() => {
    if (!editor || editor.isDestroyed) return;
    const updateState = () => {
      const state = SLASH_MENU_PLUGIN_KEY.getState(editor.pm.state);
      setMenuState(state ?? null);
      setSelectedIndex(0);
    };
    updateState();
    const unsubscribe = editor.on("transaction", updateState);
    return unsubscribe;
  }, [editor]);
  useEffect(() => {
    if (!editor || editor.isDestroyed || !(menuState == null ? void 0 : menuState.active)) return;
    const handleKeyDown = (event) => {
      switch (event.key) {
        case "ArrowDown":
          event.preventDefault();
          event.stopPropagation();
          setSelectedIndex((i) => Math.min(i + 1, filteredItems.length - 1));
          break;
        case "ArrowUp":
          event.preventDefault();
          event.stopPropagation();
          setSelectedIndex((i) => Math.max(i - 1, 0));
          break;
        case "Enter":
          event.preventDefault();
          event.stopPropagation();
          if (filteredItems[selectedIndex]) {
            handleSelect(filteredItems[selectedIndex]);
          }
          break;
        case "Escape":
          event.preventDefault();
          event.stopPropagation();
          closeSlashMenu(editor.pm.view);
          break;
      }
    };
    const editorElement = editor.pm.view.dom;
    editorElement.addEventListener("keydown", handleKeyDown, true);
    return () => editorElement.removeEventListener("keydown", handleKeyDown, true);
  }, [menuState == null ? void 0 : menuState.active, filteredItems, selectedIndex, editor]);
  const handleSelect = useCallback(
    (item) => {
      if (!editor || editor.isDestroyed || !menuState) return;
      executeSlashCommand(editor.pm.view, menuState, item.action);
      editor.pm.view.focus();
    },
    [editor, menuState]
  );
  useLayoutEffect(() => {
    if (!(menuState == null ? void 0 : menuState.active) || !menuState.coords || !menuRef.current) {
      return;
    }
    const menuHeight = menuRef.current.offsetHeight || 300;
    const spaceBelow = window.innerHeight - menuState.coords.bottom - 8;
    const spaceAbove = menuState.coords.top - 8;
    setOpenUpward(spaceBelow < menuHeight && spaceAbove > spaceBelow);
  }, [menuState == null ? void 0 : menuState.active, menuState == null ? void 0 : menuState.coords, filteredItems.length]);
  if (!(menuState == null ? void 0 : menuState.active) || !menuState.coords) {
    return null;
  }
  const style = {
    position: "fixed",
    left: menuState.coords.left,
    zIndex: 1e3,
    ...openUpward ? { bottom: window.innerHeight - menuState.coords.top + 4 } : { top: menuState.coords.bottom + 4 }
  };
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: menuRef,
      className: `ob-slash-menu ${className || ""}`,
      style,
      role: "listbox",
      children: filteredItems.length === 0 ? /* @__PURE__ */ jsx("div", { className: "ob-slash-menu-empty", children: "No results" }) : filteredItems.map((item, index) => {
        const isSelected = index === selectedIndex;
        const Icon = item.icon ? Icons$2[item.icon] : null;
        if (renderItem) {
          return /* @__PURE__ */ jsx(
            "div",
            {
              onClick: () => handleSelect(item),
              role: "option",
              "aria-selected": isSelected,
              children: renderItem(item, isSelected)
            },
            item.id
          );
        }
        return /* @__PURE__ */ jsxs(
          "div",
          {
            className: `ob-slash-menu-item ${isSelected ? "ob-slash-menu-item--selected" : ""}`,
            onClick: () => handleSelect(item),
            onMouseEnter: () => setSelectedIndex(index),
            role: "option",
            "aria-selected": isSelected,
            children: [
              Icon && /* @__PURE__ */ jsx("span", { className: "ob-slash-menu-item-icon", children: /* @__PURE__ */ jsx(Icon, {}) }),
              /* @__PURE__ */ jsxs("div", { className: "ob-slash-menu-item-content", children: [
                /* @__PURE__ */ jsx("span", { className: "ob-slash-menu-item-title", children: item.title }),
                item.description && /* @__PURE__ */ jsx("span", { className: "ob-slash-menu-item-description", children: item.description })
              ] })
            ]
          },
          item.id
        );
      })
    }
  );
}
function isValidUrl(url) {
  if (!url.trim()) return false;
  const urlToTest = url.startsWith("http://") || url.startsWith("https://") ? url : `https://${url}`;
  try {
    new URL(urlToTest);
    return true;
  } catch {
    return false;
  }
}
function normalizeUrl(url) {
  const trimmed = url.trim();
  if (!trimmed) return "";
  if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
    return trimmed;
  }
  return `https://${trimmed}`;
}
function LinkPopover({
  editor,
  currentUrl,
  onClose,
  position,
  className,
  triggerRef
}) {
  const [url, setUrl] = useState(currentUrl || "");
  const [error, setError] = useState(null);
  const inputRef = useRef(null);
  const popoverRef = useRef(null);
  useEffect(() => {
    var _a, _b;
    (_a = inputRef.current) == null ? void 0 : _a.focus();
    (_b = inputRef.current) == null ? void 0 : _b.select();
  }, []);
  useEffect(() => {
    const handleClickOutside = (event) => {
      const target = event.target;
      if (popoverRef.current && popoverRef.current.contains(target)) {
        return;
      }
      if ((triggerRef == null ? void 0 : triggerRef.current) && triggerRef.current.contains(target)) {
        return;
      }
      onClose();
    };
    const frameId = requestAnimationFrame(() => {
      document.addEventListener("mousedown", handleClickOutside);
    });
    return () => {
      cancelAnimationFrame(frameId);
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [onClose, triggerRef]);
  useEffect(() => {
    const handleKeyDown = (event) => {
      if (event.key === "Escape") {
        onClose();
      }
    };
    document.addEventListener("keydown", handleKeyDown);
    return () => document.removeEventListener("keydown", handleKeyDown);
  }, [onClose]);
  const handleSubmit = useCallback(
    (e) => {
      e.preventDefault();
      if (!url.trim()) {
        setError("Please enter a URL");
        return;
      }
      if (!isValidUrl(url)) {
        setError("Please enter a valid URL");
        return;
      }
      const normalizedUrl = normalizeUrl(url);
      editor.setLink(normalizedUrl);
      editor.pm.view.focus();
      onClose();
    },
    [editor, url, onClose]
  );
  const handleRemove = useCallback(() => {
    editor.removeLink();
    editor.pm.view.focus();
    onClose();
  }, [editor, onClose]);
  const handleOpenLink = useCallback(() => {
    if (currentUrl) {
      window.open(currentUrl, "_blank", "noopener,noreferrer");
    }
  }, [currentUrl]);
  const handleInputChange = useCallback((e) => {
    setUrl(e.target.value);
    setError(null);
  }, []);
  const isEditing = currentUrl !== null;
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: popoverRef,
      className: `ob-link-popover ${className || ""}`,
      style: {
        position: "fixed",
        left: position.left,
        top: position.top,
        zIndex: 1001
      },
      role: "dialog",
      "aria-label": isEditing ? "Edit link" : "Add link",
      children: /* @__PURE__ */ jsxs("form", { onSubmit: handleSubmit, className: "ob-link-popover-form", children: [
        /* @__PURE__ */ jsx("div", { className: "ob-link-popover-input-row", children: /* @__PURE__ */ jsxs("div", { className: `ob-link-popover-input-wrapper ${error ? "ob-link-popover-input-wrapper--error" : ""}`, children: [
          /* @__PURE__ */ jsxs(
            "svg",
            {
              className: "ob-link-popover-input-icon",
              viewBox: "0 0 24 24",
              fill: "none",
              stroke: "currentColor",
              children: [
                /* @__PURE__ */ jsx("path", { d: "M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71" }),
                /* @__PURE__ */ jsx("path", { d: "M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71" })
              ]
            }
          ),
          /* @__PURE__ */ jsx(
            "input",
            {
              ref: inputRef,
              type: "text",
              className: "ob-link-popover-input",
              value: url,
              onChange: handleInputChange,
              placeholder: "Enter URL...",
              "aria-label": "Link URL",
              "aria-invalid": !!error
            }
          ),
          isEditing && /* @__PURE__ */ jsxs(Fragment, { children: [
            /* @__PURE__ */ jsx(
              "button",
              {
                type: "button",
                className: "ob-link-popover-inline-btn",
                onClick: handleOpenLink,
                title: "Open link",
                children: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
                  /* @__PURE__ */ jsx("path", { d: "M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6" }),
                  /* @__PURE__ */ jsx("polyline", { points: "15 3 21 3 21 9" }),
                  /* @__PURE__ */ jsx("line", { x1: "10", y1: "14", x2: "21", y2: "3" })
                ] })
              }
            ),
            /* @__PURE__ */ jsx(
              "button",
              {
                type: "button",
                className: "ob-link-popover-inline-btn ob-link-popover-inline-btn--danger",
                onClick: handleRemove,
                title: "Remove link",
                children: /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: /* @__PURE__ */ jsx("path", { d: "M18 6 6 18M6 6l12 12" }) })
              }
            )
          ] }),
          /* @__PURE__ */ jsx(
            "button",
            {
              type: "submit",
              className: "ob-link-popover-inline-btn ob-link-popover-inline-btn--primary",
              title: isEditing ? "Update link" : "Add link",
              children: /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: /* @__PURE__ */ jsx("polyline", { points: "20 6 9 17 4 12" }) })
            }
          )
        ] }) }),
        error && /* @__PURE__ */ jsx("p", { className: "ob-link-popover-error", children: error })
      ] })
    }
  );
}
const DEFAULT_TEXT_COLORS = [
  { value: "", label: "Default" },
  { value: "#374151", label: "Gray" },
  { value: "#dc2626", label: "Red" },
  { value: "#ea580c", label: "Orange" },
  { value: "#ca8a04", label: "Yellow" },
  { value: "#16a34a", label: "Green" },
  { value: "#2563eb", label: "Blue" },
  { value: "#7c3aed", label: "Purple" },
  { value: "#db2777", label: "Pink" }
];
const DEFAULT_BACKGROUND_COLORS = [
  { value: "", label: "Default" },
  { value: "#f3f4f6", label: "Gray" },
  { value: "#fee2e2", label: "Red" },
  { value: "#ffedd5", label: "Orange" },
  { value: "#fef9c3", label: "Yellow" },
  { value: "#dcfce7", label: "Green" },
  { value: "#dbeafe", label: "Blue" },
  { value: "#ede9fe", label: "Purple" },
  { value: "#fce7f3", label: "Pink" }
];
function ColorPicker({
  editor,
  currentTextColor,
  currentBackgroundColor,
  textColors = DEFAULT_TEXT_COLORS,
  backgroundColors = DEFAULT_BACKGROUND_COLORS,
  textColorLabel = "Color",
  backgroundColorLabel = "Background",
  onClose
}) {
  const [isOpen, setIsOpen] = useState(false);
  const [dropdownPosition, setDropdownPosition] = useState({ left: 0 });
  const containerRef = useRef(null);
  const buttonRef = useRef(null);
  const dropdownRef = useRef(null);
  const hasAnyColor = currentTextColor || currentBackgroundColor;
  useEffect(() => {
    const handleClickOutside = (event) => {
      if (containerRef.current && !containerRef.current.contains(event.target)) {
        setIsOpen(false);
        onClose == null ? void 0 : onClose();
      }
    };
    if (isOpen) {
      document.addEventListener("mousedown", handleClickOutside);
    }
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [isOpen, onClose]);
  useLayoutEffect(() => {
    if (!isOpen || !buttonRef.current || !dropdownRef.current) return;
    const buttonRect = buttonRef.current.getBoundingClientRect();
    const dropdownHeight = dropdownRef.current.offsetHeight || 250;
    const spaceBelow = window.innerHeight - buttonRect.bottom - 8;
    const spaceAbove = buttonRect.top - 8;
    const shouldOpenUpward = spaceBelow < dropdownHeight && spaceAbove > spaceBelow;
    setDropdownPosition({
      left: buttonRect.left + buttonRect.width / 2,
      ...shouldOpenUpward ? { bottom: window.innerHeight - buttonRect.top + 8 } : { top: buttonRect.bottom + 8 }
    });
  }, [isOpen]);
  const handleToggle = useCallback(() => {
    setIsOpen(!isOpen);
  }, [isOpen]);
  const handleSelectTextColor = useCallback(
    (color) => {
      if (color === "") {
        editor.removeTextColor();
      } else {
        editor.setTextColor(color);
      }
      editor.pm.view.focus();
    },
    [editor]
  );
  const handleSelectBackgroundColor = useCallback(
    (color) => {
      if (color === "") {
        editor.removeBackgroundColor();
      } else {
        editor.setBackgroundColor(color);
      }
      editor.pm.view.focus();
    },
    [editor]
  );
  return /* @__PURE__ */ jsxs("div", { className: "ob-color-picker", ref: containerRef, children: [
    /* @__PURE__ */ jsx(
      "button",
      {
        ref: buttonRef,
        type: "button",
        className: `ob-bubble-menu-btn ${hasAnyColor ? "ob-bubble-menu-btn--active" : ""}`,
        onClick: handleToggle,
        onMouseDown: (e) => e.preventDefault(),
        title: "Colors",
        "aria-expanded": isOpen,
        "aria-haspopup": "listbox",
        children: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
          /* @__PURE__ */ jsx(
            "path",
            {
              d: "M5 18h14",
              strokeWidth: "3",
              stroke: currentTextColor || currentBackgroundColor || "currentColor"
            }
          ),
          /* @__PURE__ */ jsx("path", { d: "M6 15l6-12 6 12" }),
          /* @__PURE__ */ jsx("path", { d: "M8.5 11h7" })
        ] })
      }
    ),
    isOpen && /* @__PURE__ */ jsxs(
      "div",
      {
        ref: dropdownRef,
        className: "ob-color-picker-dropdown",
        role: "listbox",
        style: {
          position: "fixed",
          left: dropdownPosition.left,
          top: dropdownPosition.top,
          bottom: dropdownPosition.bottom,
          transform: "translateX(-50%)"
        },
        children: [
          /* @__PURE__ */ jsxs("div", { className: "ob-color-picker-section", children: [
            /* @__PURE__ */ jsx("div", { className: "ob-color-picker-label", children: textColorLabel }),
            /* @__PURE__ */ jsx("div", { className: "ob-color-picker-grid", children: textColors.map((color) => {
              const isActive = color.value === "" ? !currentTextColor : currentTextColor === color.value;
              return /* @__PURE__ */ jsx(
                "button",
                {
                  type: "button",
                  className: `ob-color-picker-option ${isActive ? "ob-color-picker-option--active" : ""}`,
                  onClick: () => handleSelectTextColor(color.value),
                  onMouseDown: (e) => e.preventDefault(),
                  role: "option",
                  "aria-selected": isActive,
                  title: color.label,
                  children: color.value === "" ? /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
                    /* @__PURE__ */ jsx("circle", { cx: "12", cy: "12", r: "10" }),
                    /* @__PURE__ */ jsx("path", { d: "M4 4l16 16" })
                  ] }) : /* @__PURE__ */ jsx("span", { className: "ob-color-picker-swatch ob-color-picker-swatch--text", children: /* @__PURE__ */ jsx("span", { style: { color: color.value }, children: "A" }) })
                },
                `text-${color.value || "default"}`
              );
            }) })
          ] }),
          /* @__PURE__ */ jsx("div", { className: "ob-color-picker-divider" }),
          /* @__PURE__ */ jsxs("div", { className: "ob-color-picker-section", children: [
            /* @__PURE__ */ jsx("div", { className: "ob-color-picker-label", children: backgroundColorLabel }),
            /* @__PURE__ */ jsx("div", { className: "ob-color-picker-grid", children: backgroundColors.map((color) => {
              const isActive = color.value === "" ? !currentBackgroundColor : currentBackgroundColor === color.value;
              return /* @__PURE__ */ jsx(
                "button",
                {
                  type: "button",
                  className: `ob-color-picker-option ${isActive ? "ob-color-picker-option--active" : ""}`,
                  onClick: () => handleSelectBackgroundColor(color.value),
                  onMouseDown: (e) => e.preventDefault(),
                  role: "option",
                  "aria-selected": isActive,
                  title: color.label,
                  children: color.value === "" ? /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
                    /* @__PURE__ */ jsx("circle", { cx: "12", cy: "12", r: "10" }),
                    /* @__PURE__ */ jsx("path", { d: "M4 4l16 16" })
                  ] }) : /* @__PURE__ */ jsx(
                    "span",
                    {
                      className: "ob-color-picker-swatch",
                      style: { backgroundColor: color.value }
                    }
                  )
                },
                `bg-${color.value || "default"}`
              );
            }) })
          ] })
        ]
      }
    )
  ] });
}
const Icons$1 = {
  bold: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
    /* @__PURE__ */ jsx("path", { d: "M6 4h8a4 4 0 0 1 4 4 4 4 0 0 1-4 4H6z" }),
    /* @__PURE__ */ jsx("path", { d: "M6 12h9a4 4 0 0 1 4 4 4 4 0 0 1-4 4H6z" })
  ] }),
  italic: /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: /* @__PURE__ */ jsx("path", { d: "M19 4h-9M14 20H5M15 4L9 20" }) }),
  underline: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
    /* @__PURE__ */ jsx("path", { d: "M6 4v6a6 6 0 0 0 12 0V4" }),
    /* @__PURE__ */ jsx("path", { d: "M4 20h16" })
  ] }),
  strikethrough: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
    /* @__PURE__ */ jsx("path", { d: "M16 4c-1.5 0-3-.5-4.5-.5S8 4 6.5 5.5 5 9 6.5 10.5" }),
    /* @__PURE__ */ jsx("path", { d: "M8 20c1.5 0 3 .5 4.5.5s3.5-.5 5-2 1.5-4 0-5.5" }),
    /* @__PURE__ */ jsx("path", { d: "M4 12h16" })
  ] }),
  code: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
    /* @__PURE__ */ jsx("path", { d: "m16 18 6-6-6-6" }),
    /* @__PURE__ */ jsx("path", { d: "m8 6-6 6 6 6" })
  ] }),
  link: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
    /* @__PURE__ */ jsx("path", { d: "M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71" }),
    /* @__PURE__ */ jsx("path", { d: "M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71" })
  ] }),
  alignLeft: /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: /* @__PURE__ */ jsx("path", { d: "M4 6h16M4 12h10M4 18h14" }) }),
  alignCenter: /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: /* @__PURE__ */ jsx("path", { d: "M4 6h16M7 12h10M5 18h14" }) }),
  alignRight: /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: /* @__PURE__ */ jsx("path", { d: "M4 6h16M10 12h10M6 18h14" }) })
};
const BUBBLE_MENU_ITEMS = {
  bold: {
    id: "bold",
    label: "Bold (Cmd+B)",
    icon: Icons$1.bold,
    isActive: (state) => state.activeMarks.bold,
    action: (editor) => {
      editor.toggleBold();
      editor.pm.view.focus();
    }
  },
  italic: {
    id: "italic",
    label: "Italic (Cmd+I)",
    icon: Icons$1.italic,
    isActive: (state) => state.activeMarks.italic,
    action: (editor) => {
      editor.toggleItalic();
      editor.pm.view.focus();
    }
  },
  underline: {
    id: "underline",
    label: "Underline (Cmd+U)",
    icon: Icons$1.underline,
    isActive: (state) => state.activeMarks.underline,
    action: (editor) => {
      editor.toggleUnderline();
      editor.pm.view.focus();
    }
  },
  strikethrough: {
    id: "strikethrough",
    label: "Strikethrough",
    icon: Icons$1.strikethrough,
    isActive: (state) => state.activeMarks.strikethrough,
    action: (editor) => {
      editor.toggleStrikethrough();
      editor.pm.view.focus();
    }
  },
  code: {
    id: "code",
    label: "Inline code",
    icon: Icons$1.code,
    isActive: (state) => state.activeMarks.code,
    action: (editor) => {
      editor.toggleCode();
      editor.pm.view.focus();
    }
  },
  alignLeft: {
    id: "alignLeft",
    label: "Align left",
    icon: Icons$1.alignLeft,
    isActive: (state) => state.textAlign === "left",
    action: (editor) => {
      editor.setTextAlign("left");
      editor.pm.view.focus();
    }
  },
  alignCenter: {
    id: "alignCenter",
    label: "Align center",
    icon: Icons$1.alignCenter,
    isActive: (state) => state.textAlign === "center",
    action: (editor) => {
      editor.setTextAlign("center");
      editor.pm.view.focus();
    }
  },
  alignRight: {
    id: "alignRight",
    label: "Align right",
    icon: Icons$1.alignRight,
    isActive: (state) => state.textAlign === "right",
    action: (editor) => {
      editor.setTextAlign("right");
      editor.pm.view.focus();
    }
  }
};
const DEFAULT_BUBBLE_MENU_ORDER = [
  "blockType",
  "---",
  "alignLeft",
  "alignCenter",
  "alignRight",
  "---",
  "bold",
  "italic",
  "underline",
  "strikethrough",
  "---",
  "code",
  "link",
  "---",
  "color"
];
function FormatButton({ active, onClick, title, children }) {
  return /* @__PURE__ */ jsx(
    "button",
    {
      type: "button",
      className: `ob-bubble-menu-btn ${active ? "ob-bubble-menu-btn--active" : ""}`,
      onClick,
      title,
      onMouseDown: (e) => e.preventDefault(),
      children
    }
  );
}
const BLOCK_TYPE_OPTIONS = [
  {
    type: "paragraph",
    label: "Paragraph",
    icon: /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: /* @__PURE__ */ jsx("path", { d: "M4 6h16M4 12h16M4 18h10" }) })
  },
  {
    type: "heading",
    label: "Heading 1",
    props: { level: 1 },
    icon: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
      /* @__PURE__ */ jsx("path", { d: "M4 12h8M4 6v12M12 6v12" }),
      /* @__PURE__ */ jsx("path", { d: "M20 8v8M17 8h6", strokeWidth: "1.5" })
    ] })
  },
  {
    type: "heading",
    label: "Heading 2",
    props: { level: 2 },
    icon: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
      /* @__PURE__ */ jsx("path", { d: "M4 12h8M4 6v12M12 6v12" }),
      /* @__PURE__ */ jsx("path", { d: "M16 12a3 3 0 1 1 6 0c0 1.5-3 3-3 3h3M16 18h6", strokeWidth: "1.5" })
    ] })
  },
  {
    type: "heading",
    label: "Heading 3",
    props: { level: 3 },
    icon: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
      /* @__PURE__ */ jsx("path", { d: "M4 12h8M4 6v12M12 6v12" }),
      /* @__PURE__ */ jsx("path", { d: "M16 9a2 2 0 1 1 4 1.5c-.5.5-2 1-2 1s1.5.5 2 1a2 2 0 1 1-4 1.5", strokeWidth: "1.5" })
    ] })
  },
  {
    type: "blockquote",
    label: "Quote",
    icon: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
      /* @__PURE__ */ jsx("path", { d: "M3 21c3 0 7-1 7-8V5c0-1.25-.756-2.017-2-2H4c-1.25 0-2 .75-2 1.972V11c0 1.25.75 2 2 2 1 0 1 0 1 1v1c0 1-1 2-2 2s-1 .008-1 1.031V21z" }),
      /* @__PURE__ */ jsx("path", { d: "M15 21c3 0 7-1 7-8V5c0-1.25-.757-2.017-2-2h-4c-1.25 0-2 .75-2 1.972V11c0 1.25.75 2 2 2h.75c0 2.25.25 4-2.75 4v4z" })
    ] })
  },
  {
    type: "bulletList",
    label: "Bullet List",
    icon: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
      /* @__PURE__ */ jsx("circle", { cx: "4", cy: "7", r: "1.5", fill: "currentColor", stroke: "none" }),
      /* @__PURE__ */ jsx("circle", { cx: "4", cy: "12", r: "1.5", fill: "currentColor", stroke: "none" }),
      /* @__PURE__ */ jsx("circle", { cx: "4", cy: "17", r: "1.5", fill: "currentColor", stroke: "none" }),
      /* @__PURE__ */ jsx("path", { d: "M9 7h11M9 12h11M9 17h11" })
    ] })
  },
  {
    type: "orderedList",
    label: "Numbered List",
    icon: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
      /* @__PURE__ */ jsx("path", { d: "M10 7h10M10 12h10M10 17h10" }),
      /* @__PURE__ */ jsx("path", { d: "M4 7h2M4 17h2M5 11v3h2", strokeWidth: "1.5" })
    ] })
  },
  {
    type: "codeBlock",
    label: "Code Block",
    icon: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
      /* @__PURE__ */ jsx("rect", { x: "3", y: "3", width: "18", height: "18", rx: "2" }),
      /* @__PURE__ */ jsx("path", { d: "m9 9-3 3 3 3M15 9l3 3-3 3" })
    ] })
  }
];
function getBlockTypeLabel(blockType) {
  if (blockType.type === "heading") {
    const level = blockType.props.level;
    return `Heading ${level}`;
  }
  const option = BLOCK_TYPE_OPTIONS.find((opt) => opt.type === blockType.type);
  return (option == null ? void 0 : option.label) || "Paragraph";
}
function getBlockTypeIcon(blockType) {
  if (blockType.type === "heading") {
    const level = blockType.props.level;
    const option2 = BLOCK_TYPE_OPTIONS.find(
      (opt) => {
        var _a;
        return opt.type === "heading" && ((_a = opt.props) == null ? void 0 : _a.level) === level;
      }
    );
    return (option2 == null ? void 0 : option2.icon) || BLOCK_TYPE_OPTIONS[0].icon;
  }
  const option = BLOCK_TYPE_OPTIONS.find((opt) => opt.type === blockType.type);
  return (option == null ? void 0 : option.icon) || BLOCK_TYPE_OPTIONS[0].icon;
}
function blockTypeMatches(blockType, option) {
  var _a;
  if (blockType.type !== option.type) return false;
  if (((_a = option.props) == null ? void 0 : _a.level) && blockType.props.level !== option.props.level) return false;
  return true;
}
function BlockTypeSelector({ editor, blockType }) {
  const [isOpen, setIsOpen] = useState(false);
  const [openUpward, setOpenUpward] = useState(false);
  const containerRef = useRef(null);
  const buttonRef = useRef(null);
  const dropdownRef = useRef(null);
  useEffect(() => {
    const handleClickOutside = (event) => {
      if (containerRef.current && !containerRef.current.contains(event.target)) {
        setIsOpen(false);
      }
    };
    if (isOpen) {
      document.addEventListener("mousedown", handleClickOutside);
    }
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [isOpen]);
  useLayoutEffect(() => {
    if (!isOpen || !buttonRef.current || !dropdownRef.current) return;
    const buttonRect = buttonRef.current.getBoundingClientRect();
    const dropdownHeight = dropdownRef.current.offsetHeight || 300;
    const spaceBelow = window.innerHeight - buttonRect.bottom - 8;
    const spaceAbove = buttonRect.top - 8;
    setOpenUpward(spaceBelow < dropdownHeight && spaceAbove > spaceBelow);
  }, [isOpen]);
  const handleSelect = useCallback(
    (option) => {
      editor.setBlockType(option.type, option.props || {});
      editor.pm.view.focus();
      setIsOpen(false);
    },
    [editor]
  );
  return /* @__PURE__ */ jsxs("div", { className: "ob-block-type-selector", ref: containerRef, children: [
    /* @__PURE__ */ jsxs(
      "button",
      {
        ref: buttonRef,
        type: "button",
        className: "ob-block-type-selector-btn",
        onClick: () => setIsOpen(!isOpen),
        onMouseDown: (e) => e.preventDefault(),
        "aria-expanded": isOpen,
        "aria-haspopup": "listbox",
        children: [
          /* @__PURE__ */ jsx("span", { className: "ob-block-type-selector-icon", children: getBlockTypeIcon(blockType) }),
          /* @__PURE__ */ jsx("span", { className: "ob-block-type-selector-label", children: getBlockTypeLabel(blockType) }),
          /* @__PURE__ */ jsx(
            "svg",
            {
              className: `ob-block-type-selector-chevron ${isOpen ? "ob-block-type-selector-chevron--open" : ""}`,
              viewBox: "0 0 24 24",
              fill: "none",
              stroke: "currentColor",
              children: /* @__PURE__ */ jsx("path", { d: "m6 9 6 6 6-6" })
            }
          )
        ]
      }
    ),
    isOpen && /* @__PURE__ */ jsx(
      "div",
      {
        ref: dropdownRef,
        className: `ob-block-type-dropdown ${openUpward ? "ob-block-type-dropdown--upward" : ""}`,
        role: "listbox",
        children: BLOCK_TYPE_OPTIONS.map((option, index) => {
          var _a;
          const isActive = blockTypeMatches(blockType, option);
          return /* @__PURE__ */ jsxs(
            "button",
            {
              type: "button",
              className: `ob-block-type-option ${isActive ? "ob-block-type-option--active" : ""}`,
              onClick: () => handleSelect(option),
              onMouseDown: (e) => e.preventDefault(),
              role: "option",
              "aria-selected": isActive,
              children: [
                /* @__PURE__ */ jsx("span", { className: "ob-block-type-option-icon", children: option.icon }),
                /* @__PURE__ */ jsx("span", { className: "ob-block-type-option-label", children: option.label }),
                isActive && /* @__PURE__ */ jsx(
                  "svg",
                  {
                    className: "ob-block-type-option-check",
                    viewBox: "0 0 24 24",
                    fill: "none",
                    stroke: "currentColor",
                    children: /* @__PURE__ */ jsx("path", { d: "M20 6 9 17l-5-5" })
                  }
                )
              ]
            },
            `${option.type}-${((_a = option.props) == null ? void 0 : _a.level) || index}`
          );
        })
      }
    )
  ] });
}
function BubbleMenu({
  editor,
  customItems = [],
  itemOrder,
  hideItems = [],
  children,
  className
}) {
  const [menuState, setMenuState] = useState(null);
  const [showLinkPopover, setShowLinkPopover] = useState(false);
  const linkButtonRef = useRef(null);
  useEffect(() => {
    if (!editor || editor.isDestroyed) return;
    const updateState = () => {
      const state = BUBBLE_MENU_PLUGIN_KEY.getState(editor.pm.state);
      setMenuState(state ?? null);
    };
    updateState();
    const unsubscribe = editor.on("transaction", updateState);
    return unsubscribe;
  }, [editor]);
  useEffect(() => {
    if (!(menuState == null ? void 0 : menuState.visible)) {
      setShowLinkPopover(false);
    }
  }, [menuState == null ? void 0 : menuState.visible]);
  const handleLinkClick = useCallback(() => {
    setShowLinkPopover(true);
  }, []);
  const closeLinkPopover = useCallback(() => {
    setShowLinkPopover(false);
  }, []);
  if (!editor || editor.isDestroyed || !(menuState == null ? void 0 : menuState.visible) || !menuState.coords) {
    return null;
  }
  const menuHeight = 36;
  const style = {
    position: "fixed",
    left: menuState.coords.left,
    top: menuState.coords.top - menuHeight - 8,
    zIndex: 1e3
  };
  if (children) {
    return /* @__PURE__ */ jsx("div", { className: `ob-bubble-menu ${className || ""}`, style, children: children({ editor, state: menuState }) });
  }
  const allItems = { ...BUBBLE_MENU_ITEMS };
  for (const item of customItems) {
    allItems[item.id] = item;
  }
  const order = itemOrder || [
    ...DEFAULT_BUBBLE_MENU_ORDER,
    // Add custom items at the end if not in custom order
    ...customItems.length > 0 ? ["---", ...customItems.map((i) => i.id)] : []
  ];
  const hiddenSet = new Set(hideItems);
  const { activeMarks, blockType } = menuState;
  const renderItem = (itemId, index) => {
    if (itemId === "---") {
      return /* @__PURE__ */ jsx("span", { className: "ob-bubble-menu-divider" }, `divider-${index}`);
    }
    if (hiddenSet.has(itemId)) {
      return null;
    }
    if (itemId === "blockType") {
      return /* @__PURE__ */ jsx(BlockTypeSelector, { editor, blockType }, "blockType");
    }
    if (itemId === "link") {
      return /* @__PURE__ */ jsx(
        "button",
        {
          ref: linkButtonRef,
          type: "button",
          className: `ob-bubble-menu-btn ${activeMarks.link ? "ob-bubble-menu-btn--active" : ""}`,
          onClick: handleLinkClick,
          onMouseDown: (e) => e.preventDefault(),
          title: activeMarks.link ? "Edit link" : "Add link",
          children: Icons$1.link
        },
        "link"
      );
    }
    if (itemId === "color") {
      return /* @__PURE__ */ jsx(
        ColorPicker,
        {
          editor,
          currentTextColor: activeMarks.textColor,
          currentBackgroundColor: activeMarks.backgroundColor
        },
        "color"
      );
    }
    const item = allItems[itemId];
    if (!item) {
      return null;
    }
    const isActive = item.isActive ? item.isActive(menuState, editor) : false;
    return /* @__PURE__ */ jsx(
      FormatButton,
      {
        active: isActive,
        onClick: () => item.action(editor, menuState),
        title: item.label,
        children: item.icon
      },
      item.id
    );
  };
  return /* @__PURE__ */ jsxs(
    "div",
    {
      className: `ob-bubble-menu ${className || ""}`,
      style,
      role: "toolbar",
      "aria-label": "Text formatting",
      children: [
        order.map((itemId, index) => renderItem(itemId, index)),
        showLinkPopover && linkButtonRef.current && /* @__PURE__ */ jsx(
          LinkPopover,
          {
            editor,
            currentUrl: activeMarks.link,
            onClose: closeLinkPopover,
            triggerRef: linkButtonRef,
            position: {
              left: linkButtonRef.current.getBoundingClientRect().left,
              top: linkButtonRef.current.getBoundingClientRect().bottom + 8
            }
          }
        )
      ]
    }
  );
}
function TableMenuButton({ onClick, title, children, danger }) {
  return /* @__PURE__ */ jsx(
    "button",
    {
      type: "button",
      className: `ob-table-menu-btn ${danger ? "ob-table-menu-btn--danger" : ""}`,
      onClick,
      title,
      onMouseDown: (e) => e.preventDefault(),
      children
    }
  );
}
function TableMenuDropdown({ label, icon, children }) {
  const [isOpen, setIsOpen] = useState(false);
  const containerRef = useRef(null);
  useEffect(() => {
    const handleClickOutside = (event) => {
      if (containerRef.current && !containerRef.current.contains(event.target)) {
        setIsOpen(false);
      }
    };
    if (isOpen) {
      document.addEventListener("mousedown", handleClickOutside);
    }
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [isOpen]);
  return /* @__PURE__ */ jsxs("div", { className: "ob-table-menu-dropdown", ref: containerRef, children: [
    /* @__PURE__ */ jsxs(
      "button",
      {
        type: "button",
        className: "ob-table-menu-dropdown-btn",
        onClick: () => setIsOpen(!isOpen),
        onMouseDown: (e) => e.preventDefault(),
        "aria-expanded": isOpen,
        children: [
          /* @__PURE__ */ jsx("span", { className: "ob-table-menu-dropdown-icon", children: icon }),
          /* @__PURE__ */ jsx("span", { className: "ob-table-menu-dropdown-label", children: label }),
          /* @__PURE__ */ jsx(
            "svg",
            {
              className: `ob-table-menu-dropdown-chevron ${isOpen ? "ob-table-menu-dropdown-chevron--open" : ""}`,
              viewBox: "0 0 24 24",
              fill: "none",
              stroke: "currentColor",
              children: /* @__PURE__ */ jsx("path", { d: "m6 9 6 6 6-6" })
            }
          )
        ]
      }
    ),
    isOpen && /* @__PURE__ */ jsx("div", { className: "ob-table-menu-dropdown-content", onClick: () => setIsOpen(false), children })
  ] });
}
const Icons = {
  addRowAbove: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("path", { d: "M3 14h18v7H3zM12 3v6M9 6h6" }),
    /* @__PURE__ */ jsx("rect", { x: "3", y: "14", width: "18", height: "7", rx: "1" })
  ] }),
  addRowBelow: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("rect", { x: "3", y: "3", width: "18", height: "7", rx: "1" }),
    /* @__PURE__ */ jsx("path", { d: "M12 14v6M9 17h6" })
  ] }),
  deleteRow: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("rect", { x: "3", y: "8", width: "18", height: "8", rx: "1" }),
    /* @__PURE__ */ jsx("path", { d: "M8 12h8" })
  ] }),
  addColumnLeft: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("rect", { x: "14", y: "3", width: "7", height: "18", rx: "1" }),
    /* @__PURE__ */ jsx("path", { d: "M3 12h6M6 9v6" })
  ] }),
  addColumnRight: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("rect", { x: "3", y: "3", width: "7", height: "18", rx: "1" }),
    /* @__PURE__ */ jsx("path", { d: "M15 12h6M18 9v6" })
  ] }),
  deleteColumn: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("rect", { x: "8", y: "3", width: "8", height: "18", rx: "1" }),
    /* @__PURE__ */ jsx("path", { d: "M10 12h4" })
  ] }),
  deleteTable: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("rect", { x: "3", y: "3", width: "18", height: "18", rx: "2" }),
    /* @__PURE__ */ jsx("path", { d: "M3 9h18M3 15h18M9 3v18M15 3v18" }),
    /* @__PURE__ */ jsx("path", { d: "M6 6l12 12M18 6L6 18", strokeWidth: "2.5" })
  ] }),
  rows: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("rect", { x: "3", y: "3", width: "18", height: "18", rx: "2" }),
    /* @__PURE__ */ jsx("path", { d: "M3 9h18M3 15h18" })
  ] }),
  columns: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: [
    /* @__PURE__ */ jsx("rect", { x: "3", y: "3", width: "18", height: "18", rx: "2" }),
    /* @__PURE__ */ jsx("path", { d: "M9 3v18M15 3v18" })
  ] })
};
function TableMenu({
  editor,
  className
}) {
  const [inTable, setInTable] = useState(false);
  const [tableInfo, setTableInfo] = useState(null);
  const [coords, setCoords] = useState(null);
  useEffect(() => {
    if (!editor || editor.isDestroyed) return;
    const updateState = () => {
      const state = editor.pm.state;
      const isInsideTable = isInTable(state);
      setInTable(isInsideTable);
      if (isInsideTable) {
        const info = getTableInfo(state);
        setTableInfo(info);
        const { $from } = state.selection;
        for (let depth = $from.depth; depth > 0; depth--) {
          const node = $from.node(depth);
          if (node.type.name === "table") {
            const pos = $from.before(depth);
            const domNode = editor.pm.view.nodeDOM(pos);
            if (domNode) {
              const rect = domNode.getBoundingClientRect();
              setCoords({
                left: rect.left,
                top: rect.top - 44
                // Menu height + gap
              });
            }
            break;
          }
        }
      } else {
        setTableInfo(null);
        setCoords(null);
      }
    };
    updateState();
    const unsubscribe = editor.on("transaction", updateState);
    return unsubscribe;
  }, [editor]);
  const handleAddRowBefore = useCallback(() => {
    if (!editor || editor.isDestroyed) return;
    addRowBefore(editor.pm.state, editor.pm.view.dispatch);
    editor.pm.view.focus();
  }, [editor]);
  const handleAddRowAfter = useCallback(() => {
    if (!editor || editor.isDestroyed) return;
    addRowAfter(editor.pm.state, editor.pm.view.dispatch);
    editor.pm.view.focus();
  }, [editor]);
  const handleDeleteRow = useCallback(() => {
    if (!editor || editor.isDestroyed) return;
    deleteRow(editor.pm.state, editor.pm.view.dispatch);
    editor.pm.view.focus();
  }, [editor]);
  const handleAddColumnBefore = useCallback(() => {
    if (!editor || editor.isDestroyed) return;
    addColumnBefore(editor.pm.state, editor.pm.view.dispatch);
    editor.pm.view.focus();
  }, [editor]);
  const handleAddColumnAfter = useCallback(() => {
    if (!editor || editor.isDestroyed) return;
    addColumnAfter(editor.pm.state, editor.pm.view.dispatch);
    editor.pm.view.focus();
  }, [editor]);
  const handleDeleteColumn = useCallback(() => {
    if (!editor || editor.isDestroyed) return;
    deleteColumn(editor.pm.state, editor.pm.view.dispatch);
    editor.pm.view.focus();
  }, [editor]);
  const handleDeleteTable = useCallback(() => {
    if (!editor || editor.isDestroyed) return;
    deleteTable(editor.pm.state, editor.pm.view.dispatch);
    editor.pm.view.focus();
  }, [editor]);
  if (!editor || editor.isDestroyed || !inTable || !coords) {
    return null;
  }
  const style = {
    position: "fixed",
    left: coords.left,
    top: coords.top,
    zIndex: 1e3
  };
  return /* @__PURE__ */ jsxs(
    "div",
    {
      className: `ob-table-menu ${className || ""}`,
      style,
      role: "toolbar",
      "aria-label": "Table editing",
      children: [
        /* @__PURE__ */ jsxs(TableMenuDropdown, { label: "Row", icon: Icons.rows, children: [
          /* @__PURE__ */ jsxs(
            "button",
            {
              type: "button",
              className: "ob-table-menu-dropdown-item",
              onClick: handleAddRowBefore,
              onMouseDown: (e) => e.preventDefault(),
              children: [
                Icons.addRowAbove,
                /* @__PURE__ */ jsx("span", { children: "Insert row above" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            "button",
            {
              type: "button",
              className: "ob-table-menu-dropdown-item",
              onClick: handleAddRowAfter,
              onMouseDown: (e) => e.preventDefault(),
              children: [
                Icons.addRowBelow,
                /* @__PURE__ */ jsx("span", { children: "Insert row below" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            "button",
            {
              type: "button",
              className: "ob-table-menu-dropdown-item ob-table-menu-dropdown-item--danger",
              onClick: handleDeleteRow,
              onMouseDown: (e) => e.preventDefault(),
              disabled: (tableInfo == null ? void 0 : tableInfo.rowCount) === 1,
              children: [
                Icons.deleteRow,
                /* @__PURE__ */ jsx("span", { children: "Delete row" })
              ]
            }
          )
        ] }),
        /* @__PURE__ */ jsxs(TableMenuDropdown, { label: "Column", icon: Icons.columns, children: [
          /* @__PURE__ */ jsxs(
            "button",
            {
              type: "button",
              className: "ob-table-menu-dropdown-item",
              onClick: handleAddColumnBefore,
              onMouseDown: (e) => e.preventDefault(),
              children: [
                Icons.addColumnLeft,
                /* @__PURE__ */ jsx("span", { children: "Insert column left" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            "button",
            {
              type: "button",
              className: "ob-table-menu-dropdown-item",
              onClick: handleAddColumnAfter,
              onMouseDown: (e) => e.preventDefault(),
              children: [
                Icons.addColumnRight,
                /* @__PURE__ */ jsx("span", { children: "Insert column right" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            "button",
            {
              type: "button",
              className: "ob-table-menu-dropdown-item ob-table-menu-dropdown-item--danger",
              onClick: handleDeleteColumn,
              onMouseDown: (e) => e.preventDefault(),
              disabled: (tableInfo == null ? void 0 : tableInfo.colCount) === 1,
              children: [
                Icons.deleteColumn,
                /* @__PURE__ */ jsx("span", { children: "Delete column" })
              ]
            }
          )
        ] }),
        /* @__PURE__ */ jsx("span", { className: "ob-table-menu-divider" }),
        /* @__PURE__ */ jsx(
          TableMenuButton,
          {
            onClick: handleDeleteTable,
            title: "Delete table",
            danger: true,
            children: Icons.deleteTable
          }
        ),
        tableInfo && /* @__PURE__ */ jsxs("span", { className: "ob-table-menu-info", children: [
          "Row ",
          tableInfo.rowIndex + 1,
          "/",
          tableInfo.rowCount,
          ", Col ",
          tableInfo.cellIndex + 1,
          "/",
          tableInfo.colCount
        ] })
      ]
    }
  );
}
function getTableState(editor, tableElement) {
  const view = editor.pm.view;
  let tablePos = -1;
  view.state.doc.descendants((node, pos) => {
    if (node.type.name === "table" && tablePos === -1) {
      const domNode = view.nodeDOM(pos);
      if (domNode === tableElement || tableElement.contains(domNode)) {
        tablePos = pos;
        return false;
      }
    }
    return true;
  });
  if (tablePos === -1) return null;
  const table = view.state.doc.nodeAt(tablePos);
  if (!table || table.type.name !== "table") return null;
  const rowElements = tableElement.querySelectorAll(":scope > tr, :scope > tbody > tr, :scope > thead > tr");
  const rows = [];
  const tableRect = tableElement.getBoundingClientRect();
  rowElements.forEach((row) => {
    const rect = row.getBoundingClientRect();
    rows.push({
      element: row,
      top: rect.top - tableRect.top,
      height: rect.height
    });
  });
  const cols = [];
  if (rowElements.length > 0) {
    const firstRowCells = rowElements[0].querySelectorAll(":scope > td, :scope > th");
    firstRowCells.forEach((cell) => {
      const rect = cell.getBoundingClientRect();
      cols.push({
        element: cell,
        left: rect.left - tableRect.left,
        width: rect.width
      });
    });
  }
  return {
    tablePos,
    tableElement,
    rowCount: table.childCount,
    colCount: cols.length,
    rows,
    cols
  };
}
function TableHandles({
  editor,
  className
}) {
  const [tableState, setTableState] = useState(null);
  const [hoverState, setHoverState] = useState({ type: null, index: -1 });
  const [showRowMenu, setShowRowMenu] = useState(null);
  const [showColMenu, setShowColMenu] = useState(null);
  const [menuPosition, setMenuPosition] = useState(null);
  const containerRef = useRef(null);
  useEffect(() => {
    if (!editor || editor.isDestroyed) return;
    let hideTimeout = null;
    const clearHideTimeout = () => {
      if (hideTimeout) {
        clearTimeout(hideTimeout);
        hideTimeout = null;
      }
    };
    const scheduleHide = () => {
      clearHideTimeout();
      hideTimeout = setTimeout(() => {
        setTableState(null);
        setHoverState({ type: null, index: -1 });
      }, 100);
    };
    const handleMouseMove = (e) => {
      var _a;
      const target = e.target;
      if ((_a = containerRef.current) == null ? void 0 : _a.contains(target)) {
        clearHideTimeout();
        return;
      }
      const tableElement = target.closest("table");
      if (!tableElement) {
        if (tableState) {
          const tableRect3 = tableState.tableElement.getBoundingClientRect();
          const isInExtendZone = e.clientX >= tableRect3.right && e.clientX <= tableRect3.right + 30 && e.clientY >= tableRect3.top && e.clientY <= tableRect3.bottom || e.clientY >= tableRect3.bottom && e.clientY <= tableRect3.bottom + 30 && e.clientX >= tableRect3.left && e.clientX <= tableRect3.right;
          if (isInExtendZone) {
            clearHideTimeout();
            return;
          }
        }
        scheduleHide();
        return;
      }
      clearHideTimeout();
      const state = getTableState(editor, tableElement);
      if (!state) {
        scheduleHide();
        return;
      }
      setTableState(state);
      const tableRect2 = tableElement.getBoundingClientRect();
      const relX = e.clientX - tableRect2.left;
      const relY = e.clientY - tableRect2.top;
      if (relX < 0 && relX > -40) {
        const rowIndex = state.rows.findIndex(
          (row) => relY >= row.top && relY <= row.top + row.height
        );
        if (rowIndex !== -1) {
          setHoverState({ type: "row", index: rowIndex });
          return;
        }
      }
      if (relY < 0 && relY > -40) {
        const colIndex = state.cols.findIndex(
          (col) => relX >= col.left && relX <= col.left + col.width
        );
        if (colIndex !== -1) {
          setHoverState({ type: "col", index: colIndex });
          return;
        }
      }
      const cell = target.closest("td, th");
      if (cell) {
        const cellRect = cell.getBoundingClientRect();
        const rowElement = cell.closest("tr");
        if (rowElement) {
          const rowIndex = Array.from(
            tableElement.querySelectorAll(":scope > tr, :scope > tbody > tr, :scope > thead > tr")
          ).indexOf(rowElement);
          const colIndex = Array.from(rowElement.children).indexOf(cell);
          if (e.clientX - cellRect.left < 20) {
            setHoverState({ type: "row", index: rowIndex });
          } else if (e.clientY - cellRect.top < 20) {
            setHoverState({ type: "col", index: colIndex });
          } else {
            setHoverState({ type: null, index: -1 });
          }
        }
      }
    };
    const handleMouseLeave = () => {
      scheduleHide();
    };
    document.addEventListener("mousemove", handleMouseMove);
    document.addEventListener("mouseleave", handleMouseLeave);
    return () => {
      clearHideTimeout();
      document.removeEventListener("mousemove", handleMouseMove);
      document.removeEventListener("mouseleave", handleMouseLeave);
    };
  }, [editor, tableState]);
  useEffect(() => {
    const handleClick = (e) => {
      if (containerRef.current && !containerRef.current.contains(e.target)) {
        setShowRowMenu(null);
        setShowColMenu(null);
        setMenuPosition(null);
      }
    };
    document.addEventListener("mousedown", handleClick);
    return () => document.removeEventListener("mousedown", handleClick);
  }, []);
  const handleAddRow = useCallback(
    (index) => {
      if (!editor || editor.isDestroyed || !tableState) return;
      addRowAtIndex(editor.pm.state, editor.pm.view.dispatch, tableState.tablePos, index);
      editor.pm.view.focus();
      setShowRowMenu(null);
    },
    [editor, tableState]
  );
  const handleDeleteRow = useCallback(
    (index) => {
      if (!editor || editor.isDestroyed || !tableState) return;
      deleteRowAtIndex(editor.pm.state, editor.pm.view.dispatch, tableState.tablePos, index);
      editor.pm.view.focus();
      setShowRowMenu(null);
    },
    [editor, tableState]
  );
  const handleAddCol = useCallback(
    (index) => {
      if (!editor || editor.isDestroyed || !tableState) return;
      addColumnAtIndex(editor.pm.state, editor.pm.view.dispatch, tableState.tablePos, index);
      editor.pm.view.focus();
      setShowColMenu(null);
    },
    [editor, tableState]
  );
  const handleDeleteCol = useCallback(
    (index) => {
      if (!editor || editor.isDestroyed || !tableState) return;
      deleteColumnAtIndex(editor.pm.state, editor.pm.view.dispatch, tableState.tablePos, index);
      editor.pm.view.focus();
      setShowColMenu(null);
    },
    [editor, tableState]
  );
  if (!editor || editor.isDestroyed || !tableState) return null;
  const tableRect = tableState.tableElement.getBoundingClientRect();
  return /* @__PURE__ */ jsxs("div", { ref: containerRef, className: `ob-table-handles ${className || ""}`, children: [
    tableState.rows.map((row, index) => /* @__PURE__ */ jsxs(
      "div",
      {
        className: `ob-table-handle ob-table-handle--row ${hoverState.type === "row" && hoverState.index === index ? "ob-table-handle--visible" : ""}`,
        style: {
          position: "fixed",
          left: tableRect.left - 28,
          top: tableRect.top + row.top,
          height: row.height
        },
        children: [
          /* @__PURE__ */ jsx(
            "button",
            {
              type: "button",
              className: "ob-table-handle-btn",
              onClick: (e) => {
                if (showRowMenu === index) {
                  setShowRowMenu(null);
                  setMenuPosition(null);
                } else {
                  const btnRect = e.currentTarget.getBoundingClientRect();
                  setShowRowMenu(index);
                  setShowColMenu(null);
                  setMenuPosition({ left: btnRect.right + 4, top: btnRect.top });
                }
              },
              onMouseDown: (e) => e.preventDefault(),
              title: "Row options",
              children: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "currentColor", children: [
                /* @__PURE__ */ jsx("circle", { cx: "12", cy: "6", r: "2" }),
                /* @__PURE__ */ jsx("circle", { cx: "12", cy: "12", r: "2" }),
                /* @__PURE__ */ jsx("circle", { cx: "12", cy: "18", r: "2" })
              ] })
            }
          ),
          showRowMenu === index && menuPosition && /* @__PURE__ */ jsxs(
            "div",
            {
              className: "ob-table-handle-menu",
              style: { left: menuPosition.left, top: menuPosition.top },
              children: [
                /* @__PURE__ */ jsxs("button", { onClick: () => handleAddRow(index), children: [
                  /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M12 5v14M5 12h14" }) }),
                  "Insert above"
                ] }),
                /* @__PURE__ */ jsxs("button", { onClick: () => handleAddRow(index + 1), children: [
                  /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M12 5v14M5 12h14" }) }),
                  "Insert below"
                ] }),
                tableState.rowCount > 1 && /* @__PURE__ */ jsxs("button", { className: "ob-table-handle-menu-danger", onClick: () => handleDeleteRow(index), children: [
                  /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M18 6L6 18M6 6l12 12" }) }),
                  "Delete row"
                ] })
              ]
            }
          )
        ]
      },
      `row-${index}`
    )),
    tableState.cols.map((col, index) => /* @__PURE__ */ jsxs(
      "div",
      {
        className: `ob-table-handle ob-table-handle--col ${hoverState.type === "col" && hoverState.index === index ? "ob-table-handle--visible" : ""}`,
        style: {
          position: "fixed",
          left: tableRect.left + col.left,
          top: tableRect.top - 28,
          width: col.width
        },
        children: [
          /* @__PURE__ */ jsx(
            "button",
            {
              type: "button",
              className: "ob-table-handle-btn",
              onClick: (e) => {
                if (showColMenu === index) {
                  setShowColMenu(null);
                  setMenuPosition(null);
                } else {
                  const btnRect = e.currentTarget.getBoundingClientRect();
                  setShowColMenu(index);
                  setShowRowMenu(null);
                  setMenuPosition({ left: btnRect.left + btnRect.width / 2 - 75, top: btnRect.bottom + 4 });
                }
              },
              onMouseDown: (e) => e.preventDefault(),
              title: "Column options",
              children: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "currentColor", children: [
                /* @__PURE__ */ jsx("circle", { cx: "6", cy: "12", r: "2" }),
                /* @__PURE__ */ jsx("circle", { cx: "12", cy: "12", r: "2" }),
                /* @__PURE__ */ jsx("circle", { cx: "18", cy: "12", r: "2" })
              ] })
            }
          ),
          showColMenu === index && menuPosition && /* @__PURE__ */ jsxs(
            "div",
            {
              className: "ob-table-handle-menu",
              style: { left: menuPosition.left, top: menuPosition.top },
              children: [
                /* @__PURE__ */ jsxs("button", { onClick: () => handleAddCol(index), children: [
                  /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M12 5v14M5 12h14" }) }),
                  "Insert left"
                ] }),
                /* @__PURE__ */ jsxs("button", { onClick: () => handleAddCol(index + 1), children: [
                  /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M12 5v14M5 12h14" }) }),
                  "Insert right"
                ] }),
                tableState.colCount > 1 && /* @__PURE__ */ jsxs("button", { className: "ob-table-handle-menu-danger", onClick: () => handleDeleteCol(index), children: [
                  /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M18 6L6 18M6 6l12 12" }) }),
                  "Delete column"
                ] })
              ]
            }
          )
        ]
      },
      `col-${index}`
    )),
    /* @__PURE__ */ jsx(
      "button",
      {
        type: "button",
        className: "ob-table-extend-btn ob-table-extend-btn--row ob-table-extend-btn--visible",
        style: {
          position: "fixed",
          left: tableRect.left,
          top: tableRect.bottom + 4,
          width: tableRect.width
        },
        onClick: () => handleAddRow(tableState.rowCount),
        onMouseDown: (e) => e.preventDefault(),
        title: "Add row",
        children: /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M12 5v14M5 12h14" }) })
      }
    ),
    /* @__PURE__ */ jsx(
      "button",
      {
        type: "button",
        className: "ob-table-extend-btn ob-table-extend-btn--col ob-table-extend-btn--visible",
        style: {
          position: "fixed",
          left: tableRect.right + 4,
          top: tableRect.top,
          height: tableRect.height
        },
        onClick: () => handleAddCol(tableState.colCount),
        onMouseDown: (e) => e.preventDefault(),
        title: "Add column",
        children: /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", children: /* @__PURE__ */ jsx("path", { d: "M12 5v14M5 12h14" }) })
      }
    )
  ] });
}
function AlignmentButton({ active, onClick, title, children }) {
  return /* @__PURE__ */ jsx(
    "button",
    {
      type: "button",
      className: `ob-media-menu-btn ${active ? "ob-media-menu-btn--active" : ""}`,
      onClick,
      onMouseDown: (e) => e.preventDefault(),
      title,
      children
    }
  );
}
function UrlEditPopover({ currentUrl, onSave, onClose, label }) {
  const [url, setUrl] = useState(currentUrl);
  const inputRef = useRef(null);
  useEffect(() => {
    var _a, _b;
    (_a = inputRef.current) == null ? void 0 : _a.focus();
    (_b = inputRef.current) == null ? void 0 : _b.select();
  }, []);
  const handleSubmit = (e) => {
    e.preventDefault();
    onSave(url);
    onClose();
  };
  const handleKeyDown = (e) => {
    if (e.key === "Escape") {
      onClose();
    }
  };
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: "ob-media-url-popover",
      onKeyDown: handleKeyDown,
      onMouseDown: (e) => e.stopPropagation(),
      children: /* @__PURE__ */ jsxs("form", { onSubmit: handleSubmit, children: [
        /* @__PURE__ */ jsx("label", { className: "ob-media-url-label", children: label }),
        /* @__PURE__ */ jsx(
          "input",
          {
            ref: inputRef,
            type: "url",
            className: "ob-media-url-input",
            value: url,
            onChange: (e) => setUrl(e.target.value),
            placeholder: "https://..."
          }
        ),
        /* @__PURE__ */ jsxs("div", { className: "ob-media-url-actions", children: [
          /* @__PURE__ */ jsx(
            "button",
            {
              type: "button",
              className: "ob-media-url-btn ob-media-url-btn--cancel",
              onClick: onClose,
              children: "Cancel"
            }
          ),
          /* @__PURE__ */ jsx("button", { type: "submit", className: "ob-media-url-btn ob-media-url-btn--save", children: "Save" })
        ] })
      ] })
    }
  );
}
function CaptionEditPopover({ currentCaption, onSave, onClose }) {
  const [caption, setCaption] = useState(currentCaption);
  const inputRef = useRef(null);
  useEffect(() => {
    var _a, _b;
    (_a = inputRef.current) == null ? void 0 : _a.focus();
    (_b = inputRef.current) == null ? void 0 : _b.select();
  }, []);
  const handleSubmit = (e) => {
    e.preventDefault();
    onSave(caption);
    onClose();
  };
  const handleKeyDown = (e) => {
    if (e.key === "Escape") {
      onClose();
    }
  };
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: "ob-media-url-popover",
      onKeyDown: handleKeyDown,
      onMouseDown: (e) => e.stopPropagation(),
      children: /* @__PURE__ */ jsxs("form", { onSubmit: handleSubmit, children: [
        /* @__PURE__ */ jsx("label", { className: "ob-media-url-label", children: "Caption" }),
        /* @__PURE__ */ jsx(
          "input",
          {
            ref: inputRef,
            type: "text",
            className: "ob-media-url-input",
            value: caption,
            onChange: (e) => setCaption(e.target.value),
            placeholder: "Enter caption..."
          }
        ),
        /* @__PURE__ */ jsxs("div", { className: "ob-media-url-actions", children: [
          /* @__PURE__ */ jsx(
            "button",
            {
              type: "button",
              className: "ob-media-url-btn ob-media-url-btn--cancel",
              onClick: onClose,
              children: "Cancel"
            }
          ),
          /* @__PURE__ */ jsx("button", { type: "submit", className: "ob-media-url-btn ob-media-url-btn--save", children: "Save" })
        ] })
      ] })
    }
  );
}
function MediaMenu({ editor, className }) {
  const [menuState, setMenuState] = useState(null);
  const [showUrlEdit, setShowUrlEdit] = useState(false);
  const [showCaptionEdit, setShowCaptionEdit] = useState(false);
  const menuRef = useRef(null);
  const lastValidStateRef = useRef(null);
  useEffect(() => {
    if (!editor || editor.isDestroyed) return;
    const updateState = () => {
      const state = MEDIA_MENU_PLUGIN_KEY.getState(editor.pm.state);
      if (state == null ? void 0 : state.visible) {
        lastValidStateRef.current = state;
        setMenuState(state);
      } else if (!showUrlEdit && !showCaptionEdit) {
        lastValidStateRef.current = null;
        setMenuState(state ?? null);
      }
    };
    updateState();
    const unsubscribe = editor.on("transaction", updateState);
    return unsubscribe;
  }, [editor, showUrlEdit, showCaptionEdit]);
  useEffect(() => {
    if (!(menuState == null ? void 0 : menuState.visible) && !showUrlEdit && !showCaptionEdit) {
      lastValidStateRef.current = null;
    }
  }, [menuState == null ? void 0 : menuState.visible, showUrlEdit, showCaptionEdit]);
  useEffect(() => {
    if (!editor || !(menuState == null ? void 0 : menuState.visible)) return;
    const handleScroll = () => {
      editor.pm.view.dispatch(
        editor.pm.view.state.tr.setMeta(MEDIA_MENU_PLUGIN_KEY, { hide: true })
      );
    };
    const editorElement = editor.pm.view.dom;
    const scrollContainer = editorElement.closest(".openblock-container") || editorElement.parentElement;
    window.addEventListener("scroll", handleScroll, true);
    scrollContainer == null ? void 0 : scrollContainer.addEventListener("scroll", handleScroll);
    return () => {
      window.removeEventListener("scroll", handleScroll, true);
      scrollContainer == null ? void 0 : scrollContainer.removeEventListener("scroll", handleScroll);
    };
  }, [editor, menuState == null ? void 0 : menuState.visible]);
  useEffect(() => {
    const isPopoverOpen2 = showUrlEdit || showCaptionEdit;
    if (!editor || !(menuState == null ? void 0 : menuState.visible) && !isPopoverOpen2) return;
    const handleClickOutside = (e) => {
      const target = e.target;
      if (target.closest(".ob-media-menu") || target.closest(".ob-media-url-popover")) {
        return;
      }
      if (target.closest(".openblock-image") || target.closest(".openblock-embed")) {
        return;
      }
      setShowUrlEdit(false);
      setShowCaptionEdit(false);
      lastValidStateRef.current = null;
    };
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [editor, menuState == null ? void 0 : menuState.visible, showUrlEdit, showCaptionEdit]);
  const getActiveState = useCallback(() => {
    return (menuState == null ? void 0 : menuState.visible) ? menuState : lastValidStateRef.current;
  }, [menuState]);
  const handleAlignmentChange = useCallback(
    (alignment2) => {
      const state = getActiveState();
      if (!editor || (state == null ? void 0 : state.nodePos) === null || (state == null ? void 0 : state.nodePos) === void 0) return;
      updateMediaAttrs(editor.pm.view, state.nodePos, { alignment: alignment2 });
      editor.pm.view.focus();
    },
    [editor, getActiveState]
  );
  const handleUrlSave = useCallback(
    (url) => {
      const state = getActiveState();
      if (!editor || (state == null ? void 0 : state.nodePos) === null || (state == null ? void 0 : state.nodePos) === void 0) return;
      if (state.mediaType === "image") {
        updateMediaAttrs(editor.pm.view, state.nodePos, { src: url });
      } else {
        updateMediaAttrs(editor.pm.view, state.nodePos, { url });
      }
      setShowUrlEdit(false);
      editor.pm.view.focus();
    },
    [editor, getActiveState]
  );
  const handleCaptionSave = useCallback(
    (caption) => {
      const state = getActiveState();
      if (!editor || (state == null ? void 0 : state.nodePos) === null || (state == null ? void 0 : state.nodePos) === void 0) return;
      updateMediaAttrs(editor.pm.view, state.nodePos, { caption });
      setShowCaptionEdit(false);
      editor.pm.view.focus();
    },
    [editor, getActiveState]
  );
  const handleDelete = useCallback(() => {
    const state = getActiveState();
    if (!editor || (state == null ? void 0 : state.nodePos) === null || (state == null ? void 0 : state.nodePos) === void 0) return;
    deleteMediaNode(editor.pm.view, state.nodePos);
    editor.pm.view.focus();
  }, [editor, getActiveState]);
  const isPopoverOpen = showUrlEdit || showCaptionEdit;
  const activeState = (menuState == null ? void 0 : menuState.visible) ? menuState : isPopoverOpen ? lastValidStateRef.current : null;
  if (!editor || editor.isDestroyed || !activeState || !activeState.coords) {
    return null;
  }
  const { mediaType, attrs, coords } = activeState;
  const isImage = mediaType === "image";
  const imageAttrs = isImage ? attrs : null;
  const embedAttrs = !isImage ? attrs : null;
  const alignment = (imageAttrs == null ? void 0 : imageAttrs.alignment) || "center";
  const currentUrl = isImage ? (imageAttrs == null ? void 0 : imageAttrs.src) || "" : (embedAttrs == null ? void 0 : embedAttrs.url) || "";
  const currentCaption = (attrs == null ? void 0 : attrs.caption) || "";
  const menuHeight = 44;
  const style = {
    position: "fixed",
    left: coords.left + (coords.right - coords.left) / 2,
    top: coords.top - menuHeight - 8,
    transform: "translateX(-50%)",
    zIndex: 1e3
  };
  return /* @__PURE__ */ jsxs(
    "div",
    {
      ref: menuRef,
      className: `ob-media-menu ${className || ""}`,
      style,
      role: "toolbar",
      "aria-label": `${isImage ? "Image" : "Embed"} options`,
      children: [
        isImage && /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx(
            AlignmentButton,
            {
              active: alignment === "left",
              onClick: () => handleAlignmentChange("left"),
              title: "Align left",
              children: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
                /* @__PURE__ */ jsx("rect", { x: "3", y: "5", width: "10", height: "14", rx: "1" }),
                /* @__PURE__ */ jsx("path", { d: "M17 8h4M17 12h4M17 16h4" })
              ] })
            }
          ),
          /* @__PURE__ */ jsx(
            AlignmentButton,
            {
              active: alignment === "center",
              onClick: () => handleAlignmentChange("center"),
              title: "Align center",
              children: /* @__PURE__ */ jsx("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: /* @__PURE__ */ jsx("rect", { x: "7", y: "5", width: "10", height: "14", rx: "1" }) })
            }
          ),
          /* @__PURE__ */ jsx(
            AlignmentButton,
            {
              active: alignment === "right",
              onClick: () => handleAlignmentChange("right"),
              title: "Align right",
              children: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
                /* @__PURE__ */ jsx("rect", { x: "11", y: "5", width: "10", height: "14", rx: "1" }),
                /* @__PURE__ */ jsx("path", { d: "M3 8h4M3 12h4M3 16h4" })
              ] })
            }
          ),
          /* @__PURE__ */ jsx("span", { className: "ob-media-menu-divider" })
        ] }),
        /* @__PURE__ */ jsx(
          "button",
          {
            type: "button",
            className: `ob-media-menu-btn ${showUrlEdit ? "ob-media-menu-btn--active" : ""}`,
            onClick: () => {
              setShowUrlEdit(!showUrlEdit);
              setShowCaptionEdit(false);
            },
            onMouseDown: (e) => e.preventDefault(),
            title: isImage ? "Edit image URL" : "Edit embed URL",
            children: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
              /* @__PURE__ */ jsx("path", { d: "M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71" }),
              /* @__PURE__ */ jsx("path", { d: "M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71" })
            ] })
          }
        ),
        /* @__PURE__ */ jsx(
          "button",
          {
            type: "button",
            className: `ob-media-menu-btn ${showCaptionEdit ? "ob-media-menu-btn--active" : ""}`,
            onClick: () => {
              setShowCaptionEdit(!showCaptionEdit);
              setShowUrlEdit(false);
            },
            onMouseDown: (e) => e.preventDefault(),
            title: "Edit caption",
            children: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
              /* @__PURE__ */ jsx("rect", { x: "3", y: "5", width: "18", height: "14", rx: "2" }),
              /* @__PURE__ */ jsx("path", { d: "M7 15h10M7 11h4" })
            ] })
          }
        ),
        /* @__PURE__ */ jsx("span", { className: "ob-media-menu-divider" }),
        /* @__PURE__ */ jsx(
          "button",
          {
            type: "button",
            className: "ob-media-menu-btn ob-media-menu-btn--danger",
            onClick: handleDelete,
            onMouseDown: (e) => e.preventDefault(),
            title: "Delete",
            children: /* @__PURE__ */ jsxs("svg", { viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", children: [
              /* @__PURE__ */ jsx("path", { d: "M3 6h18M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6M8 6V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2" }),
              /* @__PURE__ */ jsx("line", { x1: "10", y1: "11", x2: "10", y2: "17" }),
              /* @__PURE__ */ jsx("line", { x1: "14", y1: "11", x2: "14", y2: "17" })
            ] })
          }
        ),
        showUrlEdit && /* @__PURE__ */ jsx(
          UrlEditPopover,
          {
            currentUrl,
            onSave: handleUrlSave,
            onClose: () => setShowUrlEdit(false),
            label: isImage ? "Image URL" : "Embed URL"
          }
        ),
        showCaptionEdit && /* @__PURE__ */ jsx(
          CaptionEditPopover,
          {
            currentCaption,
            onSave: handleCaptionSave,
            onClose: () => setShowCaptionEdit(false)
          }
        )
      ]
    }
  );
}
const BlockContext = React.createContext(null);
function useBlockEditor() {
  const context = React.useContext(BlockContext);
  return (context == null ? void 0 : context.editor) ?? null;
}
function createReactBlockSpec(spec, implementation) {
  const { type, propSchema, content } = spec;
  const { render: RenderComponent, slashMenu } = implementation;
  const attrs = {
    id: { default: null }
  };
  for (const [key, value] of Object.entries(propSchema)) {
    attrs[key] = { default: value.default };
  }
  const nodeSpec = {
    group: "block",
    content: content === "inline" ? "inline*" : "",
    atom: content === "none",
    attrs,
    parseDOM: [
      {
        tag: `div[data-block-type="${type}"]`,
        getAttrs: (dom) => {
          const result = {
            id: dom.getAttribute("data-block-id")
          };
          for (const key of Object.keys(propSchema)) {
            const attr = dom.getAttribute(`data-${key}`);
            if (attr !== null) {
              try {
                result[key] = JSON.parse(attr);
              } catch {
                result[key] = attr;
              }
            }
          }
          return result;
        }
      }
    ],
    toDOM: (node) => {
      const domAttrs = {
        "data-block-type": type,
        "data-block-id": node.attrs.id || "",
        class: `openblock-custom-block openblock-${type}`,
        contenteditable: "false"
      };
      for (const key of Object.keys(propSchema)) {
        const value = node.attrs[key];
        if (value !== void 0 && value !== null) {
          domAttrs[`data-${key}`] = typeof value === "object" ? JSON.stringify(value) : String(value);
        }
      }
      if (content === "inline") {
        return ["div", domAttrs, 0];
      }
      return ["div", domAttrs];
    }
  };
  const createNodeView = (editor) => {
    return (node, _view, _getPos, _decorations, _innerDecorations) => {
      const dom = document.createElement("div");
      dom.className = `openblock-custom-block openblock-${type}`;
      dom.setAttribute("data-block-type", type);
      dom.setAttribute("data-block-id", node.attrs.id || "");
      dom.contentEditable = "false";
      let contentDOM;
      if (content === "inline") {
        contentDOM = document.createElement("div");
        contentDOM.className = "openblock-block-content";
        contentDOM.contentEditable = "true";
      }
      let root = null;
      const reactContainer = document.createElement("div");
      reactContainer.className = "openblock-react-container";
      dom.appendChild(reactContainer);
      const renderReact = (currentNode) => {
        const block = {
          id: currentNode.attrs.id || "",
          type,
          props: {}
        };
        for (const key of Object.keys(propSchema)) {
          block.props[key] = currentNode.attrs[key];
        }
        const contentRef = React.createRef();
        const element = /* @__PURE__ */ jsx(BlockContext.Provider, { value: { editor }, children: /* @__PURE__ */ jsx(
          RenderComponent,
          {
            block,
            editor,
            isEditable: editor.isEditable,
            contentRef: content === "inline" ? contentRef : void 0
          }
        ) });
        if (!root) {
          root = createRoot(reactContainer);
        }
        root.render(element);
        if (content === "inline" && contentDOM) {
          const dom2 = contentDOM;
          requestAnimationFrame(() => {
            const contentContainer = contentRef.current;
            if (contentContainer && !contentContainer.contains(dom2)) {
              contentContainer.appendChild(dom2);
            }
          });
        }
      };
      renderReact(node);
      return {
        dom,
        contentDOM,
        update: (updatedNode) => {
          if (updatedNode.type.name !== type) return false;
          renderReact(updatedNode);
          return true;
        },
        destroy: () => {
          if (root) {
            root.unmount();
            root = null;
          }
        },
        stopEvent: (event) => {
          return event.target !== dom;
        },
        ignoreMutation: () => {
          return true;
        }
      };
    };
  };
  return {
    type,
    propSchema,
    content,
    nodeSpec,
    createNodeView,
    slashMenu
  };
}
function useUpdateBlock(editor, blockId) {
  return React.useCallback(
    (updates) => {
      if (!editor || editor.isDestroyed) return;
      let pos = null;
      editor.pm.doc.descendants((node, nodePos) => {
        if (node.attrs.id === blockId) {
          pos = nodePos;
          return false;
        }
      });
      if (pos !== null) {
        const node = editor.pm.doc.nodeAt(pos);
        if (node) {
          const tr = editor.pm.createTransaction();
          tr.setNodeMarkup(pos, void 0, { ...node.attrs, ...updates });
          editor.pm.dispatch(tr);
        }
      }
    },
    [editor, blockId]
  );
}
export {
  BUBBLE_MENU_ITEMS,
  BubbleMenu,
  ColorPicker,
  DEFAULT_BACKGROUND_COLORS,
  DEFAULT_BUBBLE_MENU_ORDER,
  DEFAULT_TEXT_COLORS,
  MediaMenu,
  OpenBlockView,
  SlashMenu,
  TableHandles,
  TableMenu,
  createReactBlockSpec,
  useBlockEditor,
  useCustomSlashMenuItems,
  useEditorContent,
  useEditorFocus,
  useEditorSelection,
  useOpenBlock,
  useUpdateBlock
};
