// .ignored/@glideapps/glide-data-grid-cells/dist/js/cells/tags-cell.js
import { measureTextCached } from "@glideapps/glide-data-grid";
import styled from "styled-components";
import * as React from "react";

function roundedRect(ctx, x, y, width, height, radius) {
  ctx.moveTo(x + radius, y);
  ctx.arcTo(x + width, y, x + width, y + radius, radius);
  ctx.arcTo(x + width, y + height, x + width - radius, y + height, radius);
  ctx.arcTo(x, y + height, x, y + height - radius, radius);
  ctx.arcTo(x, y, x + radius, y, radius);
}

const tagHeight = 20;
const innerPad = 6;
const EditorWrap = styled.div.withConfig({
  displayName: "tags-cell__EditorWrap",
  componentId: "sc-zfe4jh-0"
})(["display:flex;flex-direction:column;align-items:stretch;padding-top:6px;color:", ";box-sizing:border-box;*{box-sizing:border-box;}&&&& label{display:flex;cursor:pointer;input{cursor:pointer;width:auto;}.pill{margin-left:8px;margin-right:6px;margin-bottom:6px;border-radius:", "px;min-height:", "px;padding:2px ", "px;display:flex;align-items:center;font:12px ", ";background-color:", ";transition:box-shadow 150ms;&.unselected{opacity:0.8;}}}label:hover .pill{box-shadow:0 1px 4px rgba(0,0,0,0.15);}&&&&.readonly label{cursor:default;.pill{box-shadow:none !important;}}"], p => p.theme.textDark, tagHeight / 2, tagHeight, innerPad, p => p.theme.fontFamily, p => p.theme.bgBubble);
const renderer = {
  isMatch: c => c.data.kind === "tags-cell",
  draw: (args, cell) => {
    const {
      ctx,
      theme,
      rect
    } = args;
    const {
      possibleTags,
      tags
    } = cell.data;
    const drawArea = {
      x: rect.x + theme.cellHorizontalPadding,
      y: rect.y + theme.cellVerticalPadding,
      width: rect.width - 2 * theme.cellHorizontalPadding,
      height: rect.height - 2 * theme.cellVerticalPadding
    };
    const rows = Math.max(1, Math.floor(drawArea.height / (tagHeight + innerPad)));
    let x = drawArea.x;
    let row = 1;
    let y = drawArea.y + (drawArea.height - rows * tagHeight - (rows - 1) * innerPad) / 2;

    for (const tag of tags) {
      var _possibleTags$find$co, _possibleTags$find;

      const color = (_possibleTags$find$co = (_possibleTags$find = possibleTags.find(t => t.tag === tag)) === null || _possibleTags$find === void 0 ? void 0 : _possibleTags$find.color) !== null && _possibleTags$find$co !== void 0 ? _possibleTags$find$co : theme.bgBubble;
      ctx.font = `12px ${theme.fontFamily}`;
      const metrics = measureTextCached(tag, ctx);
      const width = metrics.width + innerPad * 2;
      const textY = tagHeight / 2;

      if (x !== drawArea.x && x + width > drawArea.x + drawArea.width && row < rows) {
        row++;
        y += tagHeight + innerPad;
        x = drawArea.x;
      }

      ctx.fillStyle = color;
      ctx.beginPath();
      roundedRect(ctx, x, y, width, tagHeight, tagHeight / 2);
      ctx.fill();
      ctx.fillStyle = theme.textDark;
      ctx.fillText(tag, x + innerPad, y + textY);
      x += width + 8;
      if (x > drawArea.x + drawArea.width && row >= rows) break;
    }

    return true;
  },
  provideEditor: () => {
    return p => {
      const {
        onChange,
        value
      } = p;
      const {
        possibleTags,
        tags,
        readonly = false
      } = value.data;
      return React.createElement(EditorWrap, {
        className: readonly ? "readonly" : ""
      }, possibleTags.map(t => {
        const selected = tags.indexOf(t.tag) !== -1;
        return React.createElement("label", {
          key: t.tag
        }, !readonly && React.createElement("input", {
          type: "checkbox",
          checked: selected,
          onChange: () => {
            const newTags = selected ? tags.filter(x => x !== t.tag) : [...tags, t.tag];
            onChange({ ...p.value,
              data: { ...value.data,
                tags: newTags
              }
            });
          }
        }), React.createElement("div", {
          className: "pill " + (selected ? "selected" : "unselected"),
          style: {
            backgroundColor: selected ? t.color : undefined
          }
        }, t.tag));
      }));
    };
  }
};
export default renderer;
