// .ignored/@glideapps/glide-data-grid-cells/dist/js/cells/user-profile-cell.js
import * as React from "react";
import { measureTextCached, TextCellEntry } from "@glideapps/glide-data-grid";
const renderer = {
  isMatch: cell => cell.data.kind === "user-profile-cell",
  draw: (args, cell) => {
    const {
      ctx,
      rect,
      theme,
      imageLoader,
      col,
      row
    } = args;
    const {
      image,
      name,
      initial,
      tint
    } = cell.data;
    const xPad = theme.cellHorizontalPadding;
    const radius = Math.min(12, rect.height / 2 - theme.cellVerticalPadding);
    const drawX = rect.x + xPad;
    const imageResult = imageLoader.loadOrGetImage(image, col, row);
    ctx.save();
    ctx.beginPath();
    ctx.arc(drawX + radius, rect.y + rect.height / 2, radius, 0, Math.PI * 2);
    ctx.globalAlpha = 0.2;
    ctx.fillStyle = tint;
    ctx.fill();
    ctx.globalAlpha = 1;
    ctx.font = `600 16px ${theme.fontFamily}`;
    const metrics = measureTextCached(initial[0], ctx);
    ctx.fillText(initial[0], drawX + radius - metrics.width / 2, rect.y + rect.height / 2 + metrics.actualBoundingBoxAscent / 2);

    if (imageResult !== undefined) {
      ctx.save();
      ctx.beginPath();
      ctx.arc(drawX + radius, rect.y + rect.height / 2, radius, 0, Math.PI * 2);
      ctx.clip();
      ctx.drawImage(imageResult, drawX, rect.y + rect.height / 2 - radius, radius * 2, radius * 2);
      ctx.restore();
    }

    if (name !== undefined) {
      ctx.font = `${theme.baseFontStyle} ${theme.fontFamily}`;
      ctx.fillStyle = theme.textDark;
      ctx.textBaseline = "middle";
      ctx.fillText(name, drawX + radius * 2 + xPad, rect.y + rect.height / 2);
    }

    ctx.restore();
    return true;
  },
  provideEditor: () => p => {
    var _value$data$name;

    const {
      isHighlighted,
      onChange,
      value
    } = p;
    return React.createElement(TextCellEntry, {
      highlight: isHighlighted,
      autoFocus: true,
      value: (_value$data$name = value.data.name) !== null && _value$data$name !== void 0 ? _value$data$name : "",
      onChange: e => onChange({ ...value,
        data: { ...value.data,
          name: e.target.value
        }
      })
    });
  }
};
export default renderer;
