// .ignored/@labbs/openblock-core/dist/index.js
import { TextSelection, NodeSelection, AllSelection, PluginKey, Plugin, EditorState } from "prosemirror-state";
import { AllSelection as AllSelection2, EditorState as EditorState2, NodeSelection as NodeSelection2, Plugin as Plugin2, PluginKey as PluginKey2, Selection, TextSelection as TextSelection2, Transaction } from "prosemirror-state";
import { v4 } from "uuid";
import { keymap } from "prosemirror-keymap";
import { toggleMark, setBlockType, lift, wrapIn, baseKeymap } from "prosemirror-commands";
import { liftListItem, sinkListItem, wrapInList } from "prosemirror-schema-list";
import { redo, undo, history } from "prosemirror-history";
import { redo as redo2, undo as undo2 } from "prosemirror-history";
import { Fragment, Slice, Schema } from "prosemirror-model";
import { Fragment as Fragment2, Mark, MarkType, Node, NodeType, ResolvedPos, Schema as Schema2, Slice as Slice2 } from "prosemirror-model";
import { DecorationSet, Decoration, EditorView } from "prosemirror-view";
import { Decoration as Decoration2, DecorationSet as DecorationSet2, EditorView as EditorView2 } from "prosemirror-view";
import { dropCursor } from "prosemirror-dropcursor";
import { gapCursor } from "prosemirror-gapcursor";
import { wrappingInputRule, InputRule, textblockTypeInputRule, inputRules } from "prosemirror-inputrules";
class ProseMirrorAPI {
  constructor(editor) {
    this.editor = editor;
    this.decorations = {
      /**
       * Create a widget decoration (inserts a DOM element at a position)
       */
      widget: (pos, toDOM, spec) => {
        return Decoration.widget(pos, toDOM, spec);
      },
      /**
       * Create an inline decoration (adds attributes to text)
       */
      inline: (from, to, attrs) => {
        return Decoration.inline(from, to, attrs);
      },
      /**
       * Create a node decoration (adds attributes to a node)
       */
      node: (from, to, attrs) => {
        return Decoration.node(from, to, attrs);
      },
      /**
       * Create a decoration set from decorations
       */
      set: (doc, decorations) => {
        return DecorationSet.create(doc, decorations);
      },
      /**
       * Create an empty decoration set
       */
      empty: DecorationSet.empty
    };
  }
  // ===========================================================================
  // VIEW ACCESS
  // ===========================================================================
  /**
   * The ProseMirror EditorView instance
   *
   * This is the main view that renders the editor and handles user input.
   * Direct access - no private API hacking needed!
   */
  get view() {
    return this.editor["_pmView"];
  }
  /**
   * The current EditorState
   *
   * Contains the document, selection, and all plugin states.
   */
  get state() {
    return this.view.state;
  }
  /**
   * The ProseMirror Schema
   *
   * Defines the structure of the document (node types, marks, etc.)
   */
  get schema() {
    return this.state.schema;
  }
  /**
   * The current document
   *
   * The root node containing all content.
   */
  get doc() {
    return this.state.doc;
  }
  /**
   * Whether the editor currently has focus
   */
  get hasFocus() {
    return this.view.hasFocus();
  }
  /**
   * The DOM element containing the editor
   */
  get dom() {
    return this.view.dom;
  }
  // ===========================================================================
  // TRANSACTIONS
  // ===========================================================================
  /**
   * Create a new transaction
   *
   * Transactions are the way to modify the editor state.
   * ```typescript
   * const tr = editor.pm.createTransaction();
   * tr.insertText('Hello');
   * editor.pm.dispatch(tr);
   * ```
   */
  createTransaction() {
    return this.state.tr;
  }
  /**
   * Dispatch a transaction to update the editor state
   *
   * This is the core method for making changes to the document.
   * @param tr - The transaction to apply
   */
  dispatch(tr) {
    this.view.dispatch(tr);
  }
  /**
   * Create and dispatch a transaction in one call
   *
   * Convenient for simple transformations:
   * ```typescript
   * editor.pm.dispatchWith(tr => tr.insertText('Hello'));
   * ```
   */
  dispatchWith(fn) {
    this.dispatch(fn(this.createTransaction()));
  }
  // ===========================================================================
  // POSITION UTILITIES
  // ===========================================================================
  /**
   * Resolve a position to get detailed information
   *
   * Returns a ResolvedPos with parent, depth, and other context.
   */
  resolve(pos) {
    return this.doc.resolve(pos);
  }
  /**
   * Get screen coordinates for a document position
   *
   * Useful for positioning tooltips, menus, etc.
   */
  coordsAtPos(pos, side) {
    return this.view.coordsAtPos(pos, side);
  }
  /**
   * Get document position from screen coordinates
   *
   * The inverse of coordsAtPos. Returns null if outside the editor.
   */
  posAtCoords(coords) {
    return this.view.posAtCoords(coords);
  }
  /**
   * Get document position from a DOM node and offset
   *
   * Useful when handling DOM events.
   */
  posAtDOM(node, offset, bias) {
    return this.view.posAtDOM(node, offset, bias);
  }
  /**
   * Get the DOM node at a document position
   *
   * Returns null if no node is found.
   */
  nodeDOM(pos) {
    return this.view.nodeDOM(pos);
  }
  /**
   * Get the DOM element for a node at a position
   *
   * Convenience method that casts to HTMLElement.
   */
  domAtPos(pos) {
    return this.view.domAtPos(pos);
  }
  // ===========================================================================
  // NODE OPERATIONS
  // ===========================================================================
  /**
   * Get the node at a specific position
   *
   * Returns null if there's no node at that position.
   */
  nodeAt(pos) {
    return this.doc.nodeAt(pos);
  }
  /**
   * Get a node type by name
   */
  getNodeType(name) {
    return this.schema.nodes[name];
  }
  /**
   * Get a mark type by name
   */
  getMarkType(name) {
    return this.schema.marks[name];
  }
  /**
   * Set attributes on a node
   *
   * This is one of the key methods that was hard to do in BlockNote!
   * ```typescript
   * // Resize a column
   * editor.pm.setNodeAttrs(columnPos, { ratio: 2 });
   * ```
   */
  setNodeAttrs(pos, attrs) {
    const node = this.nodeAt(pos);
    if (!node) return;
    const tr = this.createTransaction();
    tr.setNodeMarkup(pos, void 0, { ...node.attrs, ...attrs });
    this.dispatch(tr);
  }
  /**
   * Replace a node with another node
   */
  replaceNode(pos, newNode) {
    const $pos = this.resolve(pos);
    const oldNode = $pos.nodeAfter;
    if (!oldNode) return;
    const tr = this.createTransaction();
    tr.replaceWith(pos, pos + oldNode.nodeSize, newNode);
    this.dispatch(tr);
  }
  /**
   * Delete a node at a position
   */
  deleteNode(pos) {
    const $pos = this.resolve(pos);
    const node = $pos.nodeAfter;
    if (!node) return;
    const tr = this.createTransaction();
    tr.delete(pos, pos + node.nodeSize);
    this.dispatch(tr);
  }
  /**
   * Create a node from the schema
   */
  createNode(type, attrs, content, marks) {
    const nodeType = typeof type === "string" ? this.getNodeType(type) : type;
    return nodeType.create(attrs, content, marks);
  }
  /**
   * Create a text node
   */
  createTextNode(text, marks) {
    return this.schema.text(text, marks);
  }
  // ===========================================================================
  // SELECTION
  // ===========================================================================
  /**
   * The current selection
   */
  get selection() {
    return this.state.selection;
  }
  /**
   * Set the selection
   */
  setSelection(selection) {
    const tr = this.createTransaction();
    tr.setSelection(selection);
    this.dispatch(tr);
  }
  /**
   * Create a text selection
   */
  createTextSelection(anchor, head) {
    return TextSelection.create(this.doc, anchor, head ?? anchor);
  }
  /**
   * Create a node selection
   */
  createNodeSelection(pos) {
    return NodeSelection.create(this.doc, pos);
  }
  /**
   * Create an all selection (selects entire document)
   */
  createAllSelection() {
    return new AllSelection(this.doc);
  }
  /**
   * Get the selected text as a string
   */
  getSelectedText() {
    const { from, to } = this.selection;
    return this.doc.textBetween(from, to);
  }
  /**
   * Check if there's an active text selection (not just a cursor)
   */
  hasSelection() {
    return !this.selection.empty;
  }
  // ===========================================================================
  // MARKS (Inline Formatting)
  // ===========================================================================
  /**
   * Get the active marks at the current selection
   */
  activeMarks() {
    return this.state.storedMarks || this.selection.$from.marks();
  }
  /**
   * Check if a mark is active
   */
  isMarkActive(markType) {
    const type = typeof markType === "string" ? this.getMarkType(markType) : markType;
    const { from, to, empty } = this.selection;
    if (empty) {
      return !!type.isInSet(this.activeMarks());
    }
    return this.doc.rangeHasMark(from, to, type);
  }
  /**
   * Toggle a mark on the current selection
   */
  toggleMark(markType, attrs) {
    const type = typeof markType === "string" ? this.getMarkType(markType) : markType;
    return toggleMark(type, attrs)(this.state, this.dispatch.bind(this));
  }
  /**
   * Add a mark to the current selection
   */
  addMark(markType, attrs) {
    const type = typeof markType === "string" ? this.getMarkType(markType) : markType;
    const { from, to } = this.selection;
    if (from === to) return;
    const tr = this.createTransaction();
    tr.addMark(from, to, type.create(attrs));
    this.dispatch(tr);
  }
  /**
   * Remove a mark from the current selection
   */
  removeMark(markType) {
    const type = typeof markType === "string" ? this.getMarkType(markType) : markType;
    const { from, to } = this.selection;
    if (from === to) return;
    const tr = this.createTransaction();
    tr.removeMark(from, to, type);
    this.dispatch(tr);
  }
  // ===========================================================================
  // PLUGINS
  // ===========================================================================
  /**
   * Get the state of a plugin
   */
  getPluginState(key) {
    return key.getState(this.state);
  }
  /**
   * Add a plugin dynamically
   *
   * This is another key feature that's hard in BlockNote!
   */
  addPlugin(plugin) {
    const plugins = [...this.state.plugins, plugin];
    const newState = this.state.reconfigure({ plugins });
    this.view.updateState(newState);
  }
  /**
   * Remove a plugin
   */
  removePlugin(pluginKey) {
    const plugins = this.state.plugins.filter((p) => p.spec.key !== pluginKey);
    const newState = this.state.reconfigure({ plugins });
    this.view.updateState(newState);
  }
  /**
   * Get all registered plugins
   */
  getPlugins() {
    return this.state.plugins;
  }
  // ===========================================================================
  // COMMANDS
  // ===========================================================================
  /**
   * Run a ProseMirror command
   *
   * Commands are functions that can modify the editor state.
   */
  runCommand(command) {
    return command(this.state, this.dispatch.bind(this), this.view);
  }
  // ===========================================================================
  // SLICES AND FRAGMENTS
  // ===========================================================================
  /**
   * Create a slice from nodes
   */
  createSlice(content, openStart = 0, openEnd = 0) {
    const fragment = Array.isArray(content) ? Fragment.from(content) : content instanceof Fragment ? content : Fragment.from(content);
    return new Slice(fragment, openStart, openEnd);
  }
  /**
   * Create a fragment from nodes
   */
  createFragment(nodes) {
    return Fragment.from(nodes);
  }
  // ===========================================================================
  // CONTENT INSERTION
  // ===========================================================================
  /**
   * Insert text at the current cursor position
   */
  insertText(text) {
    const tr = this.createTransaction();
    tr.insertText(text);
    this.dispatch(tr);
  }
  /**
   * Insert a node at the current cursor position
   */
  insertNode(node) {
    const tr = this.createTransaction();
    tr.replaceSelectionWith(node);
    this.dispatch(tr);
  }
  /**
   * Replace the selection with content
   */
  replaceSelection(content) {
    const tr = this.createTransaction();
    if (content instanceof Slice) {
      tr.replaceSelection(content);
    } else if (content instanceof Fragment) {
      tr.replaceSelection(new Slice(content, 0, 0));
    } else if (Array.isArray(content)) {
      tr.replaceSelection(new Slice(Fragment.from(content), 0, 0));
    } else {
      tr.replaceSelectionWith(content);
    }
    this.dispatch(tr);
  }
  // ===========================================================================
  // DOCUMENT TRAVERSAL
  // ===========================================================================
  /**
   * Iterate over all nodes in the document
   *
   * Callback receives node, position, parent, and index.
   */
  forEach(callback) {
    this.doc.descendants((node, pos, parent, index) => {
      return callback(node, pos, parent, index);
    });
  }
  /**
   * Find nodes matching a predicate
   */
  findNodes(predicate) {
    const result = [];
    this.forEach((node, pos) => {
      if (predicate(node)) {
        result.push({ node, pos });
      }
    });
    return result;
  }
  /**
   * Find nodes by type
   */
  findNodesByType(type) {
    const nodeType = typeof type === "string" ? this.getNodeType(type) : type;
    return this.findNodes((node) => node.type === nodeType);
  }
}
const defaultConfig$1 = {
  editable: true,
  autoFocus: false,
  injectStyles: true
};
const docNode = {
  content: "block+"
};
const paragraphNode = {
  content: "inline*",
  group: "block",
  attrs: {
    id: { default: null },
    textAlign: { default: "left" }
  },
  parseDOM: [{
    tag: "p",
    getAttrs(dom) {
      const element = dom;
      return {
        textAlign: element.style.textAlign || "left"
      };
    }
  }],
  toDOM(node) {
    const attrs = { "data-block-id": node.attrs.id };
    if (node.attrs.textAlign && node.attrs.textAlign !== "left") {
      attrs.style = `text-align: ${node.attrs.textAlign}`;
    }
    return ["p", attrs, 0];
  }
};
const headingNode = {
  content: "inline*",
  group: "block",
  attrs: {
    id: { default: null },
    level: { default: 1 },
    textAlign: { default: "left" }
  },
  parseDOM: [
    { tag: "h1", getAttrs: (dom) => ({ level: 1, textAlign: dom.style.textAlign || "left" }) },
    { tag: "h2", getAttrs: (dom) => ({ level: 2, textAlign: dom.style.textAlign || "left" }) },
    { tag: "h3", getAttrs: (dom) => ({ level: 3, textAlign: dom.style.textAlign || "left" }) },
    { tag: "h4", getAttrs: (dom) => ({ level: 4, textAlign: dom.style.textAlign || "left" }) },
    { tag: "h5", getAttrs: (dom) => ({ level: 5, textAlign: dom.style.textAlign || "left" }) },
    { tag: "h6", getAttrs: (dom) => ({ level: 6, textAlign: dom.style.textAlign || "left" }) }
  ],
  toDOM(node) {
    const attrs = { "data-block-id": node.attrs.id };
    if (node.attrs.textAlign && node.attrs.textAlign !== "left") {
      attrs.style = `text-align: ${node.attrs.textAlign}`;
    }
    return [`h${node.attrs.level}`, attrs, 0];
  }
};
const textNode = {
  group: "inline"
};
const hardBreakNode = {
  inline: true,
  group: "inline",
  selectable: false,
  parseDOM: [{ tag: "br" }],
  toDOM() {
    return ["br"];
  }
};
const blockquoteNode = {
  content: "inline*",
  group: "block",
  attrs: {
    id: { default: null }
  },
  parseDOM: [{ tag: "blockquote" }],
  toDOM(node) {
    return ["blockquote", { class: "openblock-blockquote", "data-block-id": node.attrs.id }, 0];
  }
};
const calloutNode = {
  content: "inline*",
  group: "block",
  attrs: {
    id: { default: null },
    calloutType: { default: "info" }
  },
  parseDOM: [
    {
      tag: "div.openblock-callout",
      getAttrs(dom) {
        const element = dom;
        return {
          calloutType: element.dataset.calloutType || "info"
        };
      }
    }
  ],
  toDOM(node) {
    const calloutType = node.attrs.calloutType;
    return [
      "div",
      {
        class: `openblock-callout openblock-callout--${calloutType}`,
        "data-block-id": node.attrs.id,
        "data-callout-type": calloutType
      },
      0
    ];
  }
};
const codeBlockNode = {
  content: "text*",
  group: "block",
  // Code blocks don't allow marks - code is plain text
  marks: "",
  // Treat as a single unit for cursor navigation
  code: true,
  // Preserve whitespace exactly as entered
  defining: true,
  attrs: {
    id: { default: null },
    language: { default: "" }
  },
  parseDOM: [
    {
      tag: "pre",
      preserveWhitespace: "full",
      getAttrs(node) {
        var _a;
        const pre = node;
        const code = pre.querySelector("code");
        const lang = (code == null ? void 0 : code.getAttribute("data-language")) || ((_a = code == null ? void 0 : code.className.match(/language-(\w+)/)) == null ? void 0 : _a[1]) || "";
        return { language: lang };
      }
    }
  ],
  toDOM(node) {
    const language = node.attrs.language || "";
    return [
      "pre",
      { class: "openblock-code-block", "data-block-id": node.attrs.id },
      ["code", { "data-language": language, class: language ? `language-${language}` : "" }, 0]
    ];
  }
};
const dividerNode = {
  // No content - this is a leaf node
  group: "block",
  attrs: {
    id: { default: null }
  },
  parseDOM: [{ tag: "hr" }],
  toDOM(node) {
    return ["hr", { class: "openblock-divider", "data-block-id": node.attrs.id }];
  }
};
const bulletListNode = {
  content: "listItem+",
  group: "block",
  attrs: {
    id: { default: null }
  },
  parseDOM: [{ tag: "ul" }],
  toDOM(node) {
    return ["ul", { class: "openblock-bullet-list", "data-block-id": node.attrs.id }, 0];
  }
};
const orderedListNode = {
  content: "listItem+",
  group: "block",
  attrs: {
    id: { default: null },
    start: { default: 1 }
  },
  parseDOM: [
    {
      tag: "ol",
      getAttrs(node) {
        const ol = node;
        return { start: ol.hasAttribute("start") ? parseInt(ol.getAttribute("start") ?? "1", 10) : 1 };
      }
    }
  ],
  toDOM(node) {
    const attrs = {
      class: "openblock-ordered-list",
      "data-block-id": node.attrs.id
    };
    if (node.attrs.start !== 1) {
      attrs.start = String(node.attrs.start);
    }
    return ["ol", attrs, 0];
  }
};
const listItemNode = {
  // Can contain a paragraph followed by optional nested lists
  content: "paragraph block*",
  attrs: {
    id: { default: null }
  },
  // Allow list items to define their own boundary for operations
  defining: true,
  parseDOM: [{ tag: "li" }],
  toDOM(node) {
    return ["li", { class: "openblock-list-item", "data-block-id": node.attrs.id }, 0];
  }
};
const columnListNode = {
  group: "block",
  content: "column+",
  defining: true,
  isolating: true,
  attrs: {
    /** Unique block identifier */
    id: { default: null },
    /** Gap between columns in pixels */
    gap: { default: 16 }
  },
  parseDOM: [
    {
      tag: "div[data-column-list]",
      getAttrs(dom) {
        const gap = dom.getAttribute("data-gap");
        const id = dom.getAttribute("data-block-id");
        return {
          id: id || null,
          gap: gap ? parseInt(gap, 10) : 16
        };
      }
    }
  ],
  toDOM(node) {
    const attrs = {
      "data-column-list": "",
      "data-gap": String(node.attrs.gap),
      class: "ob-column-list",
      style: `gap: ${node.attrs.gap}px`
    };
    if (node.attrs.id) {
      attrs["data-block-id"] = node.attrs.id;
    }
    return ["div", attrs, 0];
  }
};
const columnNode = {
  // Note: columns are not in the 'block' group to prevent them from being
  // inserted outside of a columnList
  content: "block+",
  defining: true,
  isolating: true,
  attrs: {
    /** Unique block identifier */
    id: { default: null },
    /** Width as a percentage (1-100) */
    width: { default: 50 }
  },
  parseDOM: [
    {
      tag: "div[data-column]",
      getAttrs(dom) {
        const width = dom.getAttribute("data-width");
        const id = dom.getAttribute("data-block-id");
        return {
          id: id || null,
          width: width ? parseInt(width, 10) : 50
        };
      }
    }
  ],
  toDOM(node) {
    const attrs = {
      "data-column": "",
      "data-width": String(node.attrs.width),
      class: "ob-column",
      style: `flex: ${node.attrs.width} 0 0; min-width: 0;`
    };
    if (node.attrs.id) {
      attrs["data-block-id"] = node.attrs.id;
    }
    return ["div", attrs, 0];
  }
};
const tableNode = {
  group: "block",
  content: "tableRow+",
  tableRole: "table",
  defining: true,
  isolating: true,
  attrs: {
    /** Unique block identifier */
    id: { default: null }
  },
  parseDOM: [
    {
      tag: "table",
      getAttrs(dom) {
        const id = dom.getAttribute("data-block-id");
        return {
          id: id || null
        };
      }
    }
  ],
  toDOM(node) {
    const attrs = {
      class: "ob-table"
    };
    if (node.attrs.id) {
      attrs["data-block-id"] = node.attrs.id;
    }
    return ["table", attrs, ["tbody", 0]];
  }
};
const tableRowNode = {
  content: "(tableCell | tableHeader)+",
  tableRole: "row",
  attrs: {
    /** Unique block identifier */
    id: { default: null }
  },
  parseDOM: [
    {
      tag: "tr",
      getAttrs(dom) {
        const id = dom.getAttribute("data-block-id");
        return {
          id: id || null
        };
      }
    }
  ],
  toDOM(node) {
    const attrs = {
      class: "ob-table-row"
    };
    if (node.attrs.id) {
      attrs["data-block-id"] = node.attrs.id;
    }
    return ["tr", attrs, 0];
  }
};
const tableCellNode = {
  content: "block+",
  tableRole: "cell",
  isolating: true,
  attrs: {
    /** Unique block identifier */
    id: { default: null },
    /** Number of columns this cell spans */
    colspan: { default: 1 },
    /** Number of rows this cell spans */
    rowspan: { default: 1 },
    /** Column width in pixels (null = auto) */
    colwidth: { default: null },
    /** Background color */
    backgroundColor: { default: null }
  },
  parseDOM: [
    {
      tag: "td",
      getAttrs(dom) {
        const id = dom.getAttribute("data-block-id");
        const colspan = dom.getAttribute("colspan");
        const rowspan = dom.getAttribute("rowspan");
        const colwidth = dom.getAttribute("data-colwidth");
        const backgroundColor = dom.style.backgroundColor || null;
        return {
          id: id || null,
          colspan: colspan ? parseInt(colspan, 10) : 1,
          rowspan: rowspan ? parseInt(rowspan, 10) : 1,
          colwidth: colwidth ? colwidth.split(",").map((w) => parseInt(w, 10)) : null,
          backgroundColor
        };
      }
    },
    {
      tag: "th",
      getAttrs(dom) {
        const id = dom.getAttribute("data-block-id");
        const colspan = dom.getAttribute("colspan");
        const rowspan = dom.getAttribute("rowspan");
        const colwidth = dom.getAttribute("data-colwidth");
        const backgroundColor = dom.style.backgroundColor || null;
        return {
          id: id || null,
          colspan: colspan ? parseInt(colspan, 10) : 1,
          rowspan: rowspan ? parseInt(rowspan, 10) : 1,
          colwidth: colwidth ? colwidth.split(",").map((w) => parseInt(w, 10)) : null,
          backgroundColor
        };
      }
    }
  ],
  toDOM(node) {
    const attrs = {
      class: "ob-table-cell"
    };
    if (node.attrs.id) {
      attrs["data-block-id"] = node.attrs.id;
    }
    if (node.attrs.colspan > 1) {
      attrs.colspan = String(node.attrs.colspan);
    }
    if (node.attrs.rowspan > 1) {
      attrs.rowspan = String(node.attrs.rowspan);
    }
    if (node.attrs.colwidth) {
      attrs["data-colwidth"] = node.attrs.colwidth.join(",");
      attrs.style = `width: ${node.attrs.colwidth[0]}px`;
    }
    if (node.attrs.backgroundColor) {
      attrs.style = (attrs.style || "") + `background-color: ${node.attrs.backgroundColor}`;
    }
    return ["td", attrs, 0];
  }
};
const tableHeaderNode = {
  content: "block+",
  tableRole: "header_cell",
  isolating: true,
  attrs: {
    /** Unique block identifier */
    id: { default: null },
    /** Number of columns this cell spans */
    colspan: { default: 1 },
    /** Number of rows this cell spans */
    rowspan: { default: 1 },
    /** Column width in pixels (null = auto) */
    colwidth: { default: null },
    /** Background color */
    backgroundColor: { default: null }
  },
  parseDOM: [
    {
      tag: "th",
      getAttrs(dom) {
        const id = dom.getAttribute("data-block-id");
        const colspan = dom.getAttribute("colspan");
        const rowspan = dom.getAttribute("rowspan");
        const colwidth = dom.getAttribute("data-colwidth");
        const backgroundColor = dom.style.backgroundColor || null;
        return {
          id: id || null,
          colspan: colspan ? parseInt(colspan, 10) : 1,
          rowspan: rowspan ? parseInt(rowspan, 10) : 1,
          colwidth: colwidth ? colwidth.split(",").map((w) => parseInt(w, 10)) : null,
          backgroundColor
        };
      }
    }
  ],
  toDOM(node) {
    const attrs = {
      class: "ob-table-header"
    };
    if (node.attrs.id) {
      attrs["data-block-id"] = node.attrs.id;
    }
    if (node.attrs.colspan > 1) {
      attrs.colspan = String(node.attrs.colspan);
    }
    if (node.attrs.rowspan > 1) {
      attrs.rowspan = String(node.attrs.rowspan);
    }
    if (node.attrs.colwidth) {
      attrs["data-colwidth"] = node.attrs.colwidth.join(",");
      attrs.style = `width: ${node.attrs.colwidth[0]}px`;
    }
    if (node.attrs.backgroundColor) {
      attrs.style = (attrs.style || "") + `background-color: ${node.attrs.backgroundColor}`;
    }
    return ["th", attrs, 0];
  }
};
const imageNode = {
  group: "block",
  // Image is atomic - no editable content inside
  atom: true,
  draggable: true,
  attrs: {
    id: { default: null },
    src: { default: "" },
    alt: { default: "" },
    caption: { default: "" },
    width: { default: null },
    alignment: { default: "center" }
  },
  parseDOM: [
    {
      tag: "figure.openblock-image",
      getAttrs: (dom) => {
        var _a;
        const element = dom;
        const img = element.querySelector("img");
        return {
          id: element.getAttribute("data-block-id"),
          src: (img == null ? void 0 : img.getAttribute("src")) || "",
          alt: (img == null ? void 0 : img.getAttribute("alt")) || "",
          caption: ((_a = element.querySelector("figcaption")) == null ? void 0 : _a.textContent) || "",
          width: (img == null ? void 0 : img.getAttribute("data-width")) ? parseInt(img.getAttribute("data-width") ?? "0") : null,
          alignment: element.getAttribute("data-alignment") || "center"
        };
      }
    },
    {
      tag: "img[src]",
      getAttrs: (dom) => {
        const element = dom;
        return {
          src: element.getAttribute("src") || "",
          alt: element.getAttribute("alt") || "",
          width: element.width || null
        };
      }
    }
  ],
  toDOM: (node) => {
    const { src, alt, caption, width, alignment, id } = node.attrs;
    const figureAttrs = {
      class: `openblock-image openblock-image--${alignment}`,
      "data-block-id": id || "",
      "data-alignment": alignment
    };
    if (!src) {
      const placeholderDiv = [
        "div",
        { class: "openblock-image-placeholder" },
        ["span", { class: "openblock-image-placeholder-icon" }],
        ["span", { class: "openblock-image-placeholder-text" }, "Click to add an image"]
      ];
      if (caption) {
        return ["figure", figureAttrs, placeholderDiv, ["figcaption", {}, caption]];
      }
      return ["figure", figureAttrs, placeholderDiv];
    }
    const imgAttrs = { src, alt };
    if (width) {
      imgAttrs["data-width"] = String(width);
      imgAttrs.style = `width: ${width}px`;
    }
    if (caption) {
      return [
        "figure",
        figureAttrs,
        ["img", imgAttrs],
        ["figcaption", {}, caption]
      ];
    }
    return ["figure", figureAttrs, ["img", imgAttrs]];
  }
};
const embedNode = {
  group: "block",
  atom: true,
  draggable: true,
  attrs: {
    id: { default: null },
    /** The original URL of the embed */
    url: { default: "" },
    /** The embed provider (youtube, twitter, etc.) */
    provider: { default: "generic" },
    /** The embed ID extracted from the URL (video ID, tweet ID, etc.) */
    embedId: { default: "" },
    /** Optional caption */
    caption: { default: "" },
    /** Width in pixels or percentage */
    width: { default: null },
    /** Height in pixels */
    height: { default: null },
    /** Aspect ratio (e.g., '16:9', '4:3') */
    aspectRatio: { default: "16:9" }
  },
  parseDOM: [
    {
      tag: "figure.openblock-embed",
      getAttrs: (dom) => {
        var _a;
        const element = dom;
        return {
          id: element.getAttribute("data-block-id"),
          url: element.getAttribute("data-url") || "",
          provider: element.getAttribute("data-provider") || "generic",
          embedId: element.getAttribute("data-embed-id") || "",
          caption: ((_a = element.querySelector("figcaption")) == null ? void 0 : _a.textContent) || "",
          width: element.getAttribute("data-width"),
          height: element.getAttribute("data-height"),
          aspectRatio: element.getAttribute("data-aspect-ratio") || "16:9"
        };
      }
    }
  ],
  toDOM: (node) => {
    const { url, provider, embedId, caption, width, height, aspectRatio } = node.attrs;
    const embedUrl = getEmbedUrl(provider, embedId, url);
    const style = width ? `max-width: ${typeof width === "number" ? `${width}px` : width}` : "";
    return [
      "figure",
      {
        class: `openblock-embed openblock-embed--${provider}`,
        "data-block-id": node.attrs.id || "",
        "data-url": url,
        "data-provider": provider,
        "data-embed-id": embedId,
        "data-aspect-ratio": aspectRatio,
        ...width ? { "data-width": String(width) } : {},
        ...height ? { "data-height": String(height) } : {},
        style
      },
      [
        "div",
        {
          class: "openblock-embed-container",
          style: `aspect-ratio: ${aspectRatio.replace(":", "/")}`
        },
        embedUrl ? [
          "iframe",
          {
            src: embedUrl,
            frameborder: "0",
            allowfullscreen: "true",
            allow: "accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture",
            loading: "lazy"
          }
        ] : [
          "div",
          { class: "openblock-embed-placeholder" },
          ["span", { class: "openblock-embed-placeholder-text" }, "Paste a URL to embed"]
        ]
      ],
      ...caption ? [["figcaption", { class: "openblock-embed-caption" }, caption]] : []
    ];
  }
};
function getEmbedUrl(provider, embedId, originalUrl) {
  if (!embedId && !originalUrl) return "";
  switch (provider) {
    case "youtube":
      return `https://www.youtube.com/embed/${embedId}`;
    case "vimeo":
      return `https://player.vimeo.com/video/${embedId}`;
    case "twitter":
      return "";
    case "codepen":
      return `https://codepen.io/${embedId}/embed/preview`;
    case "codesandbox":
      return `https://codesandbox.io/embed/${embedId}`;
    case "figma":
      return `https://www.figma.com/embed?embed_host=openblock&url=${encodeURIComponent(originalUrl)}`;
    case "loom":
      return `https://www.loom.com/embed/${embedId}`;
    case "spotify":
      return `https://open.spotify.com/embed/${embedId}`;
    case "soundcloud":
      return `https://w.soundcloud.com/player/?url=${encodeURIComponent(originalUrl)}&auto_play=false`;
    case "generic":
    default:
      return originalUrl;
  }
}
function parseEmbedUrl(url) {
  try {
    const urlObj = new URL(url);
    const hostname = urlObj.hostname.replace("www.", "");
    if (hostname === "youtube.com" || hostname === "youtu.be") {
      let videoId = "";
      if (hostname === "youtu.be") {
        videoId = urlObj.pathname.slice(1);
      } else {
        videoId = urlObj.searchParams.get("v") || "";
        if (!videoId && urlObj.pathname.startsWith("/embed/")) {
          videoId = urlObj.pathname.split("/embed/")[1];
        }
      }
      if (videoId) {
        return { provider: "youtube", embedId: videoId };
      }
    }
    if (hostname === "vimeo.com") {
      const match = urlObj.pathname.match(/\/(\d+)/);
      if (match) {
        return { provider: "vimeo", embedId: match[1] };
      }
    }
    if (hostname === "twitter.com" || hostname === "x.com") {
      const match = urlObj.pathname.match(/\/\w+\/status\/(\d+)/);
      if (match) {
        return { provider: "twitter", embedId: match[1] };
      }
    }
    if (hostname === "codepen.io") {
      const match = urlObj.pathname.match(/\/(\w+)\/pen\/(\w+)/);
      if (match) {
        return { provider: "codepen", embedId: `${match[1]}/pen/${match[2]}` };
      }
    }
    if (hostname === "codesandbox.io") {
      const match = urlObj.pathname.match(/\/s\/([^/]+)/);
      if (match) {
        return { provider: "codesandbox", embedId: match[1] };
      }
    }
    if (hostname === "figma.com") {
      if (urlObj.pathname.includes("/file/") || urlObj.pathname.includes("/proto/")) {
        return { provider: "figma", embedId: url };
      }
    }
    if (hostname === "loom.com") {
      const match = urlObj.pathname.match(/\/share\/([^/]+)/);
      if (match) {
        return { provider: "loom", embedId: match[1] };
      }
    }
    if (hostname === "open.spotify.com") {
      const match = urlObj.pathname.match(/\/(track|album|playlist|episode)\/([^/]+)/);
      if (match) {
        return { provider: "spotify", embedId: `${match[1]}/${match[2]}` };
      }
    }
    if (hostname === "soundcloud.com") {
      return { provider: "soundcloud", embedId: url };
    }
    return { provider: "generic", embedId: url };
  } catch {
    return null;
  }
}
const checkListNode = {
  group: "block",
  content: "checkListItem+",
  attrs: {
    id: { default: null }
  },
  parseDOM: [
    {
      tag: "ul.openblock-checklist",
      getAttrs: (dom) => ({
        id: dom.getAttribute("data-block-id")
      })
    }
  ],
  toDOM: (node) => [
    "ul",
    {
      class: "openblock-checklist",
      "data-block-id": node.attrs.id || ""
    },
    0
  ]
};
const checkListItemNode = {
  content: "inline*",
  attrs: {
    id: { default: null },
    checked: { default: false }
  },
  parseDOM: [
    {
      tag: "li.openblock-checklist-item",
      getAttrs: (dom) => {
        const element = dom;
        const checkbox = element.querySelector('input[type="checkbox"]');
        return {
          id: element.getAttribute("data-block-id"),
          checked: (checkbox == null ? void 0 : checkbox.checked) || element.getAttribute("data-checked") === "true"
        };
      }
    }
  ],
  toDOM: (node) => [
    "li",
    {
      class: `openblock-checklist-item ${node.attrs.checked ? "openblock-checklist-item--checked" : ""}`,
      "data-block-id": node.attrs.id || "",
      "data-checked": String(node.attrs.checked)
    },
    [
      "label",
      { class: "openblock-checklist-label", contenteditable: "false" },
      [
        "input",
        {
          type: "checkbox",
          class: "openblock-checklist-checkbox",
          ...node.attrs.checked ? { checked: "checked" } : {}
        }
      ]
    ],
    ["span", { class: "openblock-checklist-content" }, 0]
  ],
  // Allow dragging the entire item
  draggable: false
};
const boldMark = {
  parseDOM: [
    { tag: "strong" },
    { tag: "b" },
    { style: "font-weight=bold" },
    { style: "font-weight=700" }
  ],
  toDOM() {
    return ["strong", 0];
  }
};
const italicMark = {
  parseDOM: [
    { tag: "em" },
    { tag: "i" },
    { style: "font-style=italic" }
  ],
  toDOM() {
    return ["em", 0];
  }
};
const underlineMark = {
  parseDOM: [
    { tag: "u" },
    { style: "text-decoration=underline" }
  ],
  toDOM() {
    return ["u", 0];
  }
};
const strikethroughMark = {
  parseDOM: [
    { tag: "s" },
    { tag: "strike" },
    { style: "text-decoration=line-through" }
  ],
  toDOM() {
    return ["s", 0];
  }
};
const codeMark = {
  parseDOM: [{ tag: "code" }],
  toDOM() {
    return ["code", 0];
  }
};
const linkMark = {
  attrs: {
    href: {},
    title: { default: null },
    target: { default: null }
  },
  // Links don't extend to adjacent text typed at link boundaries
  inclusive: false,
  parseDOM: [
    {
      tag: "a[href]",
      getAttrs(dom) {
        return {
          href: dom.getAttribute("href"),
          title: dom.getAttribute("title"),
          target: dom.getAttribute("target")
        };
      }
    }
  ],
  toDOM(mark) {
    return [
      "a",
      {
        href: mark.attrs.href,
        title: mark.attrs.title,
        target: mark.attrs.target
      },
      0
    ];
  }
};
const textColorMark = {
  attrs: {
    color: {}
  },
  parseDOM: [
    {
      style: "color",
      getAttrs(value) {
        return { color: value };
      }
    },
    {
      tag: "span[data-text-color]",
      getAttrs(dom) {
        return { color: dom.getAttribute("data-text-color") };
      }
    }
  ],
  toDOM(mark) {
    return [
      "span",
      {
        "data-text-color": mark.attrs.color,
        style: `color: ${mark.attrs.color}`
      },
      0
    ];
  }
};
const backgroundColorMark = {
  attrs: {
    color: {}
  },
  parseDOM: [
    {
      style: "background-color",
      getAttrs(value) {
        return { color: value };
      }
    },
    {
      tag: "span[data-background-color]",
      getAttrs(dom) {
        return { color: dom.getAttribute("data-background-color") };
      }
    },
    {
      tag: "mark",
      getAttrs(dom) {
        const bgColor = dom.style.backgroundColor;
        return { color: bgColor || "#ffff00" };
      }
    }
  ],
  toDOM(mark) {
    return [
      "span",
      {
        "data-background-color": mark.attrs.color,
        style: `background-color: ${mark.attrs.color}; border-radius: 2px; padding: 0 2px;`
      },
      0
    ];
  }
};
const DEFAULT_NODES = {
  doc: docNode,
  paragraph: paragraphNode,
  heading: headingNode,
  text: textNode,
  hardBreak: hardBreakNode,
  blockquote: blockquoteNode,
  callout: calloutNode,
  codeBlock: codeBlockNode,
  divider: dividerNode,
  bulletList: bulletListNode,
  orderedList: orderedListNode,
  listItem: listItemNode,
  columnList: columnListNode,
  column: columnNode,
  table: tableNode,
  tableRow: tableRowNode,
  tableCell: tableCellNode,
  tableHeader: tableHeaderNode,
  image: imageNode,
  checkList: checkListNode,
  checkListItem: checkListItemNode,
  embed: embedNode
};
const DEFAULT_MARKS = {
  bold: boldMark,
  italic: italicMark,
  underline: underlineMark,
  strikethrough: strikethroughMark,
  code: codeMark,
  link: linkMark,
  textColor: textColorMark,
  backgroundColor: backgroundColorMark
};
function createSchema() {
  return new Schema({
    nodes: DEFAULT_NODES,
    marks: DEFAULT_MARKS
  });
}
const BLOCK_ID_PLUGIN_KEY = new PluginKey("blockIds");
function createBlockIdPlugin() {
  return new Plugin({
    key: BLOCK_ID_PLUGIN_KEY,
    appendTransaction(_transactions, _oldState, newState) {
      const nodesToUpdate = [];
      newState.doc.descendants((node, pos) => {
        if (node.type.name === "doc" || !node.isBlock) {
          if (node.type.name !== "column") {
            return;
          }
        }
        if (node.attrs.id === void 0) {
          return;
        }
        if (!node.attrs.id) {
          nodesToUpdate.push({
            pos,
            attrs: { ...node.attrs, id: v4() }
          });
        }
      });
      if (nodesToUpdate.length === 0) {
        return null;
      }
      let tr = newState.tr;
      for (let i = nodesToUpdate.length - 1; i >= 0; i--) {
        const { pos, attrs } = nodesToUpdate[i];
        tr = tr.setNodeMarkup(pos, void 0, attrs);
      }
      return tr;
    }
  });
}
function headingRule(nodeType, maxLevel = 6) {
  return textblockTypeInputRule(
    new RegExp(`^(#{1,${maxLevel}})\\s$`),
    nodeType,
    (match) => ({ level: match[1].length })
  );
}
function bulletListRule(listType) {
  return wrappingInputRule(
    /^\s*([-+*])\s$/,
    listType,
    void 0,
    // Join function: join with previous list if types match
    (_match, node) => node.type === listType
  );
}
function orderedListRule(listType) {
  return wrappingInputRule(
    /^\s*(\d+)\.\s$/,
    listType,
    (match) => ({ start: parseInt(match[1], 10) }),
    // Join with previous list if types match
    (_match, node) => node.type === listType
  );
}
function blockquoteRule(nodeType) {
  return wrappingInputRule(
    /^\s*>\s$/,
    nodeType
  );
}
function codeBlockRule(nodeType) {
  return textblockTypeInputRule(
    /^```([a-zA-Z0-9]*)\s$/,
    nodeType,
    (match) => ({ language: match[1] || "" })
  );
}
function dividerRule(nodeType) {
  return new InputRule(
    /^([-*_]){3,}\s$/,
    (state, _match, start, end) => {
      const tr = state.tr;
      tr.replaceWith(start, end, nodeType.create());
      return tr;
    }
  );
}
function boldRule(markType) {
  return new InputRule(
    /(?:\*\*|__)([^*_]+)(?:\*\*|__)$/,
    (state, match, start, end) => {
      const text = match[1];
      if (!text) return null;
      const tr = state.tr;
      tr.delete(start, end);
      tr.insertText(text, start);
      tr.addMark(start, start + text.length, markType.create());
      return tr;
    }
  );
}
function italicRule(markType) {
  return new InputRule(
    /(?:^|[^*_])(\*|_)([^*_]+)\1$/,
    (state, match, start, end) => {
      const text = match[2];
      if (!text) return null;
      const tr = state.tr;
      const actualStart = match[0].length > match[1].length + text.length + match[1].length ? start + 1 : start;
      tr.delete(actualStart, end);
      tr.insertText(text, actualStart);
      tr.addMark(actualStart, actualStart + text.length, markType.create());
      return tr;
    }
  );
}
function inlineCodeRule(markType) {
  return new InputRule(
    /`([^`]+)`$/,
    (state, match, start, end) => {
      const text = match[1];
      if (!text) return null;
      const tr = state.tr;
      tr.delete(start, end);
      tr.insertText(text, start);
      tr.addMark(start, start + text.length, markType.create());
      return tr;
    }
  );
}
function strikethroughRule(markType) {
  return new InputRule(
    /~~([^~]+)~~$/,
    (state, match, start, end) => {
      const text = match[1];
      if (!text) return null;
      const tr = state.tr;
      tr.delete(start, end);
      tr.insertText(text, start);
      tr.addMark(start, start + text.length, markType.create());
      return tr;
    }
  );
}
function createInputRulesPlugin(schema, config = {}) {
  const {
    // Block rules
    headings = true,
    bulletLists = true,
    orderedLists = true,
    blockquotes = true,
    codeBlocks = true,
    dividers = true,
    // Inline rules
    bold = true,
    italic = true,
    inlineCode = true,
    strikethrough = true
  } = config;
  const rules = [];
  if (headings && schema.nodes.heading) {
    rules.push(headingRule(schema.nodes.heading));
  }
  if (bulletLists && schema.nodes.bulletList && schema.nodes.listItem) {
    rules.push(bulletListRule(schema.nodes.bulletList));
  }
  if (orderedLists && schema.nodes.orderedList && schema.nodes.listItem) {
    rules.push(orderedListRule(schema.nodes.orderedList));
  }
  if (blockquotes && schema.nodes.blockquote) {
    rules.push(blockquoteRule(schema.nodes.blockquote));
  }
  if (codeBlocks && schema.nodes.codeBlock) {
    rules.push(codeBlockRule(schema.nodes.codeBlock));
  }
  if (dividers && schema.nodes.divider) {
    rules.push(dividerRule(schema.nodes.divider));
  }
  if (bold && schema.marks.bold) {
    rules.push(boldRule(schema.marks.bold));
  }
  if (italic && schema.marks.italic) {
    rules.push(italicRule(schema.marks.italic));
  }
  if (inlineCode && schema.marks.code) {
    rules.push(inlineCodeRule(schema.marks.code));
  }
  if (strikethrough && schema.marks.strikethrough) {
    rules.push(strikethroughRule(schema.marks.strikethrough));
  }
  return inputRules({ rules });
}
const SLASH_MENU_PLUGIN_KEY = new PluginKey("slashMenu");
function createSlashMenuPlugin(config = {}) {
  const { trigger = "/", onlyAtStart = true } = config;
  return new Plugin({
    key: SLASH_MENU_PLUGIN_KEY,
    state: {
      init() {
        return {
          active: false,
          query: "",
          triggerPos: 0,
          coords: null
        };
      },
      apply(tr, state, _oldEditorState, newEditorState) {
        const meta = tr.getMeta(SLASH_MENU_PLUGIN_KEY);
        if (meta) {
          if (meta.close) {
            return { active: false, query: "", triggerPos: 0, coords: null };
          }
          return { ...state, ...meta };
        }
        if (!state.active) {
          if (!tr.docChanged) return state;
          const { $from: $from2 } = newEditorState.selection;
          const textBefore2 = $from2.parent.textContent.slice(0, $from2.parentOffset);
          if (textBefore2.endsWith(trigger)) {
            if (onlyAtStart && textBefore2 !== trigger) {
              return state;
            }
            return {
              active: true,
              query: "",
              triggerPos: $from2.pos - trigger.length,
              coords: null
              // Will be set by the view
            };
          }
          return state;
        }
        if (!tr.docChanged && tr.selectionSet) {
          return { active: false, query: "", triggerPos: 0, coords: null };
        }
        const { $from } = newEditorState.selection;
        const textBefore = $from.parent.textContent.slice(0, $from.parentOffset);
        const triggerIndex = textBefore.lastIndexOf(trigger);
        if (triggerIndex === -1) {
          return { active: false, query: "", triggerPos: 0, coords: null };
        }
        const query = textBefore.slice(triggerIndex + trigger.length);
        if (query.includes(" ")) {
          return { active: false, query: "", triggerPos: 0, coords: null };
        }
        return { ...state, query };
      }
    },
    props: {
      handleKeyDown(view, event) {
        const state = SLASH_MENU_PLUGIN_KEY.getState(view.state);
        if (!(state == null ? void 0 : state.active)) return false;
        if (event.key === "Escape") {
          view.dispatch(view.state.tr.setMeta(SLASH_MENU_PLUGIN_KEY, { close: true }));
          return true;
        }
        return false;
      },
      handleClick(view) {
        const state = SLASH_MENU_PLUGIN_KEY.getState(view.state);
        if (state == null ? void 0 : state.active) {
          const { from } = view.state.selection;
          if (from < state.triggerPos || from > state.triggerPos + state.query.length + 1) {
            view.dispatch(view.state.tr.setMeta(SLASH_MENU_PLUGIN_KEY, { close: true }));
          }
        }
        return false;
      }
    },
    view(_editorView) {
      return {
        update(view) {
          const state = SLASH_MENU_PLUGIN_KEY.getState(view.state);
          if ((state == null ? void 0 : state.active) && !state.coords) {
            const coords = view.coordsAtPos(state.triggerPos);
            view.dispatch(
              view.state.tr.setMeta(SLASH_MENU_PLUGIN_KEY, {
                coords: { left: coords.left, top: coords.top, bottom: coords.bottom }
              })
            );
          }
        }
      };
    }
  });
}
function closeSlashMenu(view) {
  view.dispatch(view.state.tr.setMeta(SLASH_MENU_PLUGIN_KEY, { close: true }));
}
function executeSlashCommand(view, state, action) {
  const { triggerPos, query, insertAfterBlock } = state;
  if (insertAfterBlock !== void 0) {
    const node = view.state.doc.nodeAt(insertAfterBlock);
    if (node) {
      const insertPos = insertAfterBlock + node.nodeSize;
      const tr2 = view.state.tr;
      tr2.setSelection(TextSelection.create(view.state.doc, insertPos));
      tr2.setMeta(SLASH_MENU_PLUGIN_KEY, { close: true });
      view.dispatch(tr2);
      action(view, state);
      return;
    }
  }
  const tr = view.state.tr;
  const textAtTrigger = view.state.doc.textBetween(
    triggerPos,
    Math.min(triggerPos + 1, view.state.doc.content.size),
    ""
  );
  if (textAtTrigger === "/") {
    tr.delete(triggerPos, triggerPos + 1 + query.length);
  }
  tr.setMeta(SLASH_MENU_PLUGIN_KEY, { close: true });
  view.dispatch(tr);
  action(view, state);
}
function replaceCurrentBlock(view, node) {
  const { $from } = view.state.selection;
  const blockDepth = $from.depth > 0 ? 1 : 0;
  const blockStart = $from.start(blockDepth);
  const blockEnd = $from.end(blockDepth);
  const tr = view.state.tr;
  tr.replaceWith(blockStart - 1, blockEnd + 1, node);
  const newPos = blockStart;
  tr.setSelection(TextSelection.create(tr.doc, newPos));
  view.dispatch(tr);
}
function getDefaultSlashMenuItems(schema) {
  const items = [];
  if (schema.nodes.heading) {
    items.push(
      {
        id: "heading1",
        title: "Heading 1",
        description: "Large section heading",
        icon: "heading1",
        keywords: ["h1", "title", "large"],
        group: "Basic blocks",
        action: (view, _state) => {
          setBlockType(schema.nodes.heading, { level: 1 })(view.state, view.dispatch);
        }
      },
      {
        id: "heading2",
        title: "Heading 2",
        description: "Medium section heading",
        icon: "heading2",
        keywords: ["h2", "subtitle"],
        group: "Basic blocks",
        action: (view, _state) => {
          setBlockType(schema.nodes.heading, { level: 2 })(view.state, view.dispatch);
        }
      },
      {
        id: "heading3",
        title: "Heading 3",
        description: "Small section heading",
        icon: "heading3",
        keywords: ["h3"],
        group: "Basic blocks",
        action: (view, _state) => {
          setBlockType(schema.nodes.heading, { level: 3 })(view.state, view.dispatch);
        }
      }
    );
  }
  if (schema.nodes.bulletList && schema.nodes.listItem) {
    items.push({
      id: "bulletList",
      title: "Bullet List",
      description: "Create a bulleted list",
      icon: "list",
      keywords: ["ul", "unordered", "bullets"],
      group: "Lists",
      action: (view, _state) => {
        const item = schema.nodes.listItem.create(
          null,
          schema.nodes.paragraph.create()
        );
        const list = schema.nodes.bulletList.create(null, item);
        view.dispatch(view.state.tr.replaceSelectionWith(list));
      }
    });
  }
  if (schema.nodes.orderedList && schema.nodes.listItem) {
    items.push({
      id: "orderedList",
      title: "Numbered List",
      description: "Create a numbered list",
      icon: "listOrdered",
      keywords: ["ol", "ordered", "numbers"],
      group: "Lists",
      action: (view, _state) => {
        const item = schema.nodes.listItem.create(
          null,
          schema.nodes.paragraph.create()
        );
        const list = schema.nodes.orderedList.create(null, item);
        view.dispatch(view.state.tr.replaceSelectionWith(list));
      }
    });
  }
  if (schema.nodes.blockquote) {
    items.push({
      id: "quote",
      title: "Quote",
      description: "Capture a quotation",
      icon: "quote",
      keywords: ["blockquote", "citation"],
      group: "Basic blocks",
      action: (view, _state) => {
        const node = schema.nodes.blockquote.create();
        replaceCurrentBlock(view, node);
      }
    });
  }
  if (schema.nodes.callout) {
    const createCalloutAction = (calloutType) => (view, _state) => {
      const node = schema.nodes.callout.create({ calloutType });
      replaceCurrentBlock(view, node);
    };
    items.push(
      {
        id: "calloutInfo",
        title: "Callout",
        description: "Highlight important information",
        icon: "info",
        keywords: ["callout", "alert", "note", "info", "highlight"],
        group: "Basic blocks",
        action: createCalloutAction("info")
      },
      {
        id: "calloutWarning",
        title: "Warning",
        description: "Show a warning message",
        icon: "alertTriangle",
        keywords: ["callout", "alert", "warning", "caution"],
        group: "Basic blocks",
        action: createCalloutAction("warning")
      },
      {
        id: "calloutSuccess",
        title: "Success",
        description: "Show a success message",
        icon: "checkCircle",
        keywords: ["callout", "success", "done", "tip"],
        group: "Basic blocks",
        action: createCalloutAction("success")
      },
      {
        id: "calloutError",
        title: "Error",
        description: "Show an error message",
        icon: "xCircle",
        keywords: ["callout", "error", "danger", "important"],
        group: "Basic blocks",
        action: createCalloutAction("error")
      }
    );
  }
  if (schema.nodes.codeBlock) {
    items.push({
      id: "codeBlock",
      title: "Code Block",
      description: "Display code with syntax highlighting",
      icon: "code",
      keywords: ["pre", "programming", "snippet"],
      group: "Basic blocks",
      action: (view, _state) => {
        const node = schema.nodes.codeBlock.create(null, schema.text(" "));
        const tr = view.state.tr.replaceSelectionWith(node);
        view.dispatch(tr);
      }
    });
  }
  if (schema.nodes.divider) {
    items.push({
      id: "divider",
      title: "Divider",
      description: "Visual separator between sections",
      icon: "minus",
      keywords: ["hr", "horizontal", "line", "separator"],
      group: "Basic blocks",
      action: (view, _state) => {
        const { tr, schema: s } = view.state;
        const divider = schema.nodes.divider.create();
        const paragraph = s.nodes.paragraph.create();
        tr.replaceSelectionWith(divider);
        tr.insert(tr.selection.to, paragraph);
        view.dispatch(tr);
      }
    });
  }
  if (schema.nodes.columnList && schema.nodes.column) {
    items.push(
      {
        id: "columns2",
        title: "2 Columns",
        description: "Split into two columns",
        icon: "columns",
        keywords: ["col", "layout", "split", "two"],
        group: "Layout",
        action: (view, _state) => {
          const col1 = schema.nodes.column.create(
            { width: 50 },
            schema.nodes.paragraph.create()
          );
          const col2 = schema.nodes.column.create(
            { width: 50 },
            schema.nodes.paragraph.create()
          );
          const columnList = schema.nodes.columnList.create(null, [col1, col2]);
          view.dispatch(view.state.tr.replaceSelectionWith(columnList));
        }
      },
      {
        id: "columns3",
        title: "3 Columns",
        description: "Split into three columns",
        icon: "columns",
        keywords: ["col", "layout", "split", "three"],
        group: "Layout",
        action: (view, _state) => {
          const col1 = schema.nodes.column.create(
            { width: 33.33 },
            schema.nodes.paragraph.create()
          );
          const col2 = schema.nodes.column.create(
            { width: 33.33 },
            schema.nodes.paragraph.create()
          );
          const col3 = schema.nodes.column.create(
            { width: 33.34 },
            schema.nodes.paragraph.create()
          );
          const columnList = schema.nodes.columnList.create(null, [col1, col2, col3]);
          view.dispatch(view.state.tr.replaceSelectionWith(columnList));
        }
      },
      {
        id: "columnsSidebar",
        title: "Sidebar Left",
        description: "Small sidebar with main content",
        icon: "columns",
        keywords: ["col", "layout", "sidebar", "aside"],
        group: "Layout",
        action: (view, _state) => {
          const sidebar = schema.nodes.column.create(
            { width: 30 },
            schema.nodes.paragraph.create()
          );
          const main = schema.nodes.column.create(
            { width: 70 },
            schema.nodes.paragraph.create()
          );
          const columnList = schema.nodes.columnList.create(null, [sidebar, main]);
          view.dispatch(view.state.tr.replaceSelectionWith(columnList));
        }
      }
    );
  }
  if (schema.nodes.table && schema.nodes.tableRow && schema.nodes.tableCell) {
    items.push(
      {
        id: "table",
        title: "Table",
        description: "Insert a table (3x3)",
        icon: "table",
        keywords: ["grid", "spreadsheet", "rows", "columns"],
        group: "Layout",
        action: (view, _state) => {
          const createRow2 = (colCount) => {
            const cells = [];
            for (let i = 0; i < colCount; i++) {
              cells.push(
                schema.nodes.tableCell.create(null, schema.nodes.paragraph.create())
              );
            }
            return schema.nodes.tableRow.create(null, cells);
          };
          const rows = [];
          for (let i = 0; i < 3; i++) {
            rows.push(createRow2(3));
          }
          const table = schema.nodes.table.create(null, rows);
          view.dispatch(view.state.tr.replaceSelectionWith(table));
        }
      },
      {
        id: "table2x2",
        title: "Table 2x2",
        description: "Insert a small table (2x2)",
        icon: "table",
        keywords: ["grid", "small", "simple"],
        group: "Layout",
        action: (view, _state) => {
          const createRow2 = (colCount) => {
            const cells = [];
            for (let i = 0; i < colCount; i++) {
              cells.push(
                schema.nodes.tableCell.create(null, schema.nodes.paragraph.create())
              );
            }
            return schema.nodes.tableRow.create(null, cells);
          };
          const rows = [];
          for (let i = 0; i < 2; i++) {
            rows.push(createRow2(2));
          }
          const table = schema.nodes.table.create(null, rows);
          view.dispatch(view.state.tr.replaceSelectionWith(table));
        }
      },
      {
        id: "table4x4",
        title: "Table 4x4",
        description: "Insert a larger table (4x4)",
        icon: "table",
        keywords: ["grid", "large", "big"],
        group: "Layout",
        action: (view, _state) => {
          const createRow2 = (colCount) => {
            const cells = [];
            for (let i = 0; i < colCount; i++) {
              cells.push(
                schema.nodes.tableCell.create(null, schema.nodes.paragraph.create())
              );
            }
            return schema.nodes.tableRow.create(null, cells);
          };
          const rows = [];
          for (let i = 0; i < 4; i++) {
            rows.push(createRow2(4));
          }
          const table = schema.nodes.table.create(null, rows);
          view.dispatch(view.state.tr.replaceSelectionWith(table));
        }
      }
    );
  }
  if (schema.nodes.image) {
    items.push({
      id: "image",
      title: "Image",
      description: "Insert an image",
      icon: "image",
      keywords: ["img", "picture", "photo", "media"],
      group: "Media",
      action: (view) => {
        const node = schema.nodes.image.create({ src: "", alt: "" });
        view.dispatch(view.state.tr.replaceSelectionWith(node));
      }
    });
  }
  if (schema.nodes.checkList && schema.nodes.checkListItem) {
    items.push({
      id: "checklist",
      title: "To-do list",
      description: "Track tasks with checkboxes",
      icon: "checkSquare",
      keywords: ["todo", "task", "checkbox", "check", "list"],
      group: "Lists",
      action: (view) => {
        const item = schema.nodes.checkListItem.create({ checked: false });
        const list = schema.nodes.checkList.create(null, [item]);
        view.dispatch(view.state.tr.replaceSelectionWith(list));
      }
    });
  }
  if (schema.nodes.embed) {
    items.push(
      {
        id: "embed",
        title: "Embed",
        description: "Embed external content (YouTube, etc.)",
        icon: "embed",
        keywords: ["video", "youtube", "vimeo", "iframe", "external"],
        group: "Media",
        action: (view) => {
          const node = schema.nodes.embed.create({ url: "", provider: "generic" });
          view.dispatch(view.state.tr.replaceSelectionWith(node));
        }
      },
      {
        id: "youtube",
        title: "YouTube",
        description: "Embed a YouTube video",
        icon: "youtube",
        keywords: ["video", "embed", "media"],
        group: "Media",
        action: (view) => {
          const node = schema.nodes.embed.create({ url: "", provider: "youtube" });
          view.dispatch(view.state.tr.replaceSelectionWith(node));
        }
      }
    );
  }
  return items;
}
function filterSlashMenuItems(items, query) {
  if (!query) return items;
  const lowerQuery = query.toLowerCase();
  return items.filter((item) => {
    var _a;
    if (item.title.toLowerCase().includes(lowerQuery)) return true;
    if ((_a = item.keywords) == null ? void 0 : _a.some((kw) => kw.toLowerCase().includes(lowerQuery))) return true;
    return false;
  }).sort((a, b) => {
    const aExact = a.title.toLowerCase().startsWith(lowerQuery);
    const bExact = b.title.toLowerCase().startsWith(lowerQuery);
    if (aExact && !bExact) return -1;
    if (!aExact && bExact) return 1;
    return 0;
  });
}
const DRAG_DROP_PLUGIN_KEY = new PluginKey("dragDrop");
function createDragDropPlugin(config = {}) {
  const {
    showHandles = true,
    showAddButton = true,
    sideMenuClass = "ob-side-menu",
    handleClass = "ob-drag-handle",
    addButtonClass = "ob-add-button",
    onAddClick,
    onDrop
  } = config;
  let sideMenuEl = null;
  let editorView = null;
  let draggedBlockPos = null;
  function createSideMenuElement() {
    const menu = document.createElement("div");
    menu.className = sideMenuClass;
    menu.setAttribute("contenteditable", "false");
    if (showAddButton) {
      const addBtn = document.createElement("button");
      addBtn.className = addButtonClass;
      addBtn.type = "button";
      addBtn.setAttribute("tabindex", "-1");
      addBtn.setAttribute("aria-label", "Add block");
      addBtn.innerHTML = `
        <svg width="14" height="14" viewBox="0 0 14 14" fill="none" stroke="currentColor" stroke-width="1.75" stroke-linecap="round">
          <path d="M7 2v10M2 7h10"/>
        </svg>
      `;
      addBtn.addEventListener("mousedown", (e) => {
        e.preventDefault();
        e.stopPropagation();
        if (!editorView) return;
        const pos = parseInt(menu.dataset.blockPos || "0", 10);
        if (onAddClick) {
          onAddClick(pos, editorView);
        } else {
          handleAddBlockClick(pos, editorView);
        }
      });
      menu.appendChild(addBtn);
    }
    if (showHandles) {
      const handle = document.createElement("div");
      handle.className = handleClass;
      handle.setAttribute("data-drag-handle", "true");
      handle.innerHTML = `
        <svg width="14" height="14" viewBox="0 0 14 14" fill="currentColor" style="pointer-events: none;">
          <circle cx="4" cy="3" r="1.5"/>
          <circle cx="10" cy="3" r="1.5"/>
          <circle cx="4" cy="7" r="1.5"/>
          <circle cx="10" cy="7" r="1.5"/>
          <circle cx="4" cy="11" r="1.5"/>
          <circle cx="10" cy="11" r="1.5"/>
        </svg>
      `;
      handle.addEventListener("mousedown", (e) => {
        if (e.button !== 0) return;
        e.preventDefault();
        e.stopPropagation();
        if (!editorView) return;
        const pos = parseInt(menu.dataset.blockPos || "0", 10);
        draggedBlockPos = pos;
        sideMenuEl == null ? void 0 : sideMenuEl.classList.remove(`${sideMenuClass}--visible`);
        handle.classList.add(`${handleClass}--dragging`);
        document.body.style.cursor = "grabbing";
      });
      menu.appendChild(handle);
    }
    return menu;
  }
  function updateSideMenuPosition(view, blockPos) {
    if (!sideMenuEl) return;
    if (blockPos === null) {
      sideMenuEl.classList.remove(`${sideMenuClass}--visible`);
      return;
    }
    const node = view.state.doc.nodeAt(blockPos);
    if (!node) {
      sideMenuEl.classList.remove(`${sideMenuClass}--visible`);
      return;
    }
    const blockId = node.attrs.id;
    const blockEl = blockId ? view.dom.querySelector(`[data-block-id="${blockId}"]`) : null;
    const editorRect = view.dom.getBoundingClientRect();
    let top;
    let left;
    if (blockEl) {
      const blockRect = blockEl.getBoundingClientRect();
      top = blockRect.top - editorRect.top;
      const columnEl = blockEl.closest(".ob-column");
      const listEl = blockEl.closest("ul, ol");
      const isNestedInList = listEl && !listEl.classList.contains("openblock-checklist");
      const menuWidth = 40;
      if (columnEl) {
        const columnRect = columnEl.getBoundingClientRect();
        left = columnRect.left - editorRect.left + 4;
      } else if (isNestedInList) {
        left = 8;
      } else {
        left = blockRect.left - editorRect.left - menuWidth - 4;
        left = Math.max(4, left);
      }
    } else {
      const coords = view.coordsAtPos(blockPos);
      top = coords.top - editorRect.top;
      left = 8;
    }
    sideMenuEl.style.top = `${top}px`;
    sideMenuEl.style.left = `${left}px`;
    sideMenuEl.dataset.blockPos = String(blockPos);
    sideMenuEl.classList.add(`${sideMenuClass}--visible`);
  }
  const blockPosCache = /* @__PURE__ */ new Map();
  let lastDocVersion = -1;
  function updateBlockPosCache(view) {
    const docVersion = view.state.doc.content.size;
    if (docVersion === lastDocVersion) return;
    lastDocVersion = docVersion;
    blockPosCache.clear();
    view.state.doc.descendants((node, pos) => {
      if (node.attrs.id) {
        blockPosCache.set(node.attrs.id, pos);
      }
      return true;
    });
  }
  function findBlockPos(view, target) {
    const blockEl = target.closest("[data-block-id]");
    if (!blockEl) return null;
    if (blockEl.closest('table, .ob-table, [data-node-type="table"]')) {
      return null;
    }
    const blockId = blockEl.getAttribute("data-block-id");
    if (!blockId) return null;
    const cachedPos = blockPosCache.get(blockId);
    if (cachedPos !== void 0) {
      return { pos: cachedPos, id: blockId };
    }
    updateBlockPosCache(view);
    const pos = blockPosCache.get(blockId);
    if (pos === void 0) return null;
    return { pos, id: blockId };
  }
  let handleMouseMove = null;
  let handleDragOver = null;
  let handleDrop = null;
  let hideMenuTimer = null;
  function ensureSideMenu(view) {
    if (sideMenuEl) return;
    const parent = view.dom.parentElement;
    if (!parent) return;
    sideMenuEl = createSideMenuElement();
    parent.appendChild(sideMenuEl);
    const scheduleHideMenu = () => {
      if (hideMenuTimer) {
        clearTimeout(hideMenuTimer);
      }
      hideMenuTimer = setTimeout(() => {
        const currentState = DRAG_DROP_PLUGIN_KEY.getState(view.state);
        if ((currentState == null ? void 0 : currentState.hoveredBlockPos) !== null) {
          view.dispatch(
            view.state.tr.setMeta(DRAG_DROP_PLUGIN_KEY, {
              hoveredBlockPos: null,
              hoveredBlockId: null
            })
          );
        }
        hideMenuTimer = null;
      }, 150);
    };
    const cancelHideMenu = () => {
      if (hideMenuTimer) {
        clearTimeout(hideMenuTimer);
        hideMenuTimer = null;
      }
    };
    handleMouseMove = (event) => {
      const target = event.target;
      if (sideMenuEl == null ? void 0 : sideMenuEl.contains(target)) {
        cancelHideMenu();
        return;
      }
      const currentState = DRAG_DROP_PLUGIN_KEY.getState(view.state);
      if ((currentState == null ? void 0 : currentState.hoveredBlockId) && sideMenuEl) {
        const menuRect = sideMenuEl.getBoundingClientRect();
        const isInMenuZone = event.clientX < menuRect.right + 20 && // Within or to the left of menu + some margin
        event.clientY >= menuRect.top - 10 && event.clientY <= menuRect.bottom + 10;
        if (isInMenuZone) {
          cancelHideMenu();
          return;
        }
      }
      const blockInfo = findBlockPos(view, target);
      if (blockInfo) {
        cancelHideMenu();
        if ((currentState == null ? void 0 : currentState.hoveredBlockPos) !== blockInfo.pos) {
          view.dispatch(
            view.state.tr.setMeta(DRAG_DROP_PLUGIN_KEY, {
              hoveredBlockPos: blockInfo.pos,
              hoveredBlockId: blockInfo.id
            })
          );
        }
      } else {
        scheduleHideMenu();
      }
    };
    view.dom.addEventListener("mousemove", handleMouseMove);
    sideMenuEl.addEventListener("mouseenter", cancelHideMenu);
    sideMenuEl.addEventListener("mouseleave", () => {
      scheduleHideMenu();
    });
    handleDragOver = ((event) => {
      if (draggedBlockPos === null) return;
      const pos = view.posAtCoords({ left: event.clientX, top: event.clientY });
      if (!pos) return;
      const $pos = view.state.doc.resolve(pos.pos);
      let blockPos = null;
      let blockNode = null;
      for (let depth = $pos.depth; depth >= 1; depth--) {
        const node = $pos.node(depth);
        const parentNode = depth > 1 ? $pos.node(depth - 1) : view.state.doc;
        if (!node.isBlock || node.type.name === "column" || node.type.name === "columnList") {
          continue;
        }
        const parentName = parentNode.type.name;
        if (parentName === "doc" || parentName === "column" || parentName === "blockquote" || parentName === "listItem") {
          blockPos = $pos.before(depth);
          blockNode = node;
          break;
        }
      }
      if (blockPos === null || !blockNode) return;
      if (blockPos === draggedBlockPos) return;
      const draggedNode = view.state.doc.nodeAt(draggedBlockPos);
      if (!draggedNode) return;
      const blockCoords = view.coordsAtPos(blockPos);
      const blockEndCoords = view.coordsAtPos(blockPos + blockNode.nodeSize);
      const midY = (blockCoords.top + blockEndCoords.bottom) / 2;
      const isAbove = event.clientY < midY;
      let dropTarget;
      if (isAbove) {
        dropTarget = blockPos;
      } else {
        dropTarget = blockPos + blockNode.nodeSize;
      }
      const wouldMoveUp = dropTarget < draggedBlockPos;
      const wouldMoveDown = dropTarget > draggedBlockPos + draggedNode.nodeSize;
      if (wouldMoveUp || wouldMoveDown) {
        const currentState = DRAG_DROP_PLUGIN_KEY.getState(view.state);
        if ((currentState == null ? void 0 : currentState.dropTarget) !== dropTarget) {
          view.dispatch(
            view.state.tr.setMeta(DRAG_DROP_PLUGIN_KEY, {
              dropTarget
            })
          );
        }
      } else {
        const currentState = DRAG_DROP_PLUGIN_KEY.getState(view.state);
        if ((currentState == null ? void 0 : currentState.dropTarget) !== null) {
          view.dispatch(
            view.state.tr.setMeta(DRAG_DROP_PLUGIN_KEY, {
              dropTarget: null
            })
          );
        }
      }
    });
    handleDrop = (() => {
      document.body.style.cursor = "";
      if (draggedBlockPos === null) return;
      const currentState = DRAG_DROP_PLUGIN_KEY.getState(view.state);
      const dropTarget = currentState == null ? void 0 : currentState.dropTarget;
      if (dropTarget != null && dropTarget !== draggedBlockPos) {
        if (onDrop) {
          onDrop(draggedBlockPos, dropTarget);
        }
        moveBlock(view, draggedBlockPos, dropTarget);
      }
      draggedBlockPos = null;
      view.dispatch(
        view.state.tr.setMeta(DRAG_DROP_PLUGIN_KEY, {
          dragging: null,
          dropTarget: null
        })
      );
      const dragHandle = sideMenuEl == null ? void 0 : sideMenuEl.querySelector(`.${handleClass}`);
      dragHandle == null ? void 0 : dragHandle.classList.remove(`${handleClass}--dragging`);
    });
    document.addEventListener("mousemove", handleDragOver);
    document.addEventListener("mouseup", handleDrop);
  }
  return new Plugin({
    key: DRAG_DROP_PLUGIN_KEY,
    view(view) {
      editorView = view;
      ensureSideMenu(view);
      if (!sideMenuEl) {
        requestAnimationFrame(() => {
          ensureSideMenu(view);
        });
      }
      return {
        update(view2) {
          updateBlockPosCache(view2);
          ensureSideMenu(view2);
          const state = DRAG_DROP_PLUGIN_KEY.getState(view2.state);
          updateSideMenuPosition(view2, (state == null ? void 0 : state.hoveredBlockPos) ?? null);
        },
        destroy() {
          if (handleMouseMove) {
            view.dom.removeEventListener("mousemove", handleMouseMove);
          }
          if (handleDragOver) {
            document.removeEventListener("mousemove", handleDragOver);
          }
          if (handleDrop) {
            document.removeEventListener("mouseup", handleDrop);
          }
          if (hideMenuTimer) {
            clearTimeout(hideMenuTimer);
            hideMenuTimer = null;
          }
          sideMenuEl == null ? void 0 : sideMenuEl.remove();
          sideMenuEl = null;
          editorView = null;
          handleMouseMove = null;
          handleDragOver = null;
          handleDrop = null;
          draggedBlockPos = null;
          blockPosCache.clear();
          lastDocVersion = -1;
        }
      };
    },
    state: {
      init() {
        return {
          dragging: null,
          dropTarget: null,
          hoveredBlockPos: null,
          hoveredBlockId: null
        };
      },
      apply(tr, state) {
        const meta = tr.getMeta(DRAG_DROP_PLUGIN_KEY);
        if (meta) {
          return { ...state, ...meta };
        }
        return state;
      }
    },
    props: {
      decorations(state) {
        const pluginState = DRAG_DROP_PLUGIN_KEY.getState(state);
        const decorations = [];
        if ((pluginState == null ? void 0 : pluginState.dragging) != null) {
          const node = state.doc.nodeAt(pluginState.dragging);
          if (node) {
            decorations.push(
              Decoration.node(pluginState.dragging, pluginState.dragging + node.nodeSize, {
                class: "ob-block-dragging"
              })
            );
          }
        }
        if ((pluginState == null ? void 0 : pluginState.dropTarget) != null) {
          const dropIndicator = Decoration.widget(
            pluginState.dropTarget,
            () => createDropIndicator(),
            { side: -1, key: "drop-indicator" }
          );
          decorations.push(dropIndicator);
        }
        if (decorations.length === 0) return null;
        return DecorationSet.create(state.doc, decorations);
      }
    }
  });
}
function handleAddBlockClick(pos, view) {
  const { state } = view;
  const node = state.doc.nodeAt(pos);
  if (!node) return;
  const endOfBlockPos = pos + node.nodeSize - 1;
  const tr = state.tr.setSelection(TextSelection.create(state.doc, endOfBlockPos));
  view.dispatch(tr);
  view.focus();
  setTimeout(() => {
    const coords = view.coordsAtPos(pos + 1);
    view.dispatch(
      view.state.tr.setMeta(SLASH_MENU_PLUGIN_KEY, {
        active: true,
        query: "",
        triggerPos: endOfBlockPos,
        coords: { left: coords.left, top: coords.top, bottom: coords.bottom },
        // Custom flag to indicate this was opened via add button
        insertAfterBlock: pos
      })
    );
  }, 0);
}
function createDropIndicator() {
  const indicator = document.createElement("div");
  indicator.className = "ob-drop-indicator";
  return indicator;
}
function getBlockPosFromHandle(view, handle) {
  const pos = view.posAtDOM(handle, 0);
  if (pos === null) return null;
  const $pos = view.state.doc.resolve(pos);
  return $pos.before($pos.depth);
}
function moveBlock(view, fromPos, toPos) {
  const { state } = view;
  const node = state.doc.nodeAt(fromPos);
  if (!node) return;
  const tr = state.tr;
  const adjustedToPos = toPos > fromPos ? toPos - node.nodeSize : toPos;
  tr.delete(fromPos, fromPos + node.nodeSize);
  tr.insert(adjustedToPos, node);
  view.dispatch(tr);
}
const BUBBLE_MENU_PLUGIN_KEY = new PluginKey("bubbleMenu");
function getBlockType(state) {
  const { $from } = state.selection;
  for (let depth = $from.depth; depth > 0; depth--) {
    const node = $from.node(depth);
    if (node.isBlock && node.type.name !== "doc") {
      const type = node.type.name;
      if (type === "paragraph" || type === "listItem") {
        continue;
      }
      const props = {};
      if (type === "heading") {
        props.level = node.attrs.level;
      }
      return { type, props };
    }
  }
  return { type: "paragraph", props: {} };
}
function getTextAlign(state) {
  var _a;
  const { $from } = state.selection;
  for (let depth = $from.depth; depth > 0; depth--) {
    const node = $from.node(depth);
    if (node.isBlock && ((_a = node.type.spec.attrs) == null ? void 0 : _a.textAlign) !== void 0) {
      return node.attrs.textAlign || "left";
    }
  }
  return "left";
}
function getActiveMarks(state) {
  const { from, $from, to, empty } = state.selection;
  const marks = {
    bold: false,
    italic: false,
    underline: false,
    strikethrough: false,
    code: false,
    link: null,
    textColor: null,
    backgroundColor: null
  };
  if (empty) {
    const storedMarks = state.storedMarks || $from.marks();
    for (const mark of storedMarks) {
      if (mark.type.name === "bold") marks.bold = true;
      if (mark.type.name === "italic") marks.italic = true;
      if (mark.type.name === "underline") marks.underline = true;
      if (mark.type.name === "strikethrough") marks.strikethrough = true;
      if (mark.type.name === "code") marks.code = true;
      if (mark.type.name === "link") marks.link = mark.attrs.href;
      if (mark.type.name === "textColor") marks.textColor = mark.attrs.color;
      if (mark.type.name === "backgroundColor") marks.backgroundColor = mark.attrs.color;
    }
  } else {
    state.doc.nodesBetween(from, to, (node) => {
      if (node.isText && node.marks) {
        for (const mark of node.marks) {
          if (mark.type.name === "bold") marks.bold = true;
          if (mark.type.name === "italic") marks.italic = true;
          if (mark.type.name === "underline") marks.underline = true;
          if (mark.type.name === "strikethrough") marks.strikethrough = true;
          if (mark.type.name === "code") marks.code = true;
          if (mark.type.name === "link") marks.link = mark.attrs.href;
          if (mark.type.name === "textColor") marks.textColor = mark.attrs.color;
          if (mark.type.name === "backgroundColor") marks.backgroundColor = mark.attrs.color;
        }
      }
    });
  }
  return marks;
}
function shouldShowMenu(state, config) {
  var _a, _b;
  const { selection } = state;
  const { empty, from, to } = selection;
  const { minSelectionLength = 1, showOnNodeSelection = false } = config;
  if (empty) return false;
  if (to - from < minSelectionLength) return false;
  if (selection.constructor.name === "NodeSelection") {
    if (!showOnNodeSelection) {
      return false;
    }
    const node = selection.node;
    if (((_a = node == null ? void 0 : node.type) == null ? void 0 : _a.name) === "image" || ((_b = node == null ? void 0 : node.type) == null ? void 0 : _b.name) === "embed") {
      return false;
    }
  }
  const $from = selection.$from;
  if ($from.parent.type.name === "codeBlock") {
    return false;
  }
  let hasMediaNode = false;
  state.doc.nodesBetween(from, to, (node) => {
    if (node.type.name === "image" || node.type.name === "embed") {
      hasMediaNode = true;
      return false;
    }
    return true;
  });
  if (hasMediaNode) {
    return false;
  }
  return true;
}
function createBubbleMenuPlugin(config = {}) {
  return new Plugin({
    key: BUBBLE_MENU_PLUGIN_KEY,
    state: {
      init() {
        return {
          visible: false,
          from: 0,
          to: 0,
          coords: null,
          activeMarks: {
            bold: false,
            italic: false,
            underline: false,
            strikethrough: false,
            code: false,
            link: null,
            textColor: null,
            backgroundColor: null
          },
          blockType: { type: "paragraph", props: {} },
          textAlign: "left"
        };
      },
      apply(tr, state, _oldEditorState, newEditorState) {
        const meta = tr.getMeta(BUBBLE_MENU_PLUGIN_KEY);
        if (meta) {
          if (meta.hide) {
            return { ...state, visible: false, coords: null };
          }
          return { ...state, ...meta };
        }
        const shouldShow = shouldShowMenu(newEditorState, config);
        if (!shouldShow) {
          if (state.visible) {
            return { ...state, visible: false, coords: null };
          }
          return state;
        }
        const { from, to } = newEditorState.selection;
        const activeMarks = getActiveMarks(newEditorState);
        const blockType = getBlockType(newEditorState);
        const textAlign = getTextAlign(newEditorState);
        const selectionUnchanged = state.from === from && state.to === to;
        const coords = selectionUnchanged ? state.coords : null;
        return {
          visible: true,
          from,
          to,
          coords,
          activeMarks,
          blockType,
          textAlign
        };
      }
    },
    view(editorView) {
      let updateTimeout = null;
      const { showDelay = 0 } = config;
      const updateCoords = () => {
        const state = BUBBLE_MENU_PLUGIN_KEY.getState(editorView.state);
        if (!(state == null ? void 0 : state.visible)) return;
        const { from } = editorView.state.selection;
        const start = editorView.coordsAtPos(from);
        const left = start.left;
        const top = start.top;
        const bottom = start.bottom;
        if (!state.coords || state.coords.left !== left || state.coords.top !== top) {
          editorView.dispatch(
            editorView.state.tr.setMeta(BUBBLE_MENU_PLUGIN_KEY, {
              coords: { left, top, bottom }
            })
          );
        }
      };
      return {
        update(view) {
          const state = BUBBLE_MENU_PLUGIN_KEY.getState(view.state);
          if ((state == null ? void 0 : state.visible) && !state.coords) {
            if (showDelay > 0) {
              if (updateTimeout) clearTimeout(updateTimeout);
              updateTimeout = setTimeout(updateCoords, showDelay);
            } else {
              updateCoords();
            }
          }
        },
        destroy() {
          if (updateTimeout) clearTimeout(updateTimeout);
        }
      };
    },
    props: {
      handleDOMEvents: {
        // Hide on blur
        blur(view) {
          const state = BUBBLE_MENU_PLUGIN_KEY.getState(view.state);
          if (state == null ? void 0 : state.visible) {
            setTimeout(() => {
              const activeElement = document.activeElement;
              const isInBubbleMenu = activeElement == null ? void 0 : activeElement.closest(".ob-bubble-menu");
              const isInLinkPopover = activeElement == null ? void 0 : activeElement.closest(".ob-link-popover");
              const isInColorPicker = activeElement == null ? void 0 : activeElement.closest(".ob-color-picker-dropdown");
              if (!view.hasFocus() && !isInBubbleMenu && !isInLinkPopover && !isInColorPicker) {
                view.dispatch(
                  view.state.tr.setMeta(BUBBLE_MENU_PLUGIN_KEY, { hide: true })
                );
              }
            }, 100);
          }
          return false;
        }
      }
    }
  });
}
function hideBubbleMenu(view) {
  view.dispatch(view.state.tr.setMeta(BUBBLE_MENU_PLUGIN_KEY, { hide: true }));
}
function isMarkActive(state, markName) {
  if (!state) return false;
  const value = state.activeMarks[markName];
  if (markName === "link" || markName === "textColor" || markName === "backgroundColor") {
    return value !== null;
  }
  return Boolean(value);
}
const MULTI_BLOCK_SELECTION_KEY = new PluginKey(
  "multiBlockSelection"
);
function getBlockAtPos(doc, pos) {
  const $pos = doc.resolve(pos);
  for (let depth = $pos.depth; depth >= 1; depth--) {
    const node = $pos.node(depth);
    if (node.isBlock && depth === 1) {
      return {
        node,
        pos: $pos.before(depth),
        depth
      };
    }
  }
  return null;
}
function getAllBlockPositions(doc) {
  const positions = [];
  doc.forEach((_node, offset) => {
    positions.push(offset);
  });
  return positions;
}
function getBlocksInRange(doc, from, to) {
  const positions = [];
  const allBlocks = getAllBlockPositions(doc);
  const start = Math.min(from, to);
  const end = Math.max(from, to);
  for (const pos of allBlocks) {
    if (pos >= start && pos <= end) {
      positions.push(pos);
    }
  }
  return positions;
}
function createBlockSelectionDecorations(state, selectedBlocks) {
  const decorations = [];
  for (const pos of selectedBlocks) {
    const node = state.doc.nodeAt(pos);
    if (node) {
      decorations.push(
        Decoration.node(pos, pos + node.nodeSize, {
          class: "ob-block-selected"
        })
      );
    }
  }
  return DecorationSet.create(state.doc, decorations);
}
const defaultConfig = {
  marginClickSelect: true,
  marginWidth: 40
};
function createMultiBlockSelectionPlugin(config = {}) {
  const finalConfig = { ...defaultConfig, ...config };
  return new Plugin({
    key: MULTI_BLOCK_SELECTION_KEY,
    state: {
      init() {
        return {
          active: false,
          selectedBlocks: [],
          anchorPos: null
        };
      },
      apply(tr, state, _oldEditorState, newEditorState) {
        const meta = tr.getMeta(MULTI_BLOCK_SELECTION_KEY);
        if (meta) {
          return meta;
        }
        if (tr.docChanged && state.selectedBlocks.length > 0) {
          const mappedBlocks = [];
          for (const pos of state.selectedBlocks) {
            const mappedPos = tr.mapping.map(pos);
            const node = newEditorState.doc.nodeAt(mappedPos);
            if (node && node.isBlock) {
              mappedBlocks.push(mappedPos);
            }
          }
          if (mappedBlocks.length !== state.selectedBlocks.length) {
            return {
              ...state,
              selectedBlocks: mappedBlocks,
              active: mappedBlocks.length > 0
            };
          }
          return { ...state, selectedBlocks: mappedBlocks };
        }
        return state;
      }
    },
    props: {
      decorations(state) {
        const pluginState = MULTI_BLOCK_SELECTION_KEY.getState(state);
        if (!pluginState || !pluginState.active || pluginState.selectedBlocks.length === 0) {
          return DecorationSet.empty;
        }
        return createBlockSelectionDecorations(state, pluginState.selectedBlocks);
      },
      handleClick(view, _pos, event) {
        if (!event.shiftKey && !event.metaKey && !event.ctrlKey) {
          const state = MULTI_BLOCK_SELECTION_KEY.getState(view.state);
          if (state == null ? void 0 : state.active) {
            view.dispatch(
              view.state.tr.setMeta(MULTI_BLOCK_SELECTION_KEY, {
                active: false,
                selectedBlocks: [],
                anchorPos: null
              })
            );
          }
        }
        return false;
      },
      handleKeyDown(view, event) {
        const state = MULTI_BLOCK_SELECTION_KEY.getState(view.state);
        if (event.key === "Escape" && (state == null ? void 0 : state.active)) {
          view.dispatch(
            view.state.tr.setMeta(MULTI_BLOCK_SELECTION_KEY, {
              active: false,
              selectedBlocks: [],
              anchorPos: null
            })
          );
          return true;
        }
        if ((event.metaKey || event.ctrlKey) && event.key === "a") {
          const { from, to } = view.state.selection;
          const docSize = view.state.doc.content.size;
          if (from === 0 && to === docSize) {
            const allBlocks = getAllBlockPositions(view.state.doc);
            view.dispatch(
              view.state.tr.setMeta(MULTI_BLOCK_SELECTION_KEY, {
                active: true,
                selectedBlocks: allBlocks,
                anchorPos: allBlocks[0] ?? null
              })
            );
            return true;
          }
        }
        if ((event.key === "Delete" || event.key === "Backspace") && (state == null ? void 0 : state.active) && state.selectedBlocks.length > 0) {
          deleteSelectedBlocks(view);
          return true;
        }
        return false;
      },
      handleDOMEvents: {
        mousedown(view, event) {
          if (!finalConfig.marginClickSelect) return false;
          const editorRect = view.dom.getBoundingClientRect();
          const clickX = event.clientX - editorRect.left;
          if (clickX > finalConfig.marginWidth) return false;
          const coords = { left: event.clientX, top: event.clientY };
          const posInfo = view.posAtCoords(coords);
          if (!posInfo) return false;
          const block = getBlockAtPos(view.state.doc, posInfo.pos);
          if (!block) return false;
          const state = MULTI_BLOCK_SELECTION_KEY.getState(view.state);
          if (event.shiftKey && state && state.anchorPos !== null) {
            const blocks = getBlocksInRange(view.state.doc, state.anchorPos, block.pos);
            view.dispatch(
              view.state.tr.setMeta(MULTI_BLOCK_SELECTION_KEY, {
                active: true,
                selectedBlocks: blocks,
                anchorPos: state.anchorPos
              })
            );
          } else if (event.metaKey || event.ctrlKey) {
            const currentBlocks = (state == null ? void 0 : state.selectedBlocks) || [];
            const isSelected = currentBlocks.includes(block.pos);
            const newBlocks = isSelected ? currentBlocks.filter((p) => p !== block.pos) : [...currentBlocks, block.pos];
            view.dispatch(
              view.state.tr.setMeta(MULTI_BLOCK_SELECTION_KEY, {
                active: newBlocks.length > 0,
                selectedBlocks: newBlocks,
                anchorPos: block.pos
              })
            );
          } else {
            view.dispatch(
              view.state.tr.setMeta(MULTI_BLOCK_SELECTION_KEY, {
                active: true,
                selectedBlocks: [block.pos],
                anchorPos: block.pos
              })
            );
          }
          event.preventDefault();
          return true;
        }
      }
    }
  });
}
function deleteSelectedBlocks(view) {
  const state = MULTI_BLOCK_SELECTION_KEY.getState(view.state);
  if (!state || state.selectedBlocks.length === 0) return;
  const positions = [...state.selectedBlocks].sort((a, b) => b - a);
  let tr = view.state.tr;
  for (const pos of positions) {
    const node = tr.doc.nodeAt(pos);
    if (node) {
      tr = tr.delete(pos, pos + node.nodeSize);
    }
  }
  tr = tr.setMeta(MULTI_BLOCK_SELECTION_KEY, {
    active: false,
    selectedBlocks: [],
    anchorPos: null
  });
  view.dispatch(tr);
}
function findTableContext(state) {
  const { $from } = state.selection;
  let cell = null;
  let cellPos = -1;
  let row = null;
  let rowPos = -1;
  let table = null;
  let tablePos = -1;
  for (let depth = $from.depth; depth > 0; depth--) {
    const node = $from.node(depth);
    const pos = $from.before(depth);
    if (node.type.name === "tableCell" || node.type.name === "tableHeader") {
      cell = node;
      cellPos = pos;
    } else if (node.type.name === "tableRow") {
      row = node;
      rowPos = pos;
    } else if (node.type.name === "table") {
      table = node;
      tablePos = pos;
      break;
    }
  }
  if (!table || !row || !cell) return null;
  let rowIndex = 0;
  let cellIndex = 0;
  table.forEach((_r, offset) => {
    if (tablePos + 1 + offset < rowPos) {
      rowIndex++;
    }
  });
  row.forEach((_c, offset) => {
    if (rowPos + 1 + offset < cellPos) {
      cellIndex++;
    }
  });
  return {
    table,
    tablePos,
    row,
    rowPos,
    rowIndex,
    cell,
    cellPos,
    cellIndex,
    colCount: row.childCount,
    rowCount: table.childCount
  };
}
function createRow(state, colCount, isHeader = false) {
  const { schema } = state;
  const cellType = isHeader && schema.nodes.tableHeader ? schema.nodes.tableHeader : schema.nodes.tableCell;
  const cells = [];
  for (let i = 0; i < colCount; i++) {
    cells.push(cellType.create(null, schema.nodes.paragraph.create()));
  }
  return schema.nodes.tableRow.create(null, cells);
}
const addRowAfter = (state, dispatch) => {
  const ctx = findTableContext(state);
  if (!ctx) return false;
  if (dispatch) {
    const { tablePos, rowIndex, colCount, table } = ctx;
    const newRow = createRow(state, colCount);
    let insertPos = tablePos + 1;
    for (let i = 0; i <= rowIndex; i++) {
      insertPos += table.child(i).nodeSize;
    }
    dispatch(state.tr.insert(insertPos, newRow));
  }
  return true;
};
const addRowBefore = (state, dispatch) => {
  const ctx = findTableContext(state);
  if (!ctx) return false;
  if (dispatch) {
    const { tablePos, rowIndex, colCount, table } = ctx;
    const newRow = createRow(state, colCount);
    let insertPos = tablePos + 1;
    for (let i = 0; i < rowIndex; i++) {
      insertPos += table.child(i).nodeSize;
    }
    dispatch(state.tr.insert(insertPos, newRow));
  }
  return true;
};
const deleteRow = (state, dispatch) => {
  const ctx = findTableContext(state);
  if (!ctx) return false;
  if (ctx.rowCount <= 1) return false;
  if (dispatch) {
    const { tablePos, rowIndex, table } = ctx;
    let rowStart = tablePos + 1;
    for (let i = 0; i < rowIndex; i++) {
      rowStart += table.child(i).nodeSize;
    }
    const rowEnd = rowStart + table.child(rowIndex).nodeSize;
    dispatch(state.tr.delete(rowStart, rowEnd));
  }
  return true;
};
const addColumnAfter = (state, dispatch) => {
  const ctx = findTableContext(state);
  if (!ctx) return false;
  if (dispatch) {
    const { schema } = state;
    const { tablePos, cellIndex, table } = ctx;
    const newRows = [];
    table.forEach((row) => {
      const cells = [];
      const isHeaderRow = row.child(0).type.name === "tableHeader";
      const cellType = isHeaderRow && schema.nodes.tableHeader ? schema.nodes.tableHeader : schema.nodes.tableCell;
      row.forEach((cell, _offset, index) => {
        cells.push(cell);
        if (index === cellIndex) {
          cells.push(cellType.create(null, schema.nodes.paragraph.create()));
        }
      });
      newRows.push(schema.nodes.tableRow.create(row.attrs, cells));
    });
    const newTable = schema.nodes.table.create(table.attrs, newRows);
    const tr = state.tr.replaceWith(tablePos, tablePos + table.nodeSize, newTable);
    dispatch(tr);
  }
  return true;
};
const addColumnBefore = (state, dispatch) => {
  const ctx = findTableContext(state);
  if (!ctx) return false;
  if (dispatch) {
    const { schema } = state;
    const { tablePos, cellIndex, table } = ctx;
    const newRows = [];
    table.forEach((row) => {
      const cells = [];
      const isHeaderRow = row.child(0).type.name === "tableHeader";
      const cellType = isHeaderRow && schema.nodes.tableHeader ? schema.nodes.tableHeader : schema.nodes.tableCell;
      row.forEach((cell, _offset, index) => {
        if (index === cellIndex) {
          cells.push(cellType.create(null, schema.nodes.paragraph.create()));
        }
        cells.push(cell);
      });
      newRows.push(schema.nodes.tableRow.create(row.attrs, cells));
    });
    const newTable = schema.nodes.table.create(table.attrs, newRows);
    const tr = state.tr.replaceWith(tablePos, tablePos + table.nodeSize, newTable);
    dispatch(tr);
  }
  return true;
};
const deleteColumn = (state, dispatch) => {
  const ctx = findTableContext(state);
  if (!ctx) return false;
  if (ctx.colCount <= 1) return false;
  if (dispatch) {
    const { schema } = state;
    const { tablePos, cellIndex, table } = ctx;
    const newRows = [];
    table.forEach((row) => {
      const cells = [];
      row.forEach((cell, _offset, index) => {
        if (index !== cellIndex) {
          cells.push(cell);
        }
      });
      newRows.push(schema.nodes.tableRow.create(row.attrs, cells));
    });
    const newTable = schema.nodes.table.create(table.attrs, newRows);
    const tr = state.tr.replaceWith(tablePos, tablePos + table.nodeSize, newTable);
    dispatch(tr);
  }
  return true;
};
const deleteTable = (state, dispatch) => {
  const ctx = findTableContext(state);
  if (!ctx) return false;
  if (dispatch) {
    const { tablePos, table } = ctx;
    dispatch(state.tr.delete(tablePos, tablePos + table.nodeSize));
  }
  return true;
};
const goToNextCell = (state, dispatch) => {
  const ctx = findTableContext(state);
  if (!ctx) return false;
  const { tablePos, rowIndex, cellIndex, colCount, rowCount, table } = ctx;
  const isLastCell = cellIndex === colCount - 1;
  const isLastRow = rowIndex === rowCount - 1;
  if (dispatch) {
    let tr = state.tr;
    if (isLastCell && isLastRow) {
      const newRow = createRow(state, colCount);
      const insertPos = tablePos + table.nodeSize - 1;
      tr = tr.insert(insertPos, newRow);
      tr = tr.setSelection(TextSelection.near(tr.doc.resolve(insertPos + 3)));
    } else {
      let nextRow = rowIndex;
      let nextCell = cellIndex + 1;
      if (isLastCell) {
        nextRow++;
        nextCell = 0;
      }
      let targetPos = tablePos + 1;
      for (let r = 0; r < nextRow; r++) {
        targetPos += table.child(r).nodeSize;
      }
      targetPos += 1;
      const targetRow = table.child(nextRow);
      for (let c = 0; c < nextCell; c++) {
        targetPos += targetRow.child(c).nodeSize;
      }
      targetPos += 2;
      tr = tr.setSelection(TextSelection.near(tr.doc.resolve(targetPos)));
    }
    dispatch(tr);
  }
  return true;
};
const goToPreviousCell = (state, dispatch) => {
  const ctx = findTableContext(state);
  if (!ctx) return false;
  const { tablePos, rowIndex, cellIndex, colCount, table } = ctx;
  if (rowIndex === 0 && cellIndex === 0) return false;
  if (dispatch) {
    let prevRow = rowIndex;
    let prevCell = cellIndex - 1;
    if (cellIndex === 0) {
      prevRow--;
      prevCell = colCount - 1;
    }
    let targetPos = tablePos + 1;
    for (let r = 0; r < prevRow; r++) {
      targetPos += table.child(r).nodeSize;
    }
    targetPos += 1;
    const targetRow = table.child(prevRow);
    for (let c = 0; c < prevCell; c++) {
      targetPos += targetRow.child(c).nodeSize;
    }
    targetPos += 2;
    const tr = state.tr.setSelection(TextSelection.near(state.doc.resolve(targetPos)));
    dispatch(tr);
  }
  return true;
};
function isInTable(state) {
  return findTableContext(state) !== null;
}
function getTableInfo(state) {
  const ctx = findTableContext(state);
  if (!ctx) return null;
  return {
    rowIndex: ctx.rowIndex,
    cellIndex: ctx.cellIndex,
    rowCount: ctx.rowCount,
    colCount: ctx.colCount
  };
}
function addRowAtIndex(state, dispatch, tablePos, rowIndex) {
  const table = state.doc.nodeAt(tablePos);
  if (!table || table.type.name !== "table") return false;
  if (dispatch) {
    const colCount = table.child(0).childCount;
    const { schema } = state;
    const cells = [];
    for (let i = 0; i < colCount; i++) {
      cells.push(schema.nodes.tableCell.create(null, schema.nodes.paragraph.create()));
    }
    const newRow = schema.nodes.tableRow.create(null, cells);
    let insertPos = tablePos + 1;
    for (let i = 0; i < rowIndex; i++) {
      insertPos += table.child(i).nodeSize;
    }
    dispatch(state.tr.insert(insertPos, newRow));
  }
  return true;
}
function addColumnAtIndex(state, dispatch, tablePos, colIndex) {
  const table = state.doc.nodeAt(tablePos);
  if (!table || table.type.name !== "table") return false;
  if (dispatch) {
    const { schema } = state;
    const newRows = [];
    table.forEach((row) => {
      const cells = [];
      const firstCell = row.childCount > 0 ? row.child(0) : null;
      const isHeaderRow = (firstCell == null ? void 0 : firstCell.type.name) === "tableHeader";
      const cellType = isHeaderRow && schema.nodes.tableHeader ? schema.nodes.tableHeader : schema.nodes.tableCell;
      const createNewCell = () => {
        const newCell = cellType.createAndFill();
        if (newCell) return newCell;
        return cellType.create(null, schema.nodes.paragraph.create());
      };
      row.forEach((cell, _offset, index) => {
        if (index === colIndex) {
          cells.push(createNewCell());
        }
        cells.push(cell);
      });
      if (colIndex >= row.childCount) {
        cells.push(createNewCell());
      }
      newRows.push(schema.nodes.tableRow.create(row.attrs, cells));
    });
    const newTable = schema.nodes.table.create(table.attrs, newRows);
    dispatch(state.tr.replaceWith(tablePos, tablePos + table.nodeSize, newTable));
  }
  return true;
}
function deleteRowAtIndex(state, dispatch, tablePos, rowIndex) {
  const table = state.doc.nodeAt(tablePos);
  if (!table || table.type.name !== "table") return false;
  if (table.childCount <= 1) return false;
  if (dispatch) {
    let rowStart = tablePos + 1;
    for (let i = 0; i < rowIndex; i++) {
      rowStart += table.child(i).nodeSize;
    }
    const rowEnd = rowStart + table.child(rowIndex).nodeSize;
    dispatch(state.tr.delete(rowStart, rowEnd));
  }
  return true;
}
function deleteColumnAtIndex(state, dispatch, tablePos, colIndex) {
  const table = state.doc.nodeAt(tablePos);
  if (!table || table.type.name !== "table") return false;
  if (table.child(0).childCount <= 1) return false;
  if (dispatch) {
    const { schema } = state;
    const newRows = [];
    table.forEach((row) => {
      const cells = [];
      row.forEach((cell, _offset, index) => {
        if (index !== colIndex) {
          cells.push(cell);
        }
      });
      newRows.push(schema.nodes.tableRow.create(row.attrs, cells));
    });
    const newTable = schema.nodes.table.create(table.attrs, newRows);
    dispatch(state.tr.replaceWith(tablePos, tablePos + table.nodeSize, newTable));
  }
  return true;
}
const TABLE_PLUGIN_KEY = new PluginKey("tablePlugin");
function createTablePlugin(config = {}) {
  const { tabNavigation = true } = config;
  const keys = {};
  if (tabNavigation) {
    keys["Tab"] = goToNextCell;
    keys["Shift-Tab"] = goToPreviousCell;
  }
  keys["Mod-Alt-ArrowDown"] = addRowAfter;
  keys["Mod-Alt-ArrowUp"] = addRowBefore;
  keys["Mod-Alt-ArrowRight"] = addColumnAfter;
  keys["Mod-Alt-ArrowLeft"] = addColumnBefore;
  keys["Mod-Shift-Backspace"] = deleteRow;
  return keymap(keys);
}
const DEFAULT_KEYBOARD_SHORTCUTS = [
  // Text formatting
  { key: "Mod-b", description: "Bold", action: "bold" },
  { key: "Mod-i", description: "Italic", action: "italic" },
  { key: "Mod-u", description: "Underline", action: "underline" },
  { key: "Mod-Shift-s", description: "Strikethrough", action: "strikethrough" },
  { key: "Mod-e", description: "Inline code", action: "code" },
  // Undo/redo
  { key: "Mod-z", description: "Undo", action: "undo" },
  { key: "Mod-y", description: "Redo", action: "redo" },
  { key: "Mod-Shift-z", description: "Redo", action: "redo" },
  // Block types
  { key: "Mod-Alt-0", description: "Paragraph", action: "paragraph" },
  { key: "Mod-Alt-1", description: "Heading 1", action: "heading1" },
  { key: "Mod-Alt-2", description: "Heading 2", action: "heading2" },
  { key: "Mod-Alt-3", description: "Heading 3", action: "heading3" },
  { key: "Mod-Shift-7", description: "Ordered list", action: "orderedList" },
  { key: "Mod-Shift-8", description: "Bullet list", action: "bulletList" },
  { key: "Mod-Shift-9", description: "Checklist", action: "checkList" },
  { key: "Mod-Shift-b", description: "Blockquote", action: "blockquote" },
  { key: "Mod-Shift-c", description: "Code block", action: "codeBlock" },
  // List operations
  { key: "Tab", description: "Indent list item", action: "indentListItem" },
  { key: "Shift-Tab", description: "Outdent list item", action: "outdentListItem" }
];
function createKeyboardShortcutsPlugin(schema, config = {}) {
  const { shortcuts = {}, includeDefaults = true, disabledShortcuts = [] } = config;
  const keymapObj = {};
  if (includeDefaults) {
    for (const shortcut of DEFAULT_KEYBOARD_SHORTCUTS) {
      if (disabledShortcuts.includes(shortcut.key)) continue;
      const command = resolveAction(schema, shortcut.action);
      if (command) {
        keymapObj[shortcut.key] = command;
      }
    }
  }
  for (const [key, action] of Object.entries(shortcuts)) {
    if (typeof action === "function") {
      keymapObj[key] = action;
    } else {
      const command = resolveAction(schema, action);
      if (command) {
        keymapObj[key] = command;
      }
    }
  }
  return keymap(keymapObj);
}
function resolveAction(schema, action) {
  if (typeof action === "function") {
    return action;
  }
  const markActions = {
    bold: "bold",
    italic: "italic",
    underline: "underline",
    strikethrough: "strikethrough",
    code: "code"
  };
  if (markActions[action] && schema.marks[markActions[action]]) {
    return toggleMark(schema.marks[markActions[action]]);
  }
  switch (action) {
    case "undo":
      return undo;
    case "redo":
      return redo;
    case "paragraph":
      if (schema.nodes.paragraph) {
        return setBlockType(schema.nodes.paragraph);
      }
      break;
    case "heading1":
      if (schema.nodes.heading) {
        return setBlockType(schema.nodes.heading, { level: 1 });
      }
      break;
    case "heading2":
      if (schema.nodes.heading) {
        return setBlockType(schema.nodes.heading, { level: 2 });
      }
      break;
    case "heading3":
      if (schema.nodes.heading) {
        return setBlockType(schema.nodes.heading, { level: 3 });
      }
      break;
    case "bulletList":
      if (schema.nodes.bulletList && schema.nodes.listItem) {
        return wrapInList(schema.nodes.bulletList);
      }
      break;
    case "orderedList":
      if (schema.nodes.orderedList && schema.nodes.listItem) {
        return wrapInList(schema.nodes.orderedList);
      }
      break;
    case "checkList":
      if (schema.nodes.checkList && schema.nodes.checkListItem) {
        return (state, dispatch) => {
          const { $from, $to } = state.selection;
          const range = $from.blockRange($to);
          if (!range) return false;
          const tr = state.tr;
          const checkListItem = schema.nodes.checkListItem.create({ checked: false }, $from.parent.content);
          const checkList = schema.nodes.checkList.create(null, [checkListItem]);
          tr.replaceRangeWith(range.start, range.end, checkList);
          if (dispatch) dispatch(tr);
          return true;
        };
      }
      break;
    case "blockquote":
      if (schema.nodes.blockquote) {
        return wrapIn(schema.nodes.blockquote);
      }
      break;
    case "codeBlock":
      if (schema.nodes.codeBlock) {
        return setBlockType(schema.nodes.codeBlock);
      }
      break;
    case "indentListItem":
      if (schema.nodes.listItem) {
        return sinkListItem(schema.nodes.listItem);
      }
      break;
    case "outdentListItem":
      if (schema.nodes.listItem) {
        return liftListItem(schema.nodes.listItem);
      }
      break;
    case "lift":
      return lift;
  }
  return null;
}
function createChecklistPlugin(config = {}) {
  return new Plugin({
    props: {
      handleKeyDown(view, event) {
        const { state } = view;
        const { selection } = state;
        const { $from } = selection;
        let checkListItemDepth = null;
        for (let depth = $from.depth; depth >= 0; depth--) {
          if ($from.node(depth).type.name === "checkListItem") {
            checkListItemDepth = depth;
            break;
          }
        }
        if (checkListItemDepth === null) return false;
        const checkListItem = $from.node(checkListItemDepth);
        const checkListItemPos = $from.before(checkListItemDepth);
        if (event.key === "Enter" && event.shiftKey) {
          const hardBreak = state.schema.nodes.hardBreak;
          if (hardBreak) {
            const tr = state.tr.replaceSelectionWith(hardBreak.create());
            view.dispatch(tr);
            return true;
          }
        }
        if (event.key === "Enter" && !event.shiftKey) {
          const isEmpty = checkListItem.textContent.trim() === "";
          if (isEmpty) {
            const checkListDepth = checkListItemDepth - 1;
            if (checkListDepth >= 0 && $from.node(checkListDepth).type.name === "checkList") {
              const checkList = $from.node(checkListDepth);
              const checkListPos = $from.before(checkListDepth);
              if (checkList.childCount === 1) {
                const paragraph2 = state.schema.nodes.paragraph.create();
                const tr2 = state.tr.replaceWith(
                  checkListPos,
                  checkListPos + checkList.nodeSize,
                  paragraph2
                );
                view.dispatch(tr2.scrollIntoView());
                return true;
              }
              const tr = state.tr;
              tr.delete(checkListItemPos, checkListItemPos + checkListItem.nodeSize);
              const newCheckListPos = tr.mapping.map(checkListPos);
              const $newPos = tr.doc.resolve(newCheckListPos);
              let newCheckListEndPos = newCheckListPos;
              for (let d = $newPos.depth; d >= 0; d--) {
                if ($newPos.node(d).type.name === "checkList") {
                  newCheckListEndPos = $newPos.after(d);
                  break;
                }
              }
              const paragraph = state.schema.nodes.paragraph.create();
              tr.insert(newCheckListEndPos, paragraph);
              const $insertPos = tr.doc.resolve(newCheckListEndPos + 1);
              tr.setSelection(TextSelection.near($insertPos));
              view.dispatch(tr.scrollIntoView());
              return true;
            }
          }
        }
        return false;
      },
      handleDOMEvents: {
        click(view, event) {
          var _a;
          const mouseEvent = event;
          const target = mouseEvent.target;
          if (target.tagName === "INPUT" && target.getAttribute("type") === "checkbox" && target.classList.contains("openblock-checklist-checkbox")) {
            const listItem = target.closest(".openblock-checklist-item");
            if (!listItem) return false;
            const pos = view.posAtDOM(listItem, 0);
            if (pos === null || pos === void 0) return false;
            const $pos = view.state.doc.resolve(pos);
            let checkListItemPos = null;
            let checkListItemNode2 = null;
            for (let depth = $pos.depth; depth >= 0; depth--) {
              const node = $pos.node(depth);
              if (node.type.name === "checkListItem") {
                checkListItemPos = $pos.before(depth);
                checkListItemNode2 = node;
                break;
              }
            }
            if (checkListItemPos === null || !checkListItemNode2) return false;
            const newChecked = !checkListItemNode2.attrs.checked;
            const tr = view.state.tr.setNodeMarkup(checkListItemPos, void 0, {
              ...checkListItemNode2.attrs,
              checked: newChecked
            });
            view.dispatch(tr);
            (_a = config.onToggle) == null ? void 0 : _a.call(config, newChecked, checkListItemNode2.attrs.id);
            event.preventDefault();
            return true;
          }
          return false;
        }
      }
    }
  });
}
const MEDIA_MENU_PLUGIN_KEY = new PluginKey("mediaMenu");
function isMediaNode(node) {
  if (node.type.name === "image") return "image";
  if (node.type.name === "embed") return "embed";
  return null;
}
function getMediaSelection(state) {
  const { selection } = state;
  if (selection instanceof NodeSelection) {
    const mediaType = isMediaNode(selection.node);
    if (mediaType) {
      return {
        mediaType,
        pos: selection.from,
        node: selection.node
      };
    }
  }
  return null;
}
function createMediaMenuPlugin() {
  return new Plugin({
    key: MEDIA_MENU_PLUGIN_KEY,
    state: {
      init() {
        return {
          visible: false,
          mediaType: null,
          nodePos: null,
          node: null,
          attrs: null,
          coords: null
        };
      },
      apply(tr, state, _oldEditorState, newEditorState) {
        const meta = tr.getMeta(MEDIA_MENU_PLUGIN_KEY);
        if (meta) {
          if (meta.hide) {
            return { ...state, visible: false, coords: null, node: null, attrs: null };
          }
          return { ...state, ...meta };
        }
        const mediaSelection = getMediaSelection(newEditorState);
        if (!mediaSelection) {
          if (state.visible) {
            return {
              visible: false,
              mediaType: null,
              nodePos: null,
              node: null,
              attrs: null,
              coords: null
            };
          }
          return state;
        }
        const { mediaType, pos, node } = mediaSelection;
        const selectionUnchanged = state.nodePos === pos && state.visible;
        const coords = selectionUnchanged ? state.coords : null;
        return {
          visible: true,
          mediaType,
          nodePos: pos,
          node,
          attrs: node.attrs,
          coords
        };
      }
    },
    view(editorView) {
      const updateCoords = () => {
        const state = MEDIA_MENU_PLUGIN_KEY.getState(editorView.state);
        if (!(state == null ? void 0 : state.visible) || state.nodePos === null) return;
        const domNode = editorView.nodeDOM(state.nodePos);
        if (!domNode || !(domNode instanceof HTMLElement)) return;
        const rect = domNode.getBoundingClientRect();
        const coords = {
          left: rect.left,
          top: rect.top,
          right: rect.right,
          bottom: rect.bottom
        };
        if (!state.coords || state.coords.left !== coords.left || state.coords.top !== coords.top) {
          editorView.dispatch(
            editorView.state.tr.setMeta(MEDIA_MENU_PLUGIN_KEY, { coords })
          );
        }
      };
      return {
        update(view) {
          const state = MEDIA_MENU_PLUGIN_KEY.getState(view.state);
          if ((state == null ? void 0 : state.visible) && !state.coords) {
            requestAnimationFrame(updateCoords);
          }
        },
        destroy() {
        }
      };
    },
    props: {
      handleDOMEvents: {
        // Hide on blur (but allow clicks on menu)
        blur(view) {
          const state = MEDIA_MENU_PLUGIN_KEY.getState(view.state);
          if (state == null ? void 0 : state.visible) {
            setTimeout(() => {
              const activeElement = document.activeElement;
              const isInMediaMenu = activeElement == null ? void 0 : activeElement.closest(".ob-media-menu");
              const isInMediaPopover = activeElement == null ? void 0 : activeElement.closest(".ob-media-url-popover");
              if (!view.hasFocus() && !isInMediaMenu && !isInMediaPopover) {
                view.dispatch(
                  view.state.tr.setMeta(MEDIA_MENU_PLUGIN_KEY, { hide: true })
                );
              }
            }, 150);
          }
          return false;
        }
      }
    }
  });
}
function hideMediaMenu(view) {
  view.dispatch(view.state.tr.setMeta(MEDIA_MENU_PLUGIN_KEY, { hide: true }));
}
function updateMediaAttrs(view, pos, attrs) {
  const node = view.state.doc.nodeAt(pos);
  if (!node) return;
  const tr = view.state.tr.setNodeMarkup(pos, void 0, {
    ...node.attrs,
    ...attrs
  });
  view.dispatch(tr);
}
function deleteMediaNode(view, pos) {
  const node = view.state.doc.nodeAt(pos);
  if (!node) return;
  const tr = view.state.tr.delete(pos, pos + node.nodeSize);
  view.dispatch(tr);
}
function createPlugins(options = {}) {
  const { schema, toggleMark: toggleMark2, inputRules: inputRules2, dragDrop, slashMenu, bubbleMenu, multiBlockSelection, table, keyboardShortcuts, checklist, mediaMenu, additionalPlugins = [] } = options;
  const slashMenuPlugin = slashMenu !== false ? createSlashMenuPlugin(typeof slashMenu === "object" ? slashMenu : {}) : null;
  const plugins = [
    // History for undo/redo
    history(),
    // Slash menu plugin must come before baseKeymap to handle Enter when menu is active
    ...slashMenuPlugin ? [slashMenuPlugin] : [],
    // Checklist plugin must come before baseKeymap to handle Enter/Shift+Enter in checklists
    ...checklist !== false ? [createChecklistPlugin(typeof checklist === "object" ? checklist : {})] : [],
    // Standard editing commands
    keymap(baseKeymap),
    // Drop cursor visual feedback
    dropCursor(),
    // Gap cursor for block boundaries
    gapCursor(),
    // Automatic block ID assignment
    createBlockIdPlugin()
  ];
  if (schema && keyboardShortcuts !== false) {
    const keyboardConfig = typeof keyboardShortcuts === "object" ? keyboardShortcuts : {};
    plugins.splice(2, 0, createKeyboardShortcutsPlugin(schema, keyboardConfig));
  } else if (toggleMark2) {
    plugins.splice(2, 0, keymap({
      "Mod-b": () => toggleMark2("bold"),
      "Mod-i": () => toggleMark2("italic"),
      "Mod-u": () => toggleMark2("underline")
    }));
    plugins.splice(2, 0, keymap({
      "Mod-z": undo,
      "Mod-y": redo,
      "Mod-Shift-z": redo
    }));
  }
  if (schema && inputRules2 !== false) {
    const rulesConfig = typeof inputRules2 === "object" ? inputRules2 : {};
    plugins.push(createInputRulesPlugin(schema, rulesConfig));
  }
  if (dragDrop !== false) {
    const dragDropConfig = typeof dragDrop === "object" ? dragDrop : {};
    plugins.push(createDragDropPlugin(dragDropConfig));
  }
  if (bubbleMenu !== false) {
    const bubbleMenuConfig = typeof bubbleMenu === "object" ? bubbleMenu : {};
    plugins.push(createBubbleMenuPlugin(bubbleMenuConfig));
  }
  if (multiBlockSelection !== false) {
    const multiBlockConfig = typeof multiBlockSelection === "object" ? multiBlockSelection : {};
    plugins.push(createMultiBlockSelectionPlugin(multiBlockConfig));
  }
  if (table !== false) {
    const tableConfig = typeof table === "object" ? table : {};
    plugins.push(createTablePlugin(tableConfig));
  }
  if (mediaMenu !== false) {
    plugins.push(createMediaMenuPlugin());
  }
  plugins.push(...additionalPlugins);
  return plugins;
}
const STYLE_ID = "openblock-styles";
let stylesInjected = false;
const EDITOR_STYLES = `
/* OpenBlock Editor Styles - Auto-injected */

.openblock-container,
.openblock-editor {
  --ob-foreground: var(--foreground, 25 5% 22%);
  --ob-background: var(--background, 0 0% 100%);
  --ob-muted: var(--muted, 40 6% 96%);
  --ob-muted-foreground: var(--muted-foreground, 25 2% 57%);
  --ob-border: var(--border, 40 6% 90%);
  --ob-primary: var(--primary, 25 5% 22%);
  --ob-radius: var(--radius, 0.25rem);
  --ob-font-family: ui-sans-serif, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif;
  --ob-font-mono: ui-monospace, SFMono-Regular, "SF Mono", Menlo, Consolas, monospace;
  --ob-font-size: 16px;
  --ob-line-height: 1.6;
  --ob-block-spacing: 0.75em;
  --ob-content-padding: 1rem;
}

.openblock-container {
  position: relative;
  width: 100%;
}

.openblock-editor {
  font-family: var(--ob-font-family);
  font-size: var(--ob-font-size);
  line-height: var(--ob-line-height);
  color: hsl(var(--ob-foreground));
  background: hsl(var(--ob-background));
  padding: var(--ob-content-padding) var(--ob-content-padding) var(--ob-content-padding) calc(var(--ob-content-padding) + 48px);
  outline: none;
  min-height: 100px;
}

.openblock-editor:focus {
  outline: none;
}

/* ProseMirror Core */
.ProseMirror {
  outline: none;
  white-space: pre-wrap;
  word-wrap: break-word;
}

.ProseMirror-focused {
  outline: none;
}

.ProseMirror p.is-empty::before {
  content: attr(data-placeholder);
  color: hsl(var(--ob-muted-foreground));
  pointer-events: none;
  position: absolute;
}

.ProseMirror ::selection {
  background: hsl(var(--ob-primary) / 0.2);
}

.ProseMirror-gapcursor {
  display: none;
  pointer-events: none;
  position: absolute;
}

.ProseMirror-gapcursor:after {
  content: '';
  display: block;
  position: absolute;
  top: -2px;
  width: 20px;
  border-top: 1px solid hsl(var(--ob-foreground));
  animation: ProseMirror-cursor-blink 1.1s steps(2, start) infinite;
}

@keyframes ProseMirror-cursor-blink {
  to { visibility: hidden; }
}

.ProseMirror-focused .ProseMirror-gapcursor {
  display: block;
}

.ProseMirror-dropcursor {
  position: absolute;
  border-left: 2px solid hsl(var(--ob-primary));
  pointer-events: none;
}

/* Block Spacing */
.openblock-editor > * + * {
  margin-top: var(--ob-block-spacing);
}

/* Paragraph */
.openblock-editor p {
  margin: 0;
}

/* Headings */
.openblock-editor h1,
.openblock-editor h2,
.openblock-editor h3,
.openblock-editor h4,
.openblock-editor h5,
.openblock-editor h6 {
  margin: 0;
  font-weight: 600;
  line-height: 1.3;
}

.openblock-editor h1 {
  font-size: 2em;
  margin-top: 1em;
}

.openblock-editor h2 {
  font-size: 1.5em;
  margin-top: 0.875em;
}

.openblock-editor h3 {
  font-size: 1.25em;
  margin-top: 0.75em;
}

.openblock-editor h4 {
  font-size: 1.125em;
}

.openblock-editor h5 {
  font-size: 1em;
}

.openblock-editor h6 {
  font-size: 0.875em;
  color: hsl(var(--ob-muted-foreground));
}

/* Inline Formatting */
.openblock-editor strong {
  font-weight: 600;
}

.openblock-editor em {
  font-style: italic;
}

.openblock-editor u {
  text-decoration: underline;
}

.openblock-editor s {
  text-decoration: line-through;
}

.openblock-editor code {
  font-family: var(--ob-font-mono);
  font-size: 0.9em;
  background: hsl(var(--ob-muted));
  padding: 0.125em 0.25em;
  border-radius: calc(var(--ob-radius) / 2);
}

/* Links */
.openblock-editor a {
  color: hsl(var(--ob-primary));
  text-decoration: none;
}

.openblock-editor a:hover {
  text-decoration: underline;
}

/* Lists */
.openblock-editor ul,
.openblock-editor ol {
  margin: 0;
  padding-left: 1.5em;
}

.openblock-editor li {
  margin: 0.25em 0;
}

.openblock-editor li > p {
  margin: 0;
}

/* Blockquote */
.openblock-editor blockquote {
  margin: 0;
  padding-left: 1em;
  border-left: 3px solid hsl(var(--ob-border));
  color: hsl(var(--ob-muted-foreground));
}

/* Callout */
.openblock-editor .openblock-callout {
  margin: 0;
  padding: 0.75em 1em;
  border-radius: var(--ob-radius);
  border-left: 4px solid;
  background: hsl(var(--ob-muted) / 0.5);
}

.openblock-editor .openblock-callout--info {
  border-left-color: hsl(220 90% 56%);
  background: hsl(220 90% 56% / 0.1);
}

.openblock-editor .openblock-callout--warning {
  border-left-color: hsl(38 92% 50%);
  background: hsl(38 92% 50% / 0.1);
}

.openblock-editor .openblock-callout--success {
  border-left-color: hsl(142 76% 36%);
  background: hsl(142 76% 36% / 0.1);
}

.openblock-editor .openblock-callout--error {
  border-left-color: hsl(0 84% 60%);
  background: hsl(0 84% 60% / 0.1);
}

.openblock-editor .openblock-callout--note {
  border-left-color: hsl(var(--ob-muted-foreground));
  background: hsl(var(--ob-muted) / 0.5);
}

/* Code Block */
.openblock-editor pre {
  font-family: var(--ob-font-mono);
  font-size: 0.9em;
  background: hsl(var(--ob-muted));
  padding: 1em;
  border-radius: var(--ob-radius);
  overflow-x: auto;
  margin: 0;
}

.openblock-editor pre code {
  background: none;
  padding: 0;
}

/* Divider */
.openblock-editor hr {
  border: none;
  border-top: 1px solid hsl(var(--ob-border));
  margin: 1.5em 0;
}

/* Side Menu */
.ob-side-menu {
  position: absolute;
  display: flex;
  align-items: center;
  gap: 1px;
  opacity: 0;
  pointer-events: none;
  transition: opacity 0.15s ease;
  user-select: none;
  z-index: 100;
}

.ob-side-menu--visible {
  opacity: 1;
  pointer-events: auto;
}

.ob-add-button {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 18px;
  height: 24px;
  padding: 0;
  border: none;
  background: transparent;
  border-radius: 4px;
  color: #9ca3af;
  cursor: pointer;
  transition: background 0.15s ease, color 0.15s ease;
}

.ob-add-button svg {
  width: 14px;
  height: 14px;
}

.ob-add-button:hover {
  background: #f3f4f6;
  color: #374151;
}

.ob-add-button:active {
  background: #e5e7eb;
}

.ob-drag-handle {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 18px;
  height: 24px;
  cursor: grab;
  color: #9ca3af;
  border-radius: 4px;
  transition: background 0.15s ease, color 0.15s ease;
  user-select: none;
  -webkit-user-select: none;
}

.ob-drag-handle svg {
  width: 14px;
  height: 14px;
}

.ob-drag-handle:hover {
  background: #f3f4f6;
  color: #374151;
}

.ob-drag-handle--dragging {
  cursor: grabbing;
}

.ob-drag-handle:active {
  cursor: grabbing;
}

.ob-block-dragging {
  opacity: 0.4;
  background: hsl(var(--ob-muted));
  border-radius: var(--ob-radius);
}

.ob-drop-indicator {
  height: 3px;
  background: hsl(220 90% 56%);
  border-radius: 2px;
  margin: -2px 0;
  pointer-events: none;
  position: relative;
}

.ob-drop-indicator::before {
  content: '';
  position: absolute;
  left: -4px;
  top: 50%;
  transform: translateY(-50%);
  width: 8px;
  height: 8px;
  border-radius: 50%;
  background: hsl(220 90% 56%);
}

.openblock-editor > *,
.openblock-editor p,
.openblock-editor h1,
.openblock-editor h2,
.openblock-editor h3,
.openblock-editor h4,
.openblock-editor h5,
.openblock-editor h6,
.openblock-editor blockquote,
.openblock-editor .openblock-callout,
.openblock-editor pre,
.openblock-editor ul,
.openblock-editor ol,
.openblock-editor hr {
  position: relative;
}

.openblock-editor [data-block-id] {
  position: relative;
}

/* Slash Menu */
.ob-slash-menu {
  --ob-foreground: var(--foreground, 25 5% 22%);
  --ob-background: var(--background, 0 0% 100%);
  --ob-muted: var(--muted, 40 6% 96%);
  --ob-muted-foreground: var(--muted-foreground, 25 2% 57%);
  --ob-border: var(--border, 40 6% 90%);
  --ob-radius: var(--radius, 0.5rem);
  background: hsl(var(--ob-background));
  border: 1px solid hsl(var(--ob-border));
  border-radius: var(--ob-radius);
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.15), 0 2px 4px rgba(0, 0, 0, 0.1);
  max-height: 300px;
  min-width: 220px;
  max-width: 320px;
  overflow-y: auto;
  padding: 6px;
}

.ob-slash-menu-empty {
  padding: 8px 12px;
  color: hsl(var(--ob-muted-foreground));
  font-size: 0.875em;
  text-align: center;
}

.ob-slash-menu-item {
  display: flex;
  align-items: center;
  gap: 8px;
  padding: 8px 12px;
  border-radius: calc(var(--ob-radius) - 2px);
  cursor: pointer;
  transition: background 0.1s ease;
}

.ob-slash-menu-item:hover,
.ob-slash-menu-item--selected {
  background: hsl(var(--ob-muted));
}

.ob-slash-menu-item-icon {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  flex-shrink: 0;
  color: hsl(var(--ob-muted-foreground));
}

.ob-slash-menu-item--selected .ob-slash-menu-item-icon {
  color: hsl(var(--ob-foreground));
}

.ob-slash-menu-item-content {
  display: flex;
  flex-direction: column;
  gap: 2px;
  overflow: hidden;
}

.ob-slash-menu-item-title {
  font-size: 0.875em;
  font-weight: 500;
  color: hsl(var(--ob-foreground));
}

.ob-slash-menu-item-description {
  font-size: 0.75em;
  color: hsl(var(--ob-muted-foreground));
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

/* Bubble Menu */
.ob-bubble-menu {
  --ob-foreground: var(--foreground, 222 47% 11%);
  --ob-background: var(--background, 0 0% 100%);
  --ob-muted: var(--muted, 210 40% 96%);
  --ob-muted-foreground: var(--muted-foreground, 215 16% 47%);
  --ob-border: var(--border, 214 32% 91%);
  --ob-primary: var(--primary, 222 47% 11%);
  --ob-radius: var(--radius, 0.5rem);
  display: flex;
  align-items: center;
  gap: 1px;
  background: hsl(var(--ob-background));
  border: 1px solid hsl(var(--ob-border));
  border-radius: var(--ob-radius);
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.08), 0 4px 12px rgba(0, 0, 0, 0.05);
  padding: 4px 6px;
}

.ob-bubble-menu-btn {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 28px;
  height: 28px;
  padding: 0;
  border: none;
  background: transparent;
  border-radius: 4px;
  color: hsl(var(--ob-muted-foreground));
  cursor: pointer;
  transition: background 0.15s ease, color 0.15s ease;
}

.ob-bubble-menu-btn svg {
  width: 15px;
  height: 15px;
  stroke-width: 1.75;
  stroke-linecap: round;
  stroke-linejoin: round;
}

.ob-bubble-menu-btn:hover {
  background: hsl(var(--ob-muted));
  color: hsl(var(--ob-foreground));
}

.ob-bubble-menu-btn--active {
  background: hsl(215 20% 85%);
  color: hsl(var(--ob-foreground));
}

.ob-bubble-menu-btn--active:hover {
  background: hsl(215 20% 80%);
  color: hsl(var(--ob-foreground));
}

.ob-bubble-menu-divider {
  width: 1px;
  height: 16px;
  background: hsl(var(--ob-border));
  margin: 0 6px;
  flex-shrink: 0;
}

/* Text Alignment */
.ob-text-align-buttons {
  display: flex;
  align-items: center;
  gap: 1px;
}

/* Block Type Selector */
.ob-block-type-selector {
  position: relative;
}

.ob-block-type-selector-btn {
  display: flex;
  align-items: center;
  gap: 4px;
  padding: 4px 8px;
  border: none;
  background: transparent;
  border-radius: 4px;
  color: hsl(var(--ob-foreground));
  cursor: pointer;
  transition: background 0.15s ease;
  font-size: 13px;
  font-weight: 500;
  white-space: nowrap;
}

.ob-block-type-selector-btn:hover {
  background: hsl(var(--ob-muted));
}

.ob-block-type-selector-icon {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 16px;
  height: 16px;
  color: hsl(var(--ob-muted-foreground));
}

.ob-block-type-selector-icon svg {
  width: 14px;
  height: 14px;
  stroke-width: 1.75;
  stroke-linecap: round;
  stroke-linejoin: round;
}

.ob-block-type-selector-label {
  max-width: 100px;
  overflow: hidden;
  text-overflow: ellipsis;
}

.ob-block-type-selector-chevron {
  width: 14px;
  height: 14px;
  stroke-width: 2;
  stroke-linecap: round;
  stroke-linejoin: round;
  color: hsl(var(--ob-muted-foreground));
  transition: transform 0.15s ease;
}

.ob-block-type-selector-chevron--open {
  transform: rotate(180deg);
}

.ob-block-type-dropdown {
  position: absolute;
  top: 100%;
  left: 0;
  margin-top: 4px;
  background: hsl(var(--ob-background));
  border: 1px solid hsl(var(--ob-border));
  border-radius: var(--ob-radius);
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.12), 0 2px 4px rgba(0, 0, 0, 0.08);
  padding: 4px;
  min-width: 160px;
  max-height: 300px;
  overflow-y: auto;
  z-index: 1001;
  animation: ob-dropdown-fade-in 0.12s ease-out;
}

.ob-block-type-dropdown--upward {
  top: auto;
  bottom: 100%;
  margin-top: 0;
  margin-bottom: 4px;
  animation: ob-dropdown-fade-in-up 0.12s ease-out;
}

@keyframes ob-dropdown-fade-in {
  from {
    opacity: 0;
    transform: translateY(-4px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}

@keyframes ob-dropdown-fade-in-up {
  from {
    opacity: 0;
    transform: translateY(4px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}

.ob-block-type-option {
  display: flex;
  align-items: center;
  gap: 8px;
  width: 100%;
  padding: 8px 10px;
  border: none;
  background: transparent;
  border-radius: calc(var(--ob-radius) - 2px);
  color: hsl(var(--ob-foreground));
  cursor: pointer;
  transition: background 0.1s ease;
  text-align: left;
  font-size: 13px;
}

.ob-block-type-option:hover {
  background: hsl(var(--ob-muted));
}

.ob-block-type-option--active {
  background: hsl(var(--ob-muted) / 0.7);
}

.ob-block-type-option-icon {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 18px;
  height: 18px;
  color: hsl(var(--ob-muted-foreground));
  flex-shrink: 0;
}

.ob-block-type-option-icon svg {
  width: 16px;
  height: 16px;
  stroke-width: 1.75;
  stroke-linecap: round;
  stroke-linejoin: round;
}

.ob-block-type-option-label {
  flex: 1;
}

.ob-block-type-option-check {
  width: 16px;
  height: 16px;
  stroke-width: 2;
  stroke-linecap: round;
  stroke-linejoin: round;
  color: hsl(var(--ob-primary));
  flex-shrink: 0;
}

/* Column Layout */
.ob-column-list {
  display: flex;
  gap: 8px;
  width: 100%;
  margin: 0.5em 0;
  position: relative;
}

.ob-column {
  position: relative;
  min-width: 80px;
  min-height: 1em;
  padding: 0 8px 0 48px;
}

.ob-column:first-child {
  padding-left: 48px;
}

.ob-column:last-child {
  padding-right: 0;
}

.ob-column:not(:last-child)::after {
  content: '';
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
  width: 1px;
  background: hsl(var(--ob-border) / 0.3);
}

.ob-column:empty::before,
.ob-column > p:only-child:empty::before {
  content: 'Type something...';
  color: hsl(var(--ob-muted-foreground));
  pointer-events: none;
}

/* Table */
.openblock-editor table,
.openblock-editor .ob-table {
  border-collapse: collapse;
  width: calc(100% - 28px);
  margin: 0.5em 0 28px 0;
  table-layout: fixed;
}

.openblock-editor th,
.openblock-editor td,
.openblock-editor .ob-table-cell,
.openblock-editor .ob-table-header {
  border: 1px solid hsl(var(--ob-border));
  padding: 0.5em 0.75em;
  text-align: left;
  vertical-align: top;
  position: relative;
  min-width: 50px;
}

.openblock-editor th,
.openblock-editor .ob-table-header {
  background: hsl(var(--ob-muted));
  font-weight: 600;
}

.openblock-editor .ob-table-cell > p,
.openblock-editor .ob-table-header > p,
.openblock-editor td > p,
.openblock-editor th > p {
  margin: 0;
}

.openblock-editor .ob-table-cell > p:first-child,
.openblock-editor .ob-table-header > p:first-child {
  margin-top: 0;
}

.openblock-editor .ob-table-cell > p:last-child,
.openblock-editor .ob-table-header > p:last-child {
  margin-bottom: 0;
}

.openblock-editor .ob-table-cell.selectedCell,
.openblock-editor .ob-table-header.selectedCell {
  background: hsl(220 90% 56% / 0.1);
}

.openblock-editor .ob-table-cell-resize-handle,
.openblock-editor .ob-table-header-resize-handle {
  position: absolute;
  right: -2px;
  top: 0;
  bottom: 0;
  width: 4px;
  cursor: col-resize;
  background: transparent;
  z-index: 10;
}

.openblock-editor .ob-table-cell-resize-handle:hover,
.openblock-editor .ob-table-header-resize-handle:hover {
  background: hsl(var(--ob-primary) / 0.3);
}

.openblock-editor .ob-table-wrapper {
  overflow-x: auto;
  margin: 0.5em 0;
}

/* Link Popover */
.ob-link-popover {
  --ob-foreground: var(--foreground, 222 47% 11%);
  --ob-background: var(--background, 0 0% 100%);
  --ob-muted: var(--muted, 210 40% 96%);
  --ob-muted-foreground: var(--muted-foreground, 215 16% 47%);
  --ob-border: var(--border, 214 32% 91%);
  --ob-primary: var(--primary, 222 47% 11%);
  --ob-destructive: var(--destructive, 0 84% 60%);
  --ob-radius: var(--radius, 0.5rem);
  background: hsl(var(--ob-background));
  border: 1px solid hsl(var(--ob-border));
  border-radius: var(--ob-radius);
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.12), 0 2px 4px rgba(0, 0, 0, 0.08);
  padding: 4px;
  animation: ob-link-popover-fade-in 0.12s ease-out;
}

@keyframes ob-link-popover-fade-in {
  from {
    opacity: 0;
    transform: translateY(-4px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}

.ob-link-popover-form {
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.ob-link-popover-input-row {
  display: flex;
}

.ob-link-popover-input-wrapper {
  display: flex;
  align-items: center;
  gap: 2px;
  padding: 2px 4px 2px 8px;
  border: 1px solid hsl(var(--ob-border));
  border-radius: calc(var(--ob-radius) - 2px);
  background: hsl(var(--ob-background));
  transition: border-color 0.15s ease, box-shadow 0.15s ease;
}

.ob-link-popover-input-wrapper:focus-within {
  border-color: hsl(var(--ob-primary) / 0.5);
  box-shadow: 0 0 0 2px hsl(var(--ob-primary) / 0.1);
}

.ob-link-popover-input-wrapper--error {
  border-color: hsl(var(--ob-destructive));
}

.ob-link-popover-input-wrapper--error:focus-within {
  border-color: hsl(var(--ob-destructive));
  box-shadow: 0 0 0 2px hsl(var(--ob-destructive) / 0.1);
}

.ob-link-popover-input-icon {
  flex-shrink: 0;
  width: 14px;
  height: 14px;
  stroke-width: 1.75;
  stroke-linecap: round;
  stroke-linejoin: round;
  color: hsl(var(--ob-muted-foreground));
}

.ob-link-popover-input {
  flex: 1;
  min-width: 180px;
  padding: 6px 8px;
  font-size: 13px;
  border: none;
  background: transparent;
  color: hsl(var(--ob-foreground));
  outline: none;
}

.ob-link-popover-input::placeholder {
  color: hsl(var(--ob-muted-foreground));
}

.ob-link-popover-inline-btn {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 28px;
  height: 28px;
  padding: 0;
  border: none;
  border-radius: 4px;
  background: transparent;
  color: hsl(var(--ob-muted-foreground));
  cursor: pointer;
  transition: background 0.15s ease, color 0.15s ease;
  flex-shrink: 0;
}

.ob-link-popover-inline-btn svg {
  width: 14px;
  height: 14px;
  stroke-width: 1.75;
  stroke-linecap: round;
  stroke-linejoin: round;
}

.ob-link-popover-inline-btn:hover {
  background: hsl(var(--ob-muted));
  color: hsl(var(--ob-foreground));
}

.ob-link-popover-inline-btn--primary {
  background: hsl(var(--ob-primary));
  color: hsl(var(--ob-background));
}

.ob-link-popover-inline-btn--primary:hover {
  background: hsl(var(--ob-primary) / 0.85);
  color: hsl(var(--ob-background));
}

.ob-link-popover-inline-btn--danger:hover {
  background: hsl(var(--ob-destructive) / 0.1);
  color: hsl(var(--ob-destructive));
}

.ob-link-popover-error {
  margin: 0;
  padding: 0 8px;
  font-size: 11px;
  color: hsl(var(--ob-destructive));
}

/* Color Picker */
.ob-color-picker {
  position: relative;
}

.ob-color-picker > .ob-bubble-menu-btn {
  position: relative;
}

.ob-color-picker-indicator {
  position: absolute;
  bottom: 3px;
  left: 50%;
  transform: translateX(-50%);
  width: 14px;
  height: 3px;
  border-radius: 2px;
}

.ob-color-picker-dropdown {
  background: hsl(var(--ob-background));
  border: 1px solid hsl(var(--ob-border));
  border-radius: var(--ob-radius);
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.12), 0 2px 4px rgba(0, 0, 0, 0.08);
  padding: 8px;
  min-width: 160px;
  z-index: 1002;
}

.ob-color-picker-section {
  padding: 4px 0;
}

.ob-color-picker-section:first-child {
  padding-top: 0;
}

.ob-color-picker-section:last-child {
  padding-bottom: 0;
}

.ob-color-picker-divider {
  height: 1px;
  background: hsl(var(--ob-border));
  margin: 8px 0;
}

.ob-color-picker-label {
  font-size: 11px;
  font-weight: 500;
  color: hsl(var(--ob-muted-foreground));
  margin-bottom: 8px;
  text-transform: uppercase;
  letter-spacing: 0.05em;
}

.ob-color-picker-grid {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  gap: 4px;
}

.ob-color-picker-option {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  padding: 0;
  border: 1px solid transparent;
  background: transparent;
  border-radius: 6px;
  cursor: pointer;
  transition: border-color 0.15s ease, transform 0.1s ease;
}

.ob-color-picker-option:hover {
  border-color: hsl(var(--ob-border));
  transform: scale(1.05);
}

.ob-color-picker-option--active {
  border-color: hsl(var(--ob-primary));
  background: hsl(var(--ob-muted) / 0.5);
}

.ob-color-picker-option svg {
  width: 20px;
  height: 20px;
  stroke-width: 1.5;
  stroke-linecap: round;
  stroke-linejoin: round;
  color: hsl(var(--ob-muted-foreground));
}

.ob-color-picker-swatch {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 4px;
  font-size: 12px;
  font-weight: 600;
}

.ob-color-picker-swatch--text {
  background: transparent;
  font-size: 14px;
}

/* Multi-Block Selection */
.ob-block-selected {
  background: hsl(220 90% 56% / 0.1);
  outline: 2px solid hsl(220 90% 56% / 0.5);
  outline-offset: -2px;
  border-radius: var(--ob-radius);
}

.ob-block-selected::before {
  content: '';
  position: absolute;
  left: -24px;
  top: 50%;
  transform: translateY(-50%);
  width: 8px;
  height: 8px;
  background: hsl(220 90% 56%);
  border-radius: 50%;
}

/* Table Menu */
.ob-table-menu {
  --ob-foreground: var(--foreground, 222 47% 11%);
  --ob-background: var(--background, 0 0% 100%);
  --ob-muted: var(--muted, 210 40% 96%);
  --ob-muted-foreground: var(--muted-foreground, 215 16% 47%);
  --ob-border: var(--border, 214 32% 91%);
  --ob-primary: var(--primary, 222 47% 11%);
  --ob-destructive: var(--destructive, 0 84% 60%);
  --ob-radius: var(--radius, 0.5rem);
  display: flex;
  align-items: center;
  gap: 4px;
  background: hsl(var(--ob-background));
  border: 1px solid hsl(var(--ob-border));
  border-radius: var(--ob-radius);
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.08), 0 4px 12px rgba(0, 0, 0, 0.05);
  padding: 4px 8px;
  animation: ob-table-menu-fade-in 0.15s ease-out;
}

@keyframes ob-table-menu-fade-in {
  from {
    opacity: 0;
    transform: translateY(4px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}

.ob-table-menu-btn {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 28px;
  height: 28px;
  padding: 0;
  border: none;
  background: transparent;
  border-radius: 4px;
  color: hsl(var(--ob-muted-foreground));
  cursor: pointer;
  transition: background 0.15s ease, color 0.15s ease;
}

.ob-table-menu-btn svg {
  width: 16px;
  height: 16px;
  stroke-linecap: round;
  stroke-linejoin: round;
}

.ob-table-menu-btn:hover {
  background: hsl(var(--ob-muted));
  color: hsl(var(--ob-foreground));
}

.ob-table-menu-btn--danger {
  color: hsl(var(--ob-destructive));
}

.ob-table-menu-btn--danger:hover {
  background: hsl(var(--ob-destructive) / 0.1);
  color: hsl(var(--ob-destructive));
}

.ob-table-menu-divider {
  width: 1px;
  height: 20px;
  background: hsl(var(--ob-border));
  margin: 0 4px;
  flex-shrink: 0;
}

.ob-table-menu-info {
  font-size: 11px;
  color: hsl(var(--ob-muted-foreground));
  padding: 0 8px;
  white-space: nowrap;
}

.ob-table-menu-dropdown {
  position: relative;
}

.ob-table-menu-dropdown-btn {
  display: flex;
  align-items: center;
  gap: 4px;
  padding: 4px 8px;
  border: none;
  background: transparent;
  border-radius: 4px;
  color: hsl(var(--ob-foreground));
  cursor: pointer;
  transition: background 0.15s ease;
  font-size: 13px;
  font-weight: 500;
  white-space: nowrap;
}

.ob-table-menu-dropdown-btn:hover {
  background: hsl(var(--ob-muted));
}

.ob-table-menu-dropdown-icon {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 16px;
  height: 16px;
  color: hsl(var(--ob-muted-foreground));
}

.ob-table-menu-dropdown-icon svg {
  width: 14px;
  height: 14px;
  stroke-linecap: round;
  stroke-linejoin: round;
}

.ob-table-menu-dropdown-label {
  min-width: 40px;
}

.ob-table-menu-dropdown-chevron {
  width: 12px;
  height: 12px;
  stroke-width: 2;
  stroke-linecap: round;
  stroke-linejoin: round;
  color: hsl(var(--ob-muted-foreground));
  transition: transform 0.15s ease;
}

.ob-table-menu-dropdown-chevron--open {
  transform: rotate(180deg);
}

.ob-table-menu-dropdown-content {
  position: absolute;
  top: 100%;
  left: 0;
  margin-top: 4px;
  background: hsl(var(--ob-background));
  border: 1px solid hsl(var(--ob-border));
  border-radius: var(--ob-radius);
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.12), 0 2px 4px rgba(0, 0, 0, 0.08);
  padding: 4px;
  min-width: 180px;
  z-index: 1001;
  animation: ob-dropdown-fade-in 0.12s ease-out;
}

.ob-table-menu-dropdown-item {
  display: flex;
  align-items: center;
  gap: 10px;
  width: 100%;
  padding: 8px 12px;
  border: none;
  background: transparent;
  border-radius: calc(var(--ob-radius) - 2px);
  color: hsl(var(--ob-foreground));
  cursor: pointer;
  transition: background 0.1s ease;
  text-align: left;
  font-size: 13px;
}

.ob-table-menu-dropdown-item:hover {
  background: hsl(var(--ob-muted));
}

.ob-table-menu-dropdown-item:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.ob-table-menu-dropdown-item:disabled:hover {
  background: transparent;
}

.ob-table-menu-dropdown-item svg {
  width: 16px;
  height: 16px;
  stroke-linecap: round;
  stroke-linejoin: round;
  color: hsl(var(--ob-muted-foreground));
  flex-shrink: 0;
}

.ob-table-menu-dropdown-item span {
  flex: 1;
}

.ob-table-menu-dropdown-item--danger {
  color: hsl(var(--ob-destructive));
}

.ob-table-menu-dropdown-item--danger svg {
  color: hsl(var(--ob-destructive));
}

.ob-table-menu-dropdown-item--danger:hover {
  background: hsl(var(--ob-destructive) / 0.1);
}

/* Table Handles */
.ob-table-handles {
  pointer-events: none;
  z-index: 100;
}

.ob-table-handle {
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: 0;
  transition: opacity 0.15s ease;
  pointer-events: auto;
}

.ob-table-handle--visible {
  opacity: 1;
}

.ob-table-handle--row {
  width: 24px;
}

.ob-table-handle--col {
  height: 24px;
}

.ob-table-handle-btn {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 20px;
  height: 20px;
  padding: 0;
  border: none;
  background: hsl(var(--ob-background, 0 0% 100%));
  border: 1px solid hsl(var(--ob-border, 214 32% 91%));
  border-radius: 4px;
  color: hsl(var(--ob-muted-foreground, 215 16% 47%));
  cursor: pointer;
  transition: background 0.15s ease, color 0.15s ease, box-shadow 0.15s ease;
  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.05);
}

.ob-table-handle-btn:hover {
  background: hsl(var(--ob-muted, 210 40% 96%));
  color: hsl(var(--ob-foreground, 222 47% 11%));
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
}

.ob-table-handle-btn svg {
  width: 12px;
  height: 12px;
}

.ob-table-handle-menu {
  position: fixed;
  background: hsl(var(--ob-background, 0 0% 100%));
  border: 1px solid hsl(var(--ob-border, 214 32% 91%));
  border-radius: 8px;
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.12), 0 2px 4px rgba(0, 0, 0, 0.08);
  padding: 4px;
  min-width: 150px;
  z-index: 1001;
  animation: ob-dropdown-fade-in 0.12s ease-out;
}

.ob-table-handle-menu button {
  display: flex;
  align-items: center;
  gap: 8px;
  width: 100%;
  padding: 8px 12px;
  border: none;
  background: transparent;
  border-radius: 4px;
  color: hsl(var(--ob-foreground, 222 47% 11%));
  cursor: pointer;
  transition: background 0.1s ease;
  text-align: left;
  font-size: 13px;
  white-space: nowrap;
}

.ob-table-handle-menu button:hover {
  background: hsl(var(--ob-muted, 210 40% 96%));
}

.ob-table-handle-menu button svg {
  width: 14px;
  height: 14px;
  flex-shrink: 0;
}

.ob-table-handle-menu-danger {
  color: hsl(var(--ob-destructive, 0 84% 60%)) !important;
}

.ob-table-handle-menu-danger:hover {
  background: hsl(var(--ob-destructive, 0 84% 60%) / 0.1) !important;
}

.ob-table-extend-btn {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0;
  border: none;
  background: hsl(var(--ob-muted-foreground, 215 16% 47%) / 0.15);
  color: hsl(var(--ob-muted-foreground, 215 16% 47%));
  cursor: pointer;
  transition: all 0.15s ease;
  opacity: 0;
  pointer-events: auto;
  border-radius: 4px;
}

.ob-table-extend-btn--visible,
.ob-table-handles:hover .ob-table-extend-btn {
  opacity: 1;
}

.ob-table-extend-btn:hover {
  color: hsl(var(--ob-foreground, 222 47% 11%));
  background: hsl(var(--ob-muted-foreground, 215 16% 47%) / 0.3);
}

.ob-table-extend-btn svg {
  width: 14px;
  height: 14px;
}

.ob-table-extend-btn--col {
  width: 20px;
}

.ob-table-extend-btn--row {
  height: 20px;
}

.ob-table-cell .ob-side-menu,
.ob-table-header .ob-side-menu,
td .ob-side-menu,
th .ob-side-menu,
.openblock-editor table .ob-side-menu {
  display: none !important;
}

/* Checklist / To-do List */
.openblock-editor ul.openblock-checklist {
  padding-left: 0;
}

.openblock-checklist {
  list-style: none !important;
  list-style-type: none !important;
  margin: 0;
  padding: 0 !important;
  padding-left: 0 !important;
}

.openblock-checklist-item {
  display: flex;
  align-items: flex-start;
  gap: 0.5em;
  margin: 0.25em 0;
  padding-left: 0.25em;
  position: relative;
  list-style: none !important;
  list-style-type: none !important;
}

.openblock-checklist > .openblock-checklist-item::marker {
  content: none;
}

.openblock-checklist > .openblock-checklist-item::before {
  content: none;
  display: none;
}

.openblock-checklist-label {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  margin-top: 0.2em;
}

.openblock-checklist-checkbox {
  width: 16px;
  height: 16px;
  margin: 0;
  cursor: pointer;
  accent-color: hsl(var(--ob-primary));
}

.openblock-checklist-content {
  flex: 1;
  min-width: 0;
}

.openblock-checklist-item--checked .openblock-checklist-content {
  text-decoration: line-through;
  color: hsl(var(--ob-muted-foreground));
}

/* Image Block */
.openblock-image {
  margin: 0.5em 0;
  text-align: center;
}

.openblock-image--left {
  text-align: left;
}

.openblock-image--center {
  text-align: center;
}

.openblock-image--right {
  text-align: right;
}

.openblock-image img {
  max-width: 100%;
  height: auto;
  border-radius: var(--ob-radius);
}

.openblock-image figcaption {
  margin-top: 0.5em;
  font-size: 0.875em;
  color: hsl(var(--ob-muted-foreground));
  text-align: center;
}

/* Image Placeholder */
.openblock-image-placeholder {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-height: 150px;
  padding: 2em;
  background: hsl(var(--ob-muted));
  border: 2px dashed hsl(var(--ob-border));
  border-radius: var(--ob-radius);
  cursor: pointer;
  transition: background 0.15s ease, border-color 0.15s ease;
}

.openblock-image-placeholder:hover {
  background: hsl(var(--ob-muted) / 0.7);
  border-color: hsl(var(--ob-muted-foreground));
}

.openblock-image-placeholder-icon {
  display: block;
  width: 48px;
  height: 48px;
  margin-bottom: 0.75em;
  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24' fill='none' stroke='%239ca3af' stroke-width='1.5' stroke-linecap='round' stroke-linejoin='round'%3E%3Crect x='3' y='3' width='18' height='18' rx='2' ry='2'/%3E%3Ccircle cx='8.5' cy='8.5' r='1.5'/%3E%3Cpath d='M21 15l-5-5L5 21'/%3E%3C/svg%3E");
  background-size: contain;
  background-repeat: no-repeat;
  background-position: center;
}

.openblock-image-placeholder-text {
  font-size: 0.875em;
  color: hsl(var(--ob-muted-foreground));
}

/* Embed Block */
.openblock-embed {
  margin: 0.5em 0;
}

.openblock-embed-container {
  position: relative;
  width: 100%;
  overflow: hidden;
  border-radius: var(--ob-radius);
  background: hsl(var(--ob-muted));
}

.openblock-embed-container iframe {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border: none;
}

.openblock-embed-placeholder {
  display: flex;
  align-items: center;
  justify-content: center;
  min-height: 200px;
  color: hsl(var(--ob-muted-foreground));
}

.openblock-embed-caption {
  margin-top: 0.5em;
  font-size: 0.875em;
  color: hsl(var(--ob-muted-foreground));
  text-align: center;
}

/* Media Menu */
.ob-media-menu {
  --ob-foreground: var(--foreground, 222 47% 11%);
  --ob-background: var(--background, 0 0% 100%);
  --ob-muted: var(--muted, 210 40% 96%);
  --ob-muted-foreground: var(--muted-foreground, 215 16% 47%);
  --ob-border: var(--border, 214 32% 91%);
  --ob-primary: var(--primary, 222 47% 11%);
  --ob-radius: var(--radius, 0.5rem);
  display: flex;
  align-items: center;
  gap: 1px;
  background: hsl(var(--ob-background));
  border: 1px solid hsl(var(--ob-border));
  border-radius: var(--ob-radius);
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.08), 0 4px 12px rgba(0, 0, 0, 0.05);
  padding: 4px 6px;
  position: relative;
}

.ob-media-menu-btn {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 28px;
  height: 28px;
  padding: 0;
  border: none;
  background: transparent;
  border-radius: 4px;
  color: hsl(var(--ob-muted-foreground));
  cursor: pointer;
  transition: background 0.15s ease, color 0.15s ease;
}

.ob-media-menu-btn svg {
  width: 15px;
  height: 15px;
  stroke-width: 1.75;
  stroke-linecap: round;
  stroke-linejoin: round;
}

.ob-media-menu-btn:hover {
  background: hsl(var(--ob-muted));
  color: hsl(var(--ob-foreground));
}

.ob-media-menu-btn--active {
  background: hsl(215 20% 85%);
  color: hsl(var(--ob-foreground));
}

.ob-media-menu-btn--active:hover {
  background: hsl(215 20% 80%);
  color: hsl(var(--ob-foreground));
}

.ob-media-menu-btn--danger:hover {
  background: hsl(0 84% 60% / 0.1);
  color: hsl(0 84% 60%);
}

.ob-media-menu-divider {
  width: 1px;
  height: 16px;
  background: hsl(var(--ob-border));
  margin: 0 6px;
  flex-shrink: 0;
}

.ob-media-url-popover {
  position: absolute;
  top: calc(100% + 8px);
  left: 50%;
  transform: translateX(-50%);
  background: hsl(var(--ob-background));
  border: 1px solid hsl(var(--ob-border));
  border-radius: var(--ob-radius);
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.12), 0 2px 4px rgba(0, 0, 0, 0.08);
  padding: 12px;
  min-width: 280px;
  z-index: 1001;
}

.ob-media-url-label {
  display: block;
  font-size: 12px;
  font-weight: 500;
  color: hsl(var(--ob-muted-foreground));
  margin-bottom: 8px;
}

.ob-media-url-input {
  width: 100%;
  padding: 8px 10px;
  border: 1px solid hsl(var(--ob-border));
  border-radius: 4px;
  font-size: 13px;
  color: hsl(var(--ob-foreground));
  background: hsl(var(--ob-background));
  outline: none;
  transition: border-color 0.15s ease, box-shadow 0.15s ease;
}

.ob-media-url-input:focus {
  border-color: hsl(var(--ob-primary));
  box-shadow: 0 0 0 2px hsl(var(--ob-primary) / 0.1);
}

.ob-media-url-input::placeholder {
  color: hsl(var(--ob-muted-foreground));
}

.ob-media-url-actions {
  display: flex;
  justify-content: flex-end;
  gap: 8px;
  margin-top: 12px;
}

.ob-media-url-btn {
  padding: 6px 12px;
  border: none;
  border-radius: 4px;
  font-size: 13px;
  font-weight: 500;
  cursor: pointer;
  transition: background 0.15s ease;
}

.ob-media-url-btn--cancel {
  background: hsl(var(--ob-muted));
  color: hsl(var(--ob-foreground));
}

.ob-media-url-btn--cancel:hover {
  background: hsl(var(--ob-border));
}

.ob-media-url-btn--save {
  background: hsl(var(--ob-primary));
  color: hsl(var(--ob-background));
}

.ob-media-url-btn--save:hover {
  opacity: 0.9;
}
`;
function injectStyles() {
  if (typeof document === "undefined") {
    return false;
  }
  if (stylesInjected || document.getElementById(STYLE_ID)) {
    stylesInjected = true;
    return false;
  }
  const style = document.createElement("style");
  style.id = STYLE_ID;
  style.textContent = EDITOR_STYLES;
  document.head.appendChild(style);
  stylesInjected = true;
  return true;
}
function removeStyles() {
  if (typeof document === "undefined") {
    return;
  }
  const style = document.getElementById(STYLE_ID);
  if (style) {
    style.remove();
    stylesInjected = false;
  }
}
function areStylesInjected() {
  if (typeof document === "undefined") {
    return false;
  }
  return stylesInjected || !!document.getElementById(STYLE_ID);
}
function blockToNode(schema, block) {
  const nodeType = schema.nodes[block.type];
  if (!nodeType) {
    console.warn(`Unknown block type: ${block.type}, falling back to paragraph`);
    return schema.node("paragraph", { id: block.id || v4() });
  }
  const attrs = {
    id: block.id || v4(),
    ...block.props
  };
  if (block.children && block.children.length > 0) {
    const childNodes = block.children.map((child) => blockToNode(schema, child));
    return nodeType.create(attrs, childNodes);
  }
  if (block.type === "listItem" && block.content) {
    const paragraph = schema.node("paragraph", null, inlineContentToNodes(schema, block.content));
    return nodeType.create(attrs, [paragraph]);
  }
  if (block.type === "checkListItem" && block.content) {
    return nodeType.create(attrs, inlineContentToNodes(schema, block.content));
  }
  if (block.type === "image") {
    return nodeType.create(attrs);
  }
  if ((block.type === "tableCell" || block.type === "tableHeader") && block.content && !block.children) {
    const paragraph = schema.node("paragraph", null, inlineContentToNodes(schema, block.content));
    return nodeType.create(attrs, [paragraph]);
  }
  if (!block.content || block.content.length === 0) {
    if (nodeType.spec.content === "" || nodeType.spec.content === void 0) {
      return nodeType.create(attrs);
    }
    return nodeType.create(attrs, []);
  }
  const content = inlineContentToNodes(schema, block.content);
  return nodeType.create(attrs, content);
}
function inlineContentToNodes(schema, content) {
  return content.map((item) => {
    if (item.type === "text") {
      const marks = stylesToMarks(schema, item.styles);
      return schema.text(item.text, marks);
    }
    return schema.text("");
  });
}
function stylesToMarks(schema, styles) {
  const marks = [];
  if (styles.bold && schema.marks.bold) {
    marks.push(schema.marks.bold.create());
  }
  if (styles.italic && schema.marks.italic) {
    marks.push(schema.marks.italic.create());
  }
  if (styles.underline && schema.marks.underline) {
    marks.push(schema.marks.underline.create());
  }
  if (styles.strikethrough && schema.marks.strikethrough) {
    marks.push(schema.marks.strikethrough.create());
  }
  if (styles.code && schema.marks.code) {
    marks.push(schema.marks.code.create());
  }
  return marks;
}
function blocksToDoc(schema, blocks) {
  if (!blocks || blocks.length === 0) {
    return schema.node("doc", null, [
      schema.node("paragraph", { id: v4() })
    ]);
  }
  const nodes = blocks.map((block) => blockToNode(schema, block));
  return schema.node("doc", null, nodes);
}
const CONTAINER_TYPES = /* @__PURE__ */ new Set([
  "bulletList",
  "orderedList",
  "listItem",
  "checkList",
  "checkListItem",
  "table",
  "tableRow",
  "tableCell",
  "tableHeader"
]);
function nodeToBlock(node) {
  const { id, ...props } = node.attrs;
  const block = {
    id: id || v4(),
    type: node.type.name,
    props
  };
  if (CONTAINER_TYPES.has(node.type.name)) {
    if (node.type.name === "listItem") {
      const firstChild = node.firstChild;
      if (firstChild && firstChild.isTextblock) {
        block.content = nodeContentToInline(firstChild);
      }
      if (node.childCount > 1) {
        const children = [];
        node.forEach((child, _offset, index) => {
          if (index > 0) {
            children.push(nodeToBlock(child));
          }
        });
        block.children = children;
      }
    } else if (node.type.name === "checkListItem") {
      block.content = nodeContentToInline(node);
    } else if (node.type.name === "tableCell" || node.type.name === "tableHeader") {
      const cellChildren = [];
      node.forEach((child) => {
        cellChildren.push(nodeToBlock(child));
      });
      block.children = cellChildren;
    } else {
      const blockChildren = [];
      node.forEach((child) => {
        blockChildren.push(nodeToBlock(child));
      });
      block.children = blockChildren;
    }
    return block;
  }
  if (node.isTextblock && node.content.size > 0) {
    block.content = nodeContentToInline(node);
  }
  return block;
}
function nodeContentToInline(node) {
  const content = [];
  node.content.forEach((child) => {
    if (child.isText) {
      content.push({
        type: "text",
        text: child.text || "",
        styles: marksToStyles(child.marks)
      });
    }
  });
  return content;
}
function marksToStyles(marks) {
  const styles = {};
  for (const mark of marks) {
    switch (mark.type.name) {
      case "bold":
        styles.bold = true;
        break;
      case "italic":
        styles.italic = true;
        break;
      case "underline":
        styles.underline = true;
        break;
      case "strikethrough":
        styles.strikethrough = true;
        break;
      case "code":
        styles.code = true;
        break;
    }
  }
  return styles;
}
function docToBlocks(doc) {
  const blocks = [];
  doc.forEach((node) => {
    blocks.push(nodeToBlock(node));
  });
  return blocks;
}
class OpenBlockEditor {
  constructor(config = {}) {
    this._listeners = /* @__PURE__ */ new Map();
    this._destroyed = false;
    this._config = { ...defaultConfig$1, ...config };
    this.pm = new ProseMirrorAPI(this);
    if (this._config.injectStyles !== false) {
      injectStyles();
    }
    this._schema = createSchema();
    this._createEditor();
    if (this._config.autoFocus) {
      this.focus(this._config.autoFocus === "start" ? "start" : "end");
    }
  }
  // ===========================================================================
  // EDITOR LIFECYCLE
  // ===========================================================================
  _createEditor() {
    var _a;
    const doc = blocksToDoc(this._schema, this._config.initialContent);
    const plugins = createPlugins({
      schema: this._schema,
      toggleMark: (name) => this.pm.toggleMark(name),
      inputRules: this._config.inputRules,
      additionalPlugins: (_a = this._config.prosemirror) == null ? void 0 : _a.plugins
    });
    const state = EditorState.create({ doc, schema: this._schema, plugins });
    this._pmView = new EditorView(this._config.element ?? null, {
      state,
      editable: () => this._config.editable !== false,
      dispatchTransaction: this._handleTransaction.bind(this),
      attributes: {
        class: "openblock-editor",
        role: "textbox",
        "aria-multiline": "true"
      },
      ...this._config.prosemirror
    });
  }
  _handleTransaction(tr) {
    var _a, _b, _c, _d;
    if (this._destroyed) return;
    const newState = this._pmView.state.apply(tr);
    this._pmView.updateState(newState);
    if (tr.docChanged) {
      const blocks = this.getDocument();
      this._emit("change", { blocks });
      (_b = (_a = this._config).onUpdate) == null ? void 0 : _b.call(_a, blocks);
    }
    if (tr.selectionSet) {
      const blocks = this.getSelectedBlocks();
      this._emit("selectionChange", { blocks });
      (_d = (_c = this._config).onSelectionChange) == null ? void 0 : _d.call(_c, blocks);
    }
    this._emit("transaction", { transaction: tr });
  }
  // ===========================================================================
  // DOCUMENT OPERATIONS
  // ===========================================================================
  /**
   * Get the entire document as an array of blocks.
   *
   * @example
   * ```typescript
   * const blocks = editor.getDocument();
   * console.log(JSON.stringify(blocks, null, 2));
   * ```
   */
  getDocument() {
    return docToBlocks(this.pm.doc);
  }
  /**
   * Replace the entire document content.
   *
   * @example
   * ```typescript
   * editor.setDocument([
   *   { id: '1', type: 'heading', props: { level: 1 }, content: [{ type: 'text', text: 'Title', styles: {} }] },
   *   { id: '2', type: 'paragraph', props: {}, content: [{ type: 'text', text: 'Content', styles: {} }] },
   * ]);
   * ```
   */
  setDocument(blocks) {
    const doc = blocksToDoc(this._schema, blocks);
    const tr = this.pm.createTransaction();
    tr.replaceWith(0, this.pm.doc.content.size, doc.content);
    this.pm.dispatch(tr);
  }
  /**
   * Find a block by its ID.
   *
   * @param id - The block's unique identifier
   * @returns The block if found, undefined otherwise
   *
   * @example
   * ```typescript
   * const block = editor.getBlock('my-block-id');
   * if (block) {
   *   console.log(block.type, block.props);
   * }
   * ```
   */
  getBlock(id) {
    let result;
    this.pm.doc.descendants((node, _pos) => {
      if (node.attrs.id === id) {
        result = nodeToBlock(node);
        return false;
      }
    });
    return result;
  }
  /**
   * Get blocks that overlap with the current selection.
   *
   * @returns Array of blocks within the selection range
   */
  getSelectedBlocks() {
    const { from, to } = this.pm.selection;
    const blocks = [];
    this.pm.doc.nodesBetween(from, to, (node, _pos) => {
      if (node.isBlock && node.type.name !== "doc") {
        blocks.push(nodeToBlock(node));
      }
    });
    return blocks;
  }
  // ===========================================================================
  // BLOCK OPERATIONS
  // ===========================================================================
  /**
   * Insert new blocks relative to an existing block.
   *
   * @param blocks - Blocks to insert (IDs will be generated if not provided)
   * @param referenceBlock - The block ID or block object to position relative to
   * @param placement - Insert 'before' or 'after' the reference block
   *
   * @example
   * ```typescript
   * // Insert a paragraph after block with id 'intro'
   * editor.insertBlocks(
   *   [{ type: 'paragraph', content: [{ type: 'text', text: 'New paragraph', styles: {} }] }],
   *   'intro',
   *   'after'
   * );
   * ```
   */
  insertBlocks(blocks, referenceBlock, placement = "after") {
    const refId = typeof referenceBlock === "string" ? referenceBlock : referenceBlock.id;
    let insertPos = null;
    this.pm.doc.descendants((node, pos) => {
      if (node.attrs.id === refId) {
        insertPos = placement === "before" ? pos : pos + node.nodeSize;
        return false;
      }
    });
    if (insertPos === null) {
      console.warn(`Reference block not found: ${refId}`);
      return;
    }
    const nodes = blocks.map(
      (block) => blockToNode(this._schema, { ...block, id: block.id || v4() })
    );
    const tr = this.pm.createTransaction();
    tr.insert(insertPos, nodes);
    this.pm.dispatch(tr);
  }
  /**
   * Update a block's properties.
   *
   * @param block - The block ID or block object to update
   * @param update - Partial block with new props values
   *
   * @example
   * ```typescript
   * // Change a heading's level
   * editor.updateBlock('my-heading', { props: { level: 2 } });
   * ```
   */
  updateBlock(block, update) {
    const id = typeof block === "string" ? block : block.id;
    this.pm.doc.descendants((node, pos) => {
      if (node.attrs.id === id) {
        this.pm.setNodeAttrs(pos, { ...node.attrs, ...update.props });
        return false;
      }
    });
  }
  /**
   * Remove blocks from the document.
   *
   * @param blocks - Array of block IDs or block objects to remove
   *
   * @example
   * ```typescript
   * editor.removeBlocks(['block-1', 'block-2']);
   * ```
   */
  removeBlocks(blocks) {
    const ids = new Set(blocks.map((b) => typeof b === "string" ? b : b.id));
    const tr = this.pm.createTransaction();
    const toDelete = [];
    this.pm.doc.descendants((node, pos) => {
      if (ids.has(node.attrs.id)) {
        toDelete.push({ from: pos, to: pos + node.nodeSize });
      }
    });
    toDelete.reverse().forEach(({ from, to }) => tr.delete(from, to));
    this.pm.dispatch(tr);
  }
  // ===========================================================================
  // HISTORY (UNDO/REDO)
  // ===========================================================================
  /**
   * Undo the last change.
   *
   * @returns True if undo was performed, false if nothing to undo
   *
   * @example
   * ```typescript
   * editor.undo();
   * ```
   */
  undo() {
    return undo(this._pmView.state, this._pmView.dispatch);
  }
  /**
   * Redo the last undone change.
   *
   * @returns True if redo was performed, false if nothing to redo
   *
   * @example
   * ```typescript
   * editor.redo();
   * ```
   */
  redo() {
    return redo(this._pmView.state, this._pmView.dispatch);
  }
  // ===========================================================================
  // FORMATTING
  // ===========================================================================
  /** Toggle bold formatting on the current selection. */
  toggleBold() {
    return this.pm.toggleMark("bold");
  }
  /** Toggle italic formatting on the current selection. */
  toggleItalic() {
    return this.pm.toggleMark("italic");
  }
  /** Toggle underline formatting on the current selection. */
  toggleUnderline() {
    return this.pm.toggleMark("underline");
  }
  /** Toggle strikethrough formatting on the current selection. */
  toggleStrikethrough() {
    return this.pm.toggleMark("strikethrough");
  }
  /** Toggle inline code formatting on the current selection. */
  toggleCode() {
    return this.pm.toggleMark("code");
  }
  /**
   * Add a link to the current selection.
   *
   * @param href - The URL for the link
   * @param title - Optional title attribute
   */
  setLink(href, title) {
    this.pm.addMark("link", { href, title });
  }
  /** Remove link from the current selection. */
  removeLink() {
    this.pm.removeMark("link");
  }
  /**
   * Set text color on the current selection.
   *
   * @param color - CSS color value (hex, rgb, hsl, etc.)
   *
   * @example
   * ```typescript
   * editor.setTextColor('#ff0000');
   * editor.setTextColor('rgb(255, 0, 0)');
   * ```
   */
  setTextColor(color) {
    this.pm.addMark("textColor", { color });
  }
  /** Remove text color from the current selection. */
  removeTextColor() {
    this.pm.removeMark("textColor");
  }
  /**
   * Set background color on the current selection.
   *
   * @param color - CSS color value (hex, rgb, hsl, etc.)
   *
   * @example
   * ```typescript
   * editor.setBackgroundColor('#ffff00');
   * editor.setBackgroundColor('rgb(255, 255, 0)');
   * ```
   */
  setBackgroundColor(color) {
    this.pm.addMark("backgroundColor", { color });
  }
  /** Remove background color from the current selection. */
  removeBackgroundColor() {
    this.pm.removeMark("backgroundColor");
  }
  // ===========================================================================
  // TEXT ALIGNMENT
  // ===========================================================================
  /**
   * Set text alignment for the current block.
   *
   * @param alignment - The alignment value: 'left', 'center', or 'right'
   *
   * @example
   * ```typescript
   * editor.setTextAlign('center');
   * ```
   */
  setTextAlign(alignment) {
    var _a;
    const { $from } = this.pm.selection;
    for (let depth = $from.depth; depth > 0; depth--) {
      const node = $from.node(depth);
      if (node.isBlock && ((_a = node.type.spec.attrs) == null ? void 0 : _a.textAlign) !== void 0) {
        const pos = $from.before(depth);
        const tr = this.pm.createTransaction();
        tr.setNodeMarkup(pos, void 0, { ...node.attrs, textAlign: alignment });
        this.pm.dispatch(tr);
        return;
      }
    }
  }
  /**
   * Get the current text alignment of the block at selection.
   *
   * @returns The current alignment or 'left' as default
   */
  getTextAlign() {
    var _a;
    const { $from } = this.pm.selection;
    for (let depth = $from.depth; depth > 0; depth--) {
      const node = $from.node(depth);
      if (node.isBlock && ((_a = node.type.spec.attrs) == null ? void 0 : _a.textAlign) !== void 0) {
        return node.attrs.textAlign || "left";
      }
    }
    return "left";
  }
  // ===========================================================================
  // BLOCK TYPE CONVERSION
  // ===========================================================================
  /**
   * Convert the block at the current selection to a different type.
   *
   * @param type - The target block type (e.g., 'paragraph', 'heading')
   * @param attrs - Additional attributes for the new block type
   *
   * @example
   * ```typescript
   * // Convert current block to heading level 2
   * editor.setBlockType('heading', { level: 2 });
   *
   * // Convert back to paragraph
   * editor.setBlockType('paragraph');
   * ```
   */
  setBlockType(type, attrs = {}) {
    const { $from } = this.pm.selection;
    const nodeType = this._schema.nodes[type];
    if (!nodeType) {
      console.warn(`Unknown block type: ${type}`);
      return;
    }
    if (type === "bulletList" || type === "orderedList") {
      this._wrapInList(type);
      return;
    }
    let pos = null;
    let blockNode = null;
    for (let depth = $from.depth; depth > 0; depth--) {
      const node = $from.node(depth);
      if (node.isBlock && node.type.name !== "doc" && node.type.name !== "listItem") {
        pos = $from.before(depth);
        blockNode = node;
        break;
      }
    }
    if (pos === null || !blockNode) return;
    if (blockNode.type.name === "bulletList" || blockNode.type.name === "orderedList") {
      this._unwrapList(type, attrs);
      return;
    }
    if (type === "codeBlock") {
      this._convertToCodeBlock(pos, blockNode);
      return;
    }
    const newAttrs = {
      id: blockNode.attrs.id,
      ...attrs
    };
    const tr = this.pm.createTransaction();
    tr.setNodeMarkup(pos, nodeType, newAttrs);
    this.pm.dispatch(tr);
  }
  /**
   * Wrap the current block in a list.
   */
  _wrapInList(listType) {
    const { $from } = this.pm.selection;
    const listNodeType = this._schema.nodes[listType];
    const listItemType = this._schema.nodes.listItem;
    const paragraphType = this._schema.nodes.paragraph;
    if (!listNodeType || !listItemType) return;
    let blockPos = null;
    let blockNode = null;
    for (let depth = $from.depth; depth > 0; depth--) {
      const node = $from.node(depth);
      if (node.isBlock && node.type.name !== "doc" && node.type.name !== "listItem") {
        blockPos = $from.before(depth);
        blockNode = node;
        break;
      }
    }
    if (blockPos === null || !blockNode) return;
    if (blockNode.type.name === "bulletList" || blockNode.type.name === "orderedList") {
      if (blockNode.type.name !== listType) {
        const tr2 = this.pm.createTransaction();
        tr2.setNodeMarkup(blockPos, listNodeType, { id: blockNode.attrs.id });
        this.pm.dispatch(tr2);
      }
      return;
    }
    const content = blockNode.content;
    const blockId = blockNode.attrs.id;
    const paragraph = paragraphType.create({ id: v4() }, content);
    const listItem = listItemType.create({ id: v4() }, paragraph);
    const list = listNodeType.create({ id: blockId }, listItem);
    const tr = this.pm.createTransaction();
    tr.replaceWith(blockPos, blockPos + blockNode.nodeSize, list);
    this.pm.dispatch(tr);
  }
  /**
   * Unwrap list content to a different block type.
   */
  _unwrapList(targetType, attrs) {
    const { $from } = this.pm.selection;
    const targetNodeType = this._schema.nodes[targetType];
    if (!targetNodeType) return;
    let listPos = null;
    let listNode = null;
    for (let depth = $from.depth; depth > 0; depth--) {
      const node = $from.node(depth);
      if (node.type.name === "bulletList" || node.type.name === "orderedList") {
        listPos = $from.before(depth);
        listNode = node;
        break;
      }
    }
    if (listPos === null || !listNode) return;
    let content = this._schema.nodes.paragraph.create().content;
    if (listNode.firstChild && listNode.firstChild.firstChild) {
      content = listNode.firstChild.firstChild.content;
    }
    const newBlock = targetNodeType.create({ id: listNode.attrs.id, ...attrs }, content);
    const tr = this.pm.createTransaction();
    tr.replaceWith(listPos, listPos + listNode.nodeSize, newBlock);
    this.pm.dispatch(tr);
  }
  /**
   * Convert a block to a code block.
   */
  _convertToCodeBlock(pos, node) {
    const codeBlockType = this._schema.nodes.codeBlock;
    if (!codeBlockType) return;
    let text = "";
    node.content.forEach((child) => {
      if (child.isText) {
        text += child.text || "";
      }
    });
    const textNode2 = text ? this._schema.text(text) : null;
    const codeBlock = codeBlockType.create(
      { id: node.attrs.id },
      textNode2 ? [textNode2] : []
    );
    const tr = this.pm.createTransaction();
    tr.replaceWith(pos, pos + node.nodeSize, codeBlock);
    this.pm.dispatch(tr);
  }
  // ===========================================================================
  // COLUMNS
  // ===========================================================================
  /**
   * Insert a column layout at the current position.
   *
   * @param columns - Number of columns or array of widths (percentages)
   *
   * @example
   * ```typescript
   * // Insert 2 equal columns
   * editor.insertColumns(2);
   *
   * // Insert 3 columns with custom widths
   * editor.insertColumns([30, 40, 30]);
   * ```
   */
  insertColumns(columns) {
    const columnListType = this._schema.nodes.columnList;
    const columnType = this._schema.nodes.column;
    const paragraphType = this._schema.nodes.paragraph;
    if (!columnListType || !columnType || !paragraphType) {
      console.warn("Column nodes not available in schema");
      return;
    }
    let widths;
    if (typeof columns === "number") {
      widths = Array(columns).fill(100 / columns);
    } else {
      widths = columns;
    }
    const columnNodes = widths.map(
      (width) => columnType.create({ width }, paragraphType.create())
    );
    const columnList = columnListType.create(null, columnNodes);
    const { $from } = this.pm.selection;
    const insertPos = $from.end($from.depth);
    const tr = this.pm.createTransaction();
    tr.insert(insertPos, columnList);
    this.pm.dispatch(tr);
  }
  /**
   * Add a column to an existing column layout.
   *
   * @param columnListPos - Position of the columnList node
   * @param position - Where to add: 'start', 'end', or an index
   * @param width - Width for the new column (optional, will redistribute if not provided)
   */
  addColumn(columnListPos, position = "end", width) {
    const columnType = this._schema.nodes.column;
    const paragraphType = this._schema.nodes.paragraph;
    const columnList = this.pm.doc.nodeAt(columnListPos);
    if (!columnType || !paragraphType || !columnList || columnList.type.name !== "columnList") {
      return;
    }
    const tr = this.pm.createTransaction();
    let insertOffset;
    if (position === "start") {
      insertOffset = 1;
    } else if (position === "end") {
      insertOffset = columnList.nodeSize - 1;
    } else {
      let offset = 1;
      for (let i = 0; i < position && i < columnList.childCount; i++) {
        offset += columnList.child(i).nodeSize;
      }
      insertOffset = offset;
    }
    const newWidth = width ?? 100 / (columnList.childCount + 1);
    const newColumn = columnType.create({ width: newWidth }, paragraphType.create());
    tr.insert(columnListPos + insertOffset, newColumn);
    if (!width) {
      const evenWidth = 100 / (columnList.childCount + 1);
      let pos = columnListPos + 1;
      for (let i = 0; i < columnList.childCount; i++) {
        const col = columnList.child(i);
        tr.setNodeMarkup(tr.mapping.map(pos), void 0, { ...col.attrs, width: evenWidth });
        pos += col.nodeSize;
      }
    }
    this.pm.dispatch(tr);
  }
  /**
   * Remove a column from a column layout.
   *
   * @param columnPos - Position of the column to remove
   */
  removeColumn(columnPos) {
    const column = this.pm.doc.nodeAt(columnPos);
    if (!column || column.type.name !== "column") return;
    const $pos = this.pm.doc.resolve(columnPos);
    const columnListPos = $pos.before($pos.depth);
    const columnList = this.pm.doc.nodeAt(columnListPos);
    if (!columnList || columnList.type.name !== "columnList") return;
    const tr = this.pm.createTransaction();
    if (columnList.childCount <= 2) {
      const remainingContent = [];
      columnList.forEach((col, offset) => {
        const colPos = columnListPos + 1 + offset;
        if (colPos !== columnPos) {
          col.forEach((child) => remainingContent.push(child));
        }
      });
      tr.replaceWith(columnListPos, columnListPos + columnList.nodeSize, remainingContent);
    } else {
      tr.delete(columnPos, columnPos + column.nodeSize);
      const remainingCount = columnList.childCount - 1;
      const evenWidth = 100 / remainingCount;
      let pos = columnListPos + 1;
      columnList.forEach((col, offset) => {
        const colPos = columnListPos + 1 + offset;
        if (colPos !== columnPos) {
          tr.setNodeMarkup(tr.mapping.map(pos), void 0, { ...col.attrs, width: evenWidth });
        }
        pos += col.nodeSize;
      });
    }
    this.pm.dispatch(tr);
  }
  /**
   * Distribute column widths evenly in a column layout.
   *
   * @param columnListPos - Position of the columnList node
   */
  distributeColumns(columnListPos) {
    const columnList = this.pm.doc.nodeAt(columnListPos);
    if (!columnList || columnList.type.name !== "columnList") return;
    const evenWidth = 100 / columnList.childCount;
    const tr = this.pm.createTransaction();
    let pos = columnListPos + 1;
    columnList.forEach((col) => {
      tr.setNodeMarkup(pos, void 0, { ...col.attrs, width: evenWidth });
      pos += col.nodeSize;
    });
    this.pm.dispatch(tr);
  }
  // ===========================================================================
  // FOCUS & SELECTION
  // ===========================================================================
  /**
   * Focus the editor.
   *
   * @param position - Where to place the cursor: 'start', 'end', or a specific position
   */
  focus(position = "end") {
    this._pmView.focus();
    if (position === "start") {
      this.pm.setSelection(this.pm.createTextSelection(1));
    } else if (position === "end") {
      this.pm.setSelection(this.pm.createTextSelection(this.pm.doc.content.size - 1));
    } else {
      this.pm.setSelection(this.pm.createTextSelection(position));
    }
  }
  /** Remove focus from the editor. */
  blur() {
    this._pmView.dom.blur();
  }
  /** Whether the editor currently has focus. */
  get hasFocus() {
    return this._pmView.hasFocus();
  }
  // ===========================================================================
  // SERIALIZATION
  // ===========================================================================
  /** Serialize the document to JSON (alias for getDocument). */
  toJSON() {
    return this.getDocument();
  }
  /** Load document from JSON (alias for setDocument). */
  fromJSON(blocks) {
    this.setDocument(blocks);
  }
  // ===========================================================================
  // EVENTS
  // ===========================================================================
  /**
   * Subscribe to editor events.
   *
   * @param event - Event name: 'change', 'selectionChange', 'focus', 'blur', 'transaction'
   * @param handler - Callback function
   * @returns Unsubscribe function
   *
   * @example
   * ```typescript
   * const unsubscribe = editor.on('change', ({ blocks }) => {
   *   console.log('Document changed:', blocks);
   * });
   *
   * // Later: unsubscribe()
   * ```
   */
  on(event, handler) {
    let listeners = this._listeners.get(event);
    if (!listeners) {
      listeners = /* @__PURE__ */ new Set();
      this._listeners.set(event, listeners);
    }
    listeners.add(handler);
    return () => this.off(event, handler);
  }
  /** Unsubscribe from an editor event. */
  off(event, handler) {
    var _a;
    (_a = this._listeners.get(event)) == null ? void 0 : _a.delete(handler);
  }
  _emit(event, data) {
    var _a;
    (_a = this._listeners.get(event)) == null ? void 0 : _a.forEach((handler) => handler(data));
  }
  // ===========================================================================
  // LIFECYCLE
  // ===========================================================================
  /**
   * Mount the editor to a DOM element.
   *
   * Used by React/Vue adapters to attach the ProseMirror view to a container.
   * If the editor was created without an element, this must be called to render it.
   *
   * @param element - The container element to mount into
   */
  mount(element) {
    if (this._destroyed) {
      return;
    }
    if (this._pmView.dom.parentElement) {
      this._pmView.dom.parentElement.removeChild(this._pmView.dom);
    }
    element.appendChild(this._pmView.dom);
  }
  /**
   * Destroy the editor and clean up resources.
   *
   * After calling this, the editor instance cannot be reused.
   */
  destroy() {
    if (this._destroyed) return;
    this._destroyed = true;
    this._pmView.destroy();
    this._listeners.clear();
  }
  /** Whether the editor has been destroyed. */
  get isDestroyed() {
    return this._destroyed;
  }
  /** Whether the editor is in editable mode. */
  get isEditable() {
    return this._config.editable !== false;
  }
  /**
   * Enable or disable editing.
   *
   * @param editable - True to allow editing, false for read-only mode
   */
  setEditable(editable) {
    this._config.editable = editable;
    this._pmView.setProps({ editable: () => editable });
  }
}
export {
  AllSelection2 as AllSelection,
  BLOCK_ID_PLUGIN_KEY,
  BUBBLE_MENU_PLUGIN_KEY,
  DEFAULT_KEYBOARD_SHORTCUTS,
  DEFAULT_MARKS,
  DEFAULT_NODES,
  DRAG_DROP_PLUGIN_KEY,
  Decoration2 as Decoration,
  DecorationSet2 as DecorationSet,
  EditorState2 as EditorState,
  EditorView2 as EditorView,
  Fragment2 as Fragment,
  MEDIA_MENU_PLUGIN_KEY,
  Mark,
  MarkType,
  Node,
  NodeSelection2 as NodeSelection,
  NodeType,
  OpenBlockEditor,
  Plugin2 as Plugin,
  PluginKey2 as PluginKey,
  ProseMirrorAPI,
  ResolvedPos,
  SLASH_MENU_PLUGIN_KEY,
  Schema2 as Schema,
  Selection,
  Slice2 as Slice,
  TABLE_PLUGIN_KEY,
  TextSelection2 as TextSelection,
  Transaction,
  addColumnAfter,
  addColumnAtIndex,
  addColumnBefore,
  addRowAfter,
  addRowAtIndex,
  addRowBefore,
  areStylesInjected,
  blockToNode,
  blockquoteNode,
  blockquoteRule,
  blocksToDoc,
  boldMark,
  boldRule,
  bulletListNode,
  bulletListRule,
  calloutNode,
  checkListItemNode,
  checkListNode,
  closeSlashMenu,
  codeBlockNode,
  codeBlockRule,
  codeMark,
  columnListNode,
  columnNode,
  createBlockIdPlugin,
  createBubbleMenuPlugin,
  createChecklistPlugin,
  createDragDropPlugin,
  createInputRulesPlugin,
  createKeyboardShortcutsPlugin,
  createMediaMenuPlugin,
  createPlugins,
  createSchema,
  createSlashMenuPlugin,
  createTablePlugin,
  defaultConfig$1 as defaultConfig,
  deleteColumn,
  deleteColumnAtIndex,
  deleteMediaNode,
  deleteRow,
  deleteRowAtIndex,
  deleteTable,
  dividerNode,
  dividerRule,
  docNode,
  docToBlocks,
  embedNode,
  executeSlashCommand,
  filterSlashMenuItems,
  findTableContext,
  getBlockPosFromHandle,
  getDefaultSlashMenuItems,
  getTableInfo,
  goToNextCell,
  goToPreviousCell,
  headingNode,
  headingRule,
  hideBubbleMenu,
  hideMediaMenu,
  imageNode,
  injectStyles,
  inlineCodeRule,
  isInTable,
  isMarkActive,
  italicMark,
  italicRule,
  linkMark,
  listItemNode,
  moveBlock,
  nodeToBlock,
  orderedListNode,
  orderedListRule,
  paragraphNode,
  parseEmbedUrl,
  redo2 as redo,
  removeStyles,
  strikethroughMark,
  strikethroughRule,
  tableCellNode,
  tableHeaderNode,
  tableNode,
  tableRowNode,
  textNode,
  underlineMark,
  undo2 as undo,
  updateMediaAttrs
};
